% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/testNeutral.R
\name{testNeutral}
\alias{testNeutral}
\title{testNeutral}
\usage{
testNeutral(
  maf,
  patient.id = NULL,
  chrSilent = NULL,
  mutType = "All",
  use.indel = TRUE,
  withinTumor = FALSE,
  min.depth = 10,
  R2.threshold = 0.98,
  min.vaf = 0.1,
  max.vaf = 0.3,
  plot = TRUE,
  min.mut.count = 20
)
}
\arguments{
\item{maf}{the Maf object generated by readMaf}

\item{patient.id}{select the specific patients. Default: NULL, all patients are included.}

\item{chrSilent}{Select chromosomes needed to be dismissed. Default NULL.}

\item{mutType}{select Proper variant classification you need. Default "All". Option: "nonSilent".}

\item{use.indel}{Logical value. Whether to use INDELs besides somatic SNVs. Default: TRUE.}

\item{min.depth}{the minimun depth of coverage. Defalut: 10}

\item{R2.threshold}{the threshod of R2 to decide whether a tumor follows neutral evolution. Default: 0.98}

\item{min.vaf}{the minimum value of adjusted VAF value. Default: 0.1}

\item{max.vaf}{the maximum value of adjusted VAF value. Default: 0.3}

\item{plot}{logical, whether to print model fitting plot of each sample. Default: TRUE}

\item{min.mut.count}{the minimun number of subclonal mutations used to fit model. Default: 20}
}
\description{
Evaluat whether a tumor follows neutral evolution or under strong selection during the growth based on variant frequency distribution (VAF) of subclonal The subclonal mutant allele frequencies of a follow a simple power-law distribution predicted by neutral growth.
}
\examples{
testNeutral(maf)
}
