# MesKit

## Introduction

Tumor heterogeneity, as one of the characteristics of malignant tumors, often makes tumors' daughter cells exhibit molecular biological or genetic changes, thereby making differences in tumor features.
Therefore, understanding more about tumor heterogeneity could help us know more about how tumors develop and the structure of spatial cloning. 
Thus, more precise and accurate methods of tumor typing and treatments may be put forward. 
Currently, multi-site sampling is often used to evaluate the heterogeneity of tumors in the same patient.  
MesKit was aimed to deal with multi-sample WES-sequencing tumor data(each case should have at least one matched normal sample). 
For low-depth(below 100X) WES data, we highly recommend you perform mutation imputation to enable sensitive detection of mutations and improve inference of tumor's clonal architecture.
Meskit could summarize, visualize, evaluate the information contained in MAF files, and do functional and clonal analysis.
Further analysis is also available, you could get the results of the development process of tumor and mutational signature analysis outcome.


## Installation
### Via GitHub 
Install the latest version of this package by entering the following in R:

```r
install.packages("remotes")
remotes::install_github("Niinleslie/MesKit")
```


## Usage
## Prepare Input Data
PyClone requires that you generate prior information about the genotype of each mutation in the sample. In principle this depends on the algorithm(s) you use for copy number inference, tumour content prediction and your own beliefs. In practice specifying priors is challenging so PyClone includes some functionality to automatically generate the required prior information based on simpler input information.

To analyse with our packages you need to provide a set of input files, including:
  * An MAF file of tumors from the same person. Required.
  * Necessary details of samples. Required.
  * CCF data generated by PycClone (```loci.tsv``` & ```cluster.tsv```). Optional but recommended
  
  an optional but recommended clinical data associated with each sample/Tumor_Sample_Barcode in MAF.

## Tutorial Data

### The MAF files
MAF files contain many fields of information about chromosome and gene mutations and their annotations. The following fields are highly recommended to be contained in the MAF files.

Hugo_Symbol, Chromosome, Start_Position, End_Position, Variant_Classification, Variant_Type, Reference_Allele,	Tumor_Seq_Allele1, Tumor_Seq_Allele2,	Ref_allele_depth,	Alt_allele_depth,	VAF, CDS_Change, Protein_Change,Tumor_Sample_Barcode.

Example MAF file:
| Hugo_Symbol|	Chromosome | Start_Position |	End_Position |	Variant_Classification | Variant_Type |	Reference_Allele |	Tumor_Seq_Allele1 | Tumor_Seq_Allele2 |	Ref_allele_depth |	Alt_allele_depth |	VAF	| CDS_Change	| Protein_Change |	Tumor_Sample_Barcode |
|:-----| :------| :------ | :----- | :------ | :----- | :---- | :-----| :----- | :----- | :-------| :---- | :-----| :----- | :----- |
| LOC729737| 1 | 135207 |	135207	| RNA |	SNP |	C | C |	G |	40	| 4 | 0.0909 | NA |	NA | 311252-S |
|TTC34,ACTRT2| 1 | 2869474 | 2869474 |	IGR |INS | - | | CTCTCT |	43 | 8 | 0.1568 | NA |	NA | 311252-S |
|NBPF1|1 | 16908223 |	16908223 | Intron | SNP | T |	T |A|	142| 8 | 0.0533 | NA| NA | 311252-S|
|PRAMEF2 | 1 | 12921600 | 12921600 | Missense_Mutation | SNP | C |	C | T |73 |	3 |	0.0394 | c.C1391T |	p.P464L | 311252-S |

### Information of samples
Below is an example of the first four rows of one of the input files, sample_info.txt. The input files are located under the /inst/extdata/ folder.

> It should contain the sampleID, patientID, lesion and sampling time. Example data:
> tumors sampling across multiple spatially-distinct regions  
 |  sample  |  patient |  lesion |  time  |
 ---- | ------ | ------ | ------
 | 311252-S | 311252 |  S  |   -   |
 | 311252-V |  311252  |  V  |     -   |
 |311252-TC1 | 311252 |  TC  |     -   |
 |311252-TC2 | 311252 |  TC  |     -   |
 Note: "-" represent sampling at the same time
 
> tumors sampling across multiple time points
 |  sample  |  patient |  lesion |  time  |
 ---- | ------ | ------ | ------
 | WGC033391D | C1 |  -  |    T1   |
 | WGC033392D | C1 |  -  |    T2   |
 | WGC033393D | C1 |  -  |    T3   |
 | WGC033386D | C1 |  -  |    T4   |
 

## Maf objects

`readMaf` function reads MAF files along with specific sample information, stores it as an MAF object. The Maf class inherits from matools::MAF. We recommend users provide ccf data generated by PyClone for clone analysis.
```
	maf.File <- system.file("extdata/multi_lesion/maf", "311252.maf", package = "Meskit")
	sampleInfo.File <- system.file("extdata/multi_lesion", "sample_info.txt", package = "Meskit")
	pyCloneCluster <- system.file("extdata/multi_lesion/ccf", "311252.cluster.tsv", package = "Meskit")
	pyCloneLoci <- system.file("extdata/multi_lesion/ccf", "311252.loci.tsv", package = "Meskit")
	maf <- readMaf(patientID = "311252", mafFile = maf.File, sampleInfo = sampleInfo.File, refBuild = "hg19")
```

## ITH evaluation
MesKit offers several functions to estimate intratumoral heterogeneity(ITH) with bulk-tumor WES data, including `vaf.cluster`, `mathScore`. `vafCluster` function gives a general idea of heterogeneity by clustering variant allele frequencies(VAF). The function produces density distribution plot, which can show you the clusters of muatations with different Variant Allele Frequencies in all/selected samples.  Parameter `tsb` is set to select samples. 
`vafCluster(maf, show.MATH = T)`
Furthermore, you can use function `mathScore` to get accurate quantitative estimation of ITH.
`mathScore(maf)`
For each person, `mutSharedPrivate` function identify mutations shared by all the samples as shared mutations or clone, mutations shared by a subset of samples as partial-shared mutations, others which are unique to single sample as private mutations or subclone. This function provides users with an overview of mutations by producing a stack plot. The parameter "show.num" can be set to determine whether to show the numbers of muatations in the plot.
`mutSharedPrivate(maf, show.num = FALSE)`

## Clone analysis
`tumorClonesPlot` function can get use of ccf data generated from PyClone, visulizing the clone ccf and mutation ccf distrition in the form of radar plot and dotplot respectively. The paramete `clone.min.mut` decide the minium mutation number of a clone, and parameter `clone.min.aveCCF` specify the minium avergae CCF value of a clone.
```
	pyCloneCluster <- system.file("extdata/multi_lesion/ccf", "311252.cluster.tsv", package = "Meskit")
	pyCloneLoci <- system.file("extdata/multi_lesion/ccf", "311252.loci.tsv", package = "Meskit")
	tumorClonesPlot(ccfClusterFile = pyCloneCluster, ccfLociFile = pyCloneLoci)
```
As for tumors sampling across multiple time points, `cloneFishPlot` function infers clonal ordering based on Clonevol R package. It can also take inferred subclonal hierarchy results generated by SCHISM program as input(example). The metastatic evolvogram would be graphically represented using the R package FishPlot or TimeScape by setting `plotOption`. 
`cloneFishPlot()`


## njtree object
Based on Maf object, function `NJtree' construct phylogenetic tree into a S4 object - njtree. Njtree object contains neighbor-joining tree object(phylo), binary matrix of mutations, ccf matrix of mutations(if ccf data if available) and mutation info of each trunk/branch. Further, it can be used to perform functional analysis, mutational signature analysis and visualizing phylogentic tree.
`njtree <- NJtree(maf, use.indel = FALSE)`

## Function analysis
MesKit provide `GO.njtree` function and `Pathway.njtree` function to peform gene enrichment analysis. Both founctions return enrichment results of all mutations from the same person and of each branch mutation. By default, `Pathway.njtree` function perform kegg pathway enrichment analysis, you can also choose "reactome" via `pathway.type`. Barplot and dotplot can be controlled by `plotType` argument.
```
	GO.njtree(njtree, GO.type = "BP", savePlot = T, writeTable = T)
	Pathway.njtree(njtree, pathway.type = "KEGG", savePlot = T, writeTable = T)
```

## Mutation signature analysis

This method minimizes the total distance of the phylogenetic tree by determining the nearest or adjacent paired classification units.
The maf_file would be sorted and transformed into an NJtree object before plotting NJtree. Apart from that, the trunk and branches are colored differently, in order to better show the signatures.


`Mutational_sigs_tree` function could get mutational signature for each branch of phylogenetic tree. 


## GO anaysis
The GO database standardizes the gene products from functions, biological pathways and cell localization.
Through GO enrichment analysis, we can roughly understand where the differenal genes enrich, in what biological functions, pathways or cell localizations.
This function can offer a barplot and a dotplot to visualize the result of the GO enrichment, as well as some branch information extracted from the input njtree. You can get all/seleted type of analysis by setting the `type` parameter.
Also, `pval` and `qval` can be controlled to meet different needs.

`GO.njtree(njtree, GO.type = "BP", savePlot = T)`


## Pathway analysis
This function enables you to get metabolic pathway results and maps, which are analyzed based on the enriched genes.
The KEGG database links gene lists obtained from genomes that have been completely sequenced to higher levels of system functions at the cellular, species, and ecosystem levels.
And the Pathway analysis will offer you a clear plot showing enriched pathways.
You can choose between the kEGG analysis and Pathway analysis by controlling the `pathway.type` paramater.

`Pathway.njtree(njtree, pathway.type = "KEGG", savePlot = T)`


## Mutational Signature
`treeMutationalSig`
In order to get a phylogenetic tree, this function can define branches' set relationship by re-labeling their tumor sample barcode from the smallest set. 
And it will calcualte each branch's mutational signature weight according to cosmic reference and pick the maxium. 
Finally, a data frame includes each set/branch's mutational signature will be offered.

```	
	treeMutationalSig(maf, refBuild = "hg19")
```
Every cancer, as it progresses leaves a signature characterized by specific pattern of nucleotide substitutions. Alexandrov et.al have shown such mutational signatures, derived from over 7000 cancer samples 5. Such signatures can be extracted by decomposing matrix of nucleotide substitutions, classified into 96 substitution classes based on immediate bases surrounding the mutated base. Extracted signatures can also be compared to those validated signatures

## Phylogenetic tree plot
`plotPhyloTree` function can visulize phylogenetic tree from njtree object. If `show.mutSig` is set to ‘TRUE’(DEFAULT), muatation signature would be show in different colors of each trunk and branch. Either binary mutation heat map(indicate the presence or absence of a mutation) or ccf heat map can be displayed by parameter `heatmap.type`. Besides the njtree generated by function `NJtree`, this function also supports other rooted tree formats as input, such as `*.newick`, `*.beast`, `*PAML`(Root represent noraml sample). `ccf.mutation.id` munually specify which columns could be joint by `ccf.mutation.sep` to match mutation id in ccf data.

```
	plotPhyloTree(maf, phylotree.type = 'njtree', use.indel = FALSE, show.mutSig = TRUE)
	plotPhyloTree(phylotree.type = 'newick', phylotree.dir = "inst/extdata/")

```

## FAQ
