# MesKit
[TOC]

## 1. Introduction

Intra-tumor heterogeneity (ITH) is now thought to be a key factor contributing to the therapeutic failures and drug resistance, which have attracted increasing attention in the cancer research field. Here, we present an R package, MesKit, for characterizing cancer genomic ITH and inferring the tumorâ€™s evolutionary history. MesKit provides a wide range of analysis including ITH evaluation, enrichment, signature, clone evolution analysis via implementation of well-established computational and statistical methods.The source code and documents are freely available through Github (https://github.com/Niinleslie/MesKit). We also developed a shiny application to provide easier analysis and visualization. 


## 2. Prepare Input Data
To analyze with MesKit, you need to provide:

* An MAF file of multiple tumors samples from the same person (named `patientID.maf | patientID.maf.gz`). **Required**
* Necessary details of samples. **Required**
* CCF results generated by [PyClone](https://github.com/aroth85/pyclone) (`loci.tsv & cluster.tsv`). **Optional but recommended**
* Subclonal hierarchy inference generated by [SciClone](https://github.com/genome/sciclone). **Optional**

### 2.1 MAF files
Mutation Annotation Format (MAF) is a tab-delimited text file with aggregated mutation information from VCF Files. The following fields are required to be contained in the MAF files with MesKit.

**Mandatory fields:** `Hugo_Symbol, Chromosome, Start_Position, End_Position, Variant_Classification, Variant_Type, Reference_Allele, Tumor_Seq_Allele2, VAF, Tumor_Sample_Barcode`.

**Example MAF file**

| Hugo_Symbol|  Chromosome | Start_Position | End_Position |  Variant_Classification | Variant_Type | Reference_Allele |  Tumor_Seq_Allele1 | Tumor_Seq_Allele2 | VAF | Tumor_Sample_Barcode |
|:-----| :------| :------ | :----- | :------ | :----- | :---- | :-----| :----- | :----- | :-------|
| CDK11B| 1 | 1583495 | 1583495 | Intron| SNP | T | C | C |0.1875|SRR3670033|
|MIR4417,MIR4689|1| 5735104 | 5735104 | IGR |SNP | C | T |T|0.2142|SRR3670034|
|MST1L|1| 17083837|17083837|Missense_Mutation|SNP| T | C |C|0.1756|SRR3670035|

### 2.2 Information of samples
Below is an example of the first three rows of `sample_info.txt`. It should contain the sampleID, patientID, lesion and sampling time. 

- tumors sampling across multiple spatially-distinct regions

 |  sample  |  patient |  lesion |  time  |
 ---- | ------ | ------ | ------
 | SRR3670033 | HCC6046 |  L1  |   -   |
 | SRR3670034 | HCC6046 |  L2  |   -   |
 | SRR3670035 | HCC6046 |  L3  |   -   |

- tumors sampling across multiple time points 

 |  sample  |  patient |  lesion |  time  |
 ---- | ------ | ------ | ------
 | SRR3670033 | HCC6046 |  L1  |   T1   |
 | SRR3670034 | HCC6046 |  L2  |   T2   |
 | SRR3670035 | HCC6046 |  L3  |   T3   |

**Note:** `"-"` represents sampling at the same time or sampling from the same site. 


## 3. Installation
### Via GitHub 
Install the latest version of this package by typing the commands below in R console:
```r
install.packages("remotes")
remotes::install_github("Niinleslie/MesKit")
```

## 4. Start with the Maf object 

`readMaf` function create an MAF object by reading MAF files along with specific sample information. Parameter `refBuild` is used to set particular full genome sequences for Homo sapiens reference(`"hg19"` or `"hg38"`). Additionally, we recommend you provide cancer cell fraction (CCF) data generated by [PyClone](https://github.com/aroth85/pyclone) for visualizing clonal history. `ccf.mutation.id` manually specifies which columns could be joint by `ccf.mutation.sep` to match mutation id in ccf data.

```R
maf.File <- system.file("extdata/maf", "HCC6046.maf", package = "Meskit")
sampleInfo.File <- system.file("extdata", "HCC6046.sampleInfo.txt", package = "Meskit")
ccf.cluster.File <- system.file("extdata/ccf", "HCC6046.cluster.tsv", package = "Meskit")
ccf.loci <- system.file("extdata/ccf", "HCC6046.loci.tsv", package = "Meskit")


maf <- readMaf(mafFile = maf.File, sampleInfo = sampleInfo.File)

# Maf object with CCF information
maf <- readMaf(mafFile = maf.File, sampleInfo = sampleInfo.File, 
                ccfClusterTsvFile = ccf.cluster.File, 
                ccfLociTsvFile = ccf.loci.File,
                refBuild = "hg19")                
```


## 5. ITH evaluation
MesKit offers several functions to estimate intratumoral heterogeneity (ITH) with bulk-tumor WES data, including  `mathScore`, `vafCluster`, `mutSharedPrivate` and `JaccardSimilarity`. 

`mathScore` function estimates ITH in single sample using MATH approach. Typically, the higher the MATH score, the more heterogeneous is a tumor.
```R
mathScore(maf, minvaf = 0.02, maxvaf = 1)
```

`vafCluster` function clusters variant allele frequencies (VAFs) based on a Gaussian finite mixture model. The function produces a density distribution plot, depicting clusters of mutations with different VAFs in all/selected samples. We consider the number of clusters as a simple indicator of ITH.
Parameters `minVaf` and `maxVaf` can to set to filter mutations. In addition, we offer several options for defining the plotting styles. Parameter `plotOption` could be set as `"separate"`, `"combine"` or `"compare"`, which respectively means to output the frequency map in different documents, one documents or the ridgeline plot. 

```R
vafCluster(maf, plotOption = "combine", showMATH = T)
```

For each patient, `mutSharedPrivate` function identifies mutations shared by all samples as shared mutations (trunk mutations), mutations observed in a subset of samples as partial-shared mutations, others which are unique to single sample as private mutations (branch mutations). This function provides an overview of mutations using stack plots. Parameter `show.num` determines whether to show the numbers of mutations in the plot.
```R
mutSharedPrivate(maf, show.num = FALSE)
```

`JaccardIndex` function returns a grapgical display of Jaccard similarity correlation matrix. Jaccard similarity coefficients are calculated based on the ratio of shared mutations for paired-samples from single patient.
```
JaccardIndex(maf, type = "lower")
```


##6. Clone analysis
Subclones inference generated by [PyClone](https://github.com/aroth85/pyclone) or [SciClone](https://github.com/genome/sciclone) can be integrated to reconstruct and interpret subclonal architecture. 

`tumorClonesPlot` function can visualize the CCF distribution of subclones and mutations (`loci.tsv` and `cluster.tsv` generated by PyClone are required). Parameter `clone.min.mut` and `clone.min.aveCCF` specify the minimum mutation number of a subclone and the minimum average CCF value of a subclone, respectively.
```R
tumorClonesPlot(maf, clone.min.mut = 5, clone.min.aveCCF = 0.1)
```

For tumors sampling across multiple time points, `cloneFishPlot` function infers subclonal relationship based on R package, clonevol. With parameter `inferMethod`, `cloneFishPlot` can also take inferred subclonal hierarchy results generated by [SCHISM](https://github.com/KarchinLab/SCHISM) program as input (`*.cluster.cellularity` and `*.GA.consensusTree`). The metastatic evolvogram would be graphically represented using the R package FishPlot or TimeScape by setting `plotOption`.
```R
# clonevol method
cloneFishPlot(maf, inferMethod = "clonevol", plotOption = "fishplot")
cloneFishPlot(maf, inferMethod = "clonevol", plotOption = "timescape")

# read subclonal hierarchy results generated by SCHISM
cloneFishPlot(maf, inferMethod = "SCHISM", plotOption = "fishplot", schismCellularityFile, schismConsensusTree)
```

Using sciClone() function provided by R package SciClone, you could generate a scObject which contains the information about subclonal clusters and corresponding relationship. The scObject's RData file would be one of the inputs of tumorClonesPlot() in sciClone-input mode, and the following command would be an example for your usage.

```R
cloneFishPlot(maf=NULL, inferMethod="clonevol", plotOption="fishplot", sciCloneObject=sco, genotypePosition="space")
```

`ccfDensity` function can be used to print pairwise CCF plots of mutations which are identified across samples from single patient. Recent studies have used this method to infer which manner are metastases seeded in from primary tumors, monoclonal or polyclonal?
```R
ccfDensity(maf, show.density = TRUE)
```


##7. Phylogenetic tree object generation
Based on the Maf object, `NJtree` function reconstructs phylogenetic tree via the neighbor-joining method and stored in an njtree object, which is later utilized for functional analysis, mutational signature analysis and printing phylogenetic trees.
```R
njtree <- getNJtree(maf, use.indel = FALSE)
```

##8. Functional exploration 
MesKit conduct enrichment analysis of Gene ontology (GO) and pathway using [ClusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) R package. `GO.njtree` function and `Pathway.njtree` function  return enrichment results on both tree-level and branch-level. By default, `Pathway.njtree` function performs KEGG pathway enrichment analysis, besides, you can choose "reactome pathway" via parameter `pathway.type`. 
```R
# Required org.Hs.eg.db
library("org.Hs.eg.db")

# GO enrichment analysis
GO.njtree(njtree, GO.type = "BP", pval = 0.05, 
                pAdjustMethod = "BH", qval = 0.2, plotType = "dot", showCategory = 5)

# Pathway enrichment analysis
Pathway.njtree(njtree, pathway.type = "KEGG", pval = 0.05, 
                pAdjustMethod = "BH", qval = 0.2, plotType = "dot", showCategory = 5)
```

`treeMutationalSig` function identifies potential signatures which could be attributed to known mutational processes for trunks/branches of phylogenetic trees. This function is based on R package [deconstructSigs](https://github.com/raerose01/deconstructSigs) and you can set signatures matrix via `signature.ref` (`cosmic` or `nature2013`). Parameter `mutThreshold` specifies the minimum number of mutations (Default: 50) required to perform mutation signature analysis. Moreover, you could provide the driver genes list to `driverGenesFile` and the final result will add a column recording driver genes for particular mutation intersections. 

```R
# Requires BSgenome.Hsapiens.UCSC.hg19 if you set refBuild = "hg19"
library("BSgenome.Hsapiens.UCSC.hg19")
# Requires BSgenome.Hsapiens.UCSC.hg38 if you set refBuild = "hg38"
library("BSgenome.Hsapiens.UCSC.hg38")

treeMutationalSig(njtree, mutThreshold = 50, driverGenesFile = NULL)

```

##9. Phylogenetic tree visualization
`plotPhyloTree` function is aimed to visualize phylogenetic trees. If `show.mutSig = TRUE` (Default), mutation signature would be shown in different colors of each trunk and branch. Either binary mutation heatmap (indicates the presence or absence of a mutation) or ccf heatmap can be displayed by setting parameter `heatmap.type`. Besides njtree object, we also support other rooted tree formats as inputs, such as `*.newick"`, `*.beast`, `*.PAML` (Root represents normal samples). 

```R
plotPhyloTree(njtree, show.mutSig = TRUE, heatmap.type = 'CCF') 

# use other tree format 
newick.file <- system.file("extdata\phyloTree", "newick_example.nwk", package = "MesKit") 
plotPhyloTree(phylotree.dat = newick.file, phylotree.type = 'newick') 

beast.file <- system.file("extdata\phyloTree", "beast_example.beast", package ="MesKit") 
plotPhyloTree(phylotree.dat = beast.file, phylotree.type = 'beast') 

PAML.file <- system.file("extdata\phyloTree", "PAML_example.paml", package = "MesKit") 
plotPhyloTree(phylotree.dat = PAML.file, phylotree.type = 'PAML') 
```

## FAQ
