# MesKit

## Introduction

Intra-tumor heterogeneity(ITH) is now thought to have been a key factor contributing to the therapeutic failures and drug resistence. Today, with the rapid development of high-throughput sequencing technologies, cancer evolution and ITH have attracted increasing attention in the cancer research field. MesKit was aimed to deal with multi-sample WES-sequencing tumor data(each case should have at least one matched normal sample), which offers a host of analysis and visualization modules that are commonly used in cancer genomic ITH studies, including ITH evaluation, pathway, signature, clone evolution analysis, also allowing visualizing phylogenetic trees.


## Installation
### Via GitHub 
Install the latest version of this package by entering the following in R:
```r
install.packages("remotes")
remotes::install_github("Niinleslie/MesKit")
```


## Usage
### Prepare Input Data
To analyse with MesKit you need to provide a set of input files, including:
  * An MAF file of tumors from the same person(\*.maf|\*.maf.gz). Required.
  * Necessary details of samples. Required.
  * CCF data generated by PyClone (loci.tsv & cluster.tsv). Optional but recommended
  
### Tutorial Data
**The MAF files**

MAF files contain many fields of information about chromosome and gene mutations and their annotations. The following fields are highly recommended to be contained in the MAF files.

Hugo_Symbol, Chromosome, Start_Position, End_Position, Variant_Classification, Variant_Type, Reference_Allele,	Tumor_Seq_Allele1, Tumor_Seq_Allele2,	Ref_allele_depth,	Alt_allele_depth,	VAF, CDS_Change, Protein_Change,Tumor_Sample_Barcode.

**Example MAF file**

| Hugo_Symbol|  Chromosome | Start_Position | End_Position |  Variant_Classification | Variant_Type | Reference_Allele |  Tumor_Seq_Allele1 | Tumor_Seq_Allele2 | Ref_allele_depth |  Alt_allele_depth |  VAF | CDS_Change  | Protein_Change |  Tumor_Sample_Barcode |
|:-----| :------| :------ | :----- | :------ | :----- | :---- | :-----| :----- | :----- | :-------| :---- | :-----| :----- | :----- |
| LOC729737| 1 | 135207 | 135207  | RNA | SNP | C | C | G | 40  | 4 | 0.0909 | NA | NA | 311252-S |
|TTC34,ACTRT2| 1 | 2869474 | 2869474 |  IGR |INS | - | | CTCTCT | 43 | 8 | 0.1568 | NA |  NA | 311252-S |
|NBPF1|1 | 16908223 | 16908223 | Intron | SNP | T | T |A| 142| 8 | 0.0533 | NA| NA | 311252-S|
|PRAMEF2 | 1 | 12921600 | 12921600 | Missense_Mutation | SNP | C |  C | T |73 | 3 | 0.0394 | c.C1391T | p.P464L | 311252-S |


### Information of samples
Below is an example of the first four rows of sample_info.txt. It should contain the sampleID, patientID, lesion and sampling time The input files are located under the `/inst/extdata/` folder.

- tumors sampling across multiple spatially-distinct regions  
  
 |  sample  |  patient |  lesion |  time  |
 ---- | ------ | ------ | ------
 | 311252-S | 311252 |  S      |   -   |
 | 311252-V |  311252  |  V  |     -   |
 |311252-TC1 | 311252 |  TC  |     -   |
 |311252-TC2 | 311252 |  TC  |     -   |
 
- tumors sampling across multiple time points

 |  sample  |  patient |  lesion |  time  |
 ---- | ------ | ------ | ------
 | WGC033391D | C1 |  -  |     T1   |
 | WGC033392D | C1 |  -  |     T2   |
 | WGC033393D | C1 |  -  |     T3   |
 | WGC033386D | C1 |  -  |     T4   |

  **Note:** "-" represents sampling at the same time or sampling from the same site.


## Maf objects
`readMaf` function reads MAF files along with specific sample information, stores it as a MAF object. The Maf class inherits from matools::MAF. We also recommend users provide ccf data generated by PyClone for clone analysis.
```R
maf.File <- system.file("extdata/multi_lesion/maf", "311252.maf", package = "Meskit")
sampleInfo.File <- system.file("extdata/multi_lesion", "sample_info.txt", package = "Meskit")
pyCloneCluster <- system.file("extdata/multi_lesion/ccf", "311252.cluster.tsv", package = "Meskit")
pyCloneLoci <- system.file("extdata/multi_lesion/ccf", "311252.loci.tsv", package = "Meskit")
maf <- readMaf(patientID = "311252", mafFile = maf.File, sampleInfo = sampleInfo.File, refBuild = "hg19")
```

## ITH evaluation
MesKit offers several functions to estimate intratumoral heterogeneity(ITH) with bulk-tumor WES data, including `vaf.cluster`, `mathScore` and `mutSharedPrivate`. `vafCluster` function gives a general idea of heterogeneity by clustering variant allele frequencies(VAF). The function produces density distribution plot, depicting clusters of muatations with different Variant Allele Frequencies in all/selected samples. Parameter `tsb` is set to select samples. 
```R
vafCluster(maf, show.MATH = T)
```
Furthermore, you can use function `mathScore` to get accurate quantitative estimation of ITH.
```R
mathScore(maf)
```
For each person, `mutSharedPrivate` function identifies mutations shared by all samples as shared mutations or clone, mutations shared by a subset of samples as partial-shared mutations, others which are unique to single sample as private mutations or subclone. This function provides users with an overview of mutations by producing a stack plot. Parameter `show.num` can be set to determine whether to show the numbers of muatations in the plot.
```R
(maf, show.num = FALSE)
```

## Clone analysis
`tumorClonesPlot` function can get use of ccf data generated by PyClone, visulizing the clone ccf and mutation ccf distribution in the form of radar plot and dotplot respectively. Parameter `clone.min.mut` decides the minium mutation number of a clone and parameter `clone.min.aveCCF` specifies the minium avergae CCF value of a clone.
```R
pyCloneCluster <- system.file("extdata/multi_lesion/ccf", "311252.cluster.tsv", package = "Meskit")
pyCloneLoci <- system.file("extdata/multi_lesion/ccf", "311252.loci.tsv", package = "Meskit")
tumorClonesPlot(ccfClusterFile = pyCloneCluster, ccfLociFile = pyCloneLoci)
```
As for tumors sampling across multiple time points, `cloneFishPlot` function infers subclonal relationship based on R package, clonevol. With parameter `inferMethod`, `cloneFishPlot` can also take inferred subclonal hierarchy results generated by SCHISM program as input(*.cluster.cellularity and *.GA.consensusTree). The metastatic evolvogram would be graphically represented using the R package FishPlot or TimeScape by setting `plotOption`.
```R
# clonevol method
cloneFishPlot(maf, inferMethod="clonevol", plotOption="fishplot")
cloneFishPlot(maf, inferMethod="clonevol", plotOption="timescape")

# fishplot method
cloneFishPlot(maf, inferMethod="SCHISM", plotOption="fishplot", schismCellularityFile, schismConsensusTree)
cloneFishPlot(maf, inferMethod="SCHISM", plotOption="timescape", schismCellularityFile, schismConsensusTree)
```

## njtree object
Based on Maf object, `NJtree` function constructs phylogenetic tree into a S4 object——njtree. Njtree object contains neighbor-joining tree object(phylo), binary matrix of mutations, ccf matrix of mutations(if ccf data if available) and mutation info of each trunk/branch. Further, it can be used to perform functional analysis, mutational signature analysis and visualizing phylogentic tree.
```R
njtree <- NJtree(maf, use.indel = FALSE)
```

## Function analysis
MesKit provide `GO.njtree` function and `Pathway.njtree` function to peform gene enrichment analysis. Both founctions return enrichment results of all mutations from the same person and of each branch mutation. By default, `Pathway.njtree` function perform kegg pathway enrichment analysis, besides, you can choose "reactome" via `pathway.type`. Barplot and dotplot can be controlled by `plotType` argument.
```R
# GO enrichment analysis
GO.njtree(njtree, GO.type = "BP", savePlot = T, writeTable = T)
# Pathway enrichment analysis
Pathway.njtree(njtree, pathway.type = "KEGG", savePlot = T, writeTable = T)
```

Using `treeMutationalSig` function, you can identify potential signatures which could be attributed to known mutational processes for each branch/trunk of the NJtree. The function is implemented from R package deconstructSig. Two signatures matrices are available including "cosmic" and "nature2013" via `signature.ref`. Parameter `mutThreshold` specifies the mininum mutation number that required to perform mutation signature analysis(Default:50).
```R
treeMutationalSig(njtree, refBuild = "hg19")
```


## Phylogenetic tree visualization
`plotPhyloTree` function can visulize phylogenetic tree from njtree object. If `show.mutSig` is set to "TRUE"(DEFAULT), muatation signature would be shown in different colors of each trunk and branch. Either binary mutation heatmap(indicate the presence or absence of a mutation) or ccf heatmap can be displayed by parameter `heatmap.type`. Except for njtree object, this function also supports other rooted tree formats as input, such as "\*.newick", "\*.beast", "\*.PAML"(Root represents noraml sample). `ccf.mutation.id` munually specifies which columns could be joint by `ccf.mutation.sep` to match mutation id in ccf data.
```R
# plot phylogenetic tree based on njtree object
plotPhyloTree(maf, phylotree.type = 'njtree', use.indel = FALSE, show.mutSig = TRUE)`

# plot *.newick trees provided by users
newick.tree <- system.file("extdata/newick", "1.nwk", package = "MesKit")
plotPhyloTree(phylotree.dat = newick.tree, phylotree.type = 'newick')

# plot *.beast trees provided by users
beast.tree <- system.file("extdata/BEAST", "beast_mcc.tree", package = "MesKit")
plotPhyloTree(phylotree.dat = beast.tree , phylotree.type = 'beast')

# plot *.PAML trees provided by users
PAML.tree <- system.file("extdata/PAML_Baseml", "rst", package = "MesKit")
plotPhyloTree(phylotree.dat = brstfile , phylotree.type = 'PAML')
```

## FAQ
