# MesKit

## Introduction

Intra-tumor heterogeneity (ITH) is now thought to be a key factor contributing to therapeutic failures and drug resistance. Today, with the rapid development of high-throughput sequencing technologies, cancer evolution and ITH have attracted increasing attention in the cancer research field. MesKit was aimed to deal with multi-sample WES-sequencing tumor data (each case should have at least one matched normal sample), which offers a host of analysis and visualization modules that are commonly used in cancer genomic ITH studies, including ITH evaluation, pathway, signature, clone evolution analysis, also allowing visualizing phylogenetic trees.


## Installation
### Via GitHub 
Install the latest version of this package by typing the following in R console:
```r
install.packages("remotes")
remotes::install_github("Niinleslie/MesKit")
```


## Usage
### Prepare Input Data
To analyze with MesKit, you need to provide a set of input files, including:
  * An MAF file of tumours from the same person (\*.maf|\*.maf.gz). Required.
  * Necessary details of samples. Required.
  * CCF data generated by PyClone (loci.tsv & cluster.tsv). Optional but recommended

### Tutorial Data
**The MAF files**

MAF files contain many fields of information about chromosome and gene mutations and their annotations. The following fields are highly recommended to be contained in the MAF files.

Hugo_Symbol, Chromosome, Start_Position, End_Position, Variant_Classification, Variant_Type, Reference_Allele, Tumor_Seq_Allele1, Tumor_Seq_Allele2, Ref_allele_depth, Alt_allele_depth, VAF, CDS_Change, Protein_Change, Tumor_Sample_Barcode.

**MAF file example**

| Hugo_Symbol|  Chromosome | Start_Position | End_Position |  Variant_Classification | Variant_Type | Reference_Allele |  Tumor_Seq_Allele1 | Tumor_Seq_Allele2 | Ref_allele_depth |  Alt_allele_depth |  VAF | CDS_Change  | Protein_Change |  Tumor_Sample_Barcode |
|:-----| :------| :------ | :----- | :------ | :----- | :---- | :-----| :----- | :----- | :-------| :---- | :-----| :----- | :----- |
| LOC729737| 1 | 135207 | 135207  | RNA | SNP | C | C | G | 40  | 4 | 0.0909 | NA | NA | 311252-S |
|TTC34,ACTRT2| 1 | 2869474 | 2869474 |  IGR |INS | - | | CTCTCT | 43 | 8 | 0.1568 | NA |  NA | 311252-S |
|NBPF1|1 | 16908223 | 16908223 | Intron | SNP | T | T |A| 142| 8 | 0.0533 | NA| NA | 311252-S|
|PRAMEF2 | 1 | 12921600 | 12921600 | Missense_Mutation | SNP | C |  C | T |73 | 3 | 0.0394 | c.C1391T | p.P464L | 311252-S |


### Information of samples
Below is an example of the first four rows of sample_info.txt. It should contain the sampleID, patientID, lesion and sampling time. The input files are located under the `/inst/extdata/` folder.

- tumors sampling across multiple spatially-distinct regions  
  

 |  sample  |  patient |  lesion |  time  |
 ---- | ------ | ------ | ------
 | 311252-S | 311252 |  S      |   -   |
 | 311252-V |  311252  |  V  |     -   |
 |311252-TC1 | 311252 |  TC  |     -   |
 |311252-TC2 | 311252 |  TC  |     -   |

- tumors sampling across multiple time points

 |  sample  |  patient |  lesion |  time  |
 ---- | ------ | ------ | ------
 | WGC033391D | C1 |  -  |     T1   |
 | WGC033392D | C1 |  -  |     T2   |
 | WGC033393D | C1 |  -  |     T3   |
 | WGC033386D | C1 |  -  |     T4   |

  **Note:** "-" represents sampling at the same time or sampling from the same site.


## Maf objects
`readMaf` function reads MAF files along with specific sample information, stores it as a MAF object. To generate proper `patientID`, the names of MAF files should be correpsonding ID of patients. The Maf class inherits from maftools::MAF. CCF data generated by PyClone is recommended for clone analysis. Additionally, parameter `refBuild` is used to set particular full genome sequences for Homo sapiens (BSgenome.Hsapiens.UCSC) reference for your MAF data.

```R
maf.File <- system.file("extdata/multi_lesion/maf", "311252.maf", package = "Meskit")
sampleInfo.File <- system.file("extdata/multi_lesion", "sample_info.txt", package = "Meskit")
pyCloneCluster <- system.file("extdata/multi_lesion/ccf", "311252.cluster.tsv", package = "Meskit")
pyCloneLoci <- system.file("extdata/multi_lesion/ccf", "311252.loci.tsv", package = "Meskit")
maf <- readMaf(patientID = "311252", mafFile = maf.File, sampleInfo = sampleInfo.File, refBuild = "hg19")
```

## ITH evaluation
MesKit offers several functions to estimate intratumoral heterogeneity (ITH) with bulk-tumor WES data, including `vafCluster`, `mathScore` and `mutSharedPrivate`. The `vafCluster` function gives a general idea of heterogeneity by clustering variant allele frequencies (VAF). The function produces a density distribution plot, depicting clusters of mutations with different Variant Allele Frequencies in all/selected samples. Parameter `maf` needs the classMaf object generated by readMaf. To calculate the MATH score if required, the parameter `vafRange` would be considered as your targeted interval of VAF values, and `vafColumn` should be set as the column containing VAF values in your data set. After checking all data options, the following options would define the style and output of the plotting result. Parameter `plotOption` could be set as `"separate"`, `"combine"` or `"compare"`, which respectively means to output the frequency map in different documents, one documents or the ridgeline plot and the `themeOption` could be used to set proper colour schemes provided by ggsci. To save pictures in appropriate directories, you could set the `fileFormat` as "png", "pdf" or other file format supported by ggsave and pass the picture-saving directory to parameter `outputDir`.

```R
vafCluster(maf, vafRange=c(0,1), vafColumn="VAF", plotOption="combine", themeOption="aaas")
```
Furthermore, you can use function `mathScore` to get accurate quantitative estimations of ITH.
```R
mathScore(maf)
```
For each person, `mutSharedPrivate` function identifies mutations shared by all samples as shared mutations or clone, mutations shared by a subset of samples as partial-shared mutations, others which are unique to single sample as private mutations or subclone. This function provides users with an overview of mutations by producing a stack plot. Parameter `show.num` can be set to determine whether to show the numbers of mutations in the plot.
```R
mutSharedPrivate(maf, show.num = FALSE)
```

## Clone analysis
`tumorClonesPlot` function can get use of ccf data generated by PyClone, visualizing the clone ccf and mutation ccf distribution in the form of radar plot and dotplot respectively. Parameter `clone.min.mut` decides the minimum mutation number of a clone and parameter `clone.min.aveCCF` specifies the minimum average CCF value of a clone.
```R
pyCloneCluster <- system.file("extdata/multi_lesion/ccf", "311252.cluster.tsv", package = "Meskit")
pyCloneLoci <- system.file("extdata/multi_lesion/ccf", "311252.loci.tsv", package = "Meskit")
tumorClonesPlot(ccfClusterFile = pyCloneCluster, ccfLociFile = pyCloneLoci)
```
As for tumors sampling across multiple time points, `cloneFishPlot` function infers subclonal relationship based on an R package, clonevol. With the parameter `inferMethod`, function `cloneFishPlot` can also take inferred subclonal hierarchy results generated by SCHISM program as input (*.cluster.cellularity and *.GA.consensusTree). The metastatic evolvogram would be graphically represented using the R package FishPlot or TimeScape by setting `plotOption`.
```R
# clonevol method
cloneFishPlot(maf, inferMethod="clonevol", plotOption="fishplot")
cloneFishPlot(maf, inferMethod="clonevol", plotOption="timescape")

# fishplot method
cloneFishPlot(maf, inferMethod="SCHISM", plotOption="fishplot", schismCellularityFile, schismConsensusTree)
cloneFishPlot(maf, inferMethod="SCHISM", plotOption="timescape", schismCellularityFile, schismConsensusTree)
```

## njtree object
Based on the Maf object, `NJtree` function constructs phylogenetic tree into an S4 object——njtree. Njtree object contains neighbor-joining tree object (phylo), the binary matrix of mutations, ccf matrix of mutations (if ccf data is available) and mutation information of each trunk/branch. Furthermore, it can be used to perform functional analysis, mutational signature analysis, and visualizing the phylogenetic tree.
```R
njtree <- getNJtree(maf, use.indel = FALSE)
```

## Function analysis
MesKit provides `GO.njtree` function and `Pathway.njtree` function to perform gene enrichment analysis. Both functions return enrichment results of all mutations from the same person and of each branch mutation. By default, `Pathway.njtree` function performs KEGG pathway enrichment analysis. Besides, you can choose "Reactome" via `pathway.type`. Barplot and dotplot can be controlled by `plotType` argument.
```R
# GO enrichment analysis
GO.njtree(njtree, GO.type = "BP", savePlot = T, writeTable = T)
# Pathway enrichment analysis
Pathway.njtree(njtree, pathway.type = "KEGG", savePlot = T, writeTable = T)
```

Using `treeMutationalSig` function, you can identify potential signatures which could be attributed to known mutational processes for each branch/trunk of the NJtree. The function is implemented from R package deconstructSig. Two signature matrices are available including "cosmic" and "nature2013" via `signature.ref`. Parameter `mutThreshold` specifies the minimum mutation number that required to perform mutation signature analysis (Default:50). Moreover, you could provide the driver gene list to `driverGenesFile`, and the final result will add a column recording driver genes for particular mutation intersections. 
```R
treeMutationalSig(njtree, refBuild="hg19", driverGenesFile=NULL, mutThreshold=50)
```


## Phylogenetic tree visualization
`plotPhyloTree` function can visualize phylogenetic tree from njtree objects. If parameter `show.mutSig` is set to "TRUE" (DEFAULT), mutation signature would be shown in different colors of each trunk and branch. Either binary mutation heatmap (indicate the presence or absence of a mutation) or ccf heatmap can be displayed by parameter `heatmap.type`. Except for njtree objects, this function also supports other rooted tree formats as input, such as "\*.newick", "\*.beast", "\*.PAML" (Root represents noraml sample). `ccf.mutation.id` manually specifies which columns could be joint by `ccf.mutation.sep` to match mutation id in ccf data.
```R

# if use ccf  
plotPhyloTree(njtree, use.indel = T, heatmap.type = 'CCF') 

# use other tree format 
newick.file <- system.file("extdata/newick", "1.nwk", package="MesKit") 
plotPhyloTree(phylotree.dat = newick.file, phylotree.type = 'newick') 
beast.file <- system.file("extdata/BEAST", "sample.beast", package="MesKit") 
plotPhyloTree(phylotree.dat = beast.file , phylotree.type = 'beast') 
PAML.file <- system.file("extdata/PAML", "sample.paml", package="MesKit") 
plotPhyloTree(phylotree.dat = PAML.file , phylotree.type = 'PAML') 
```

## FAQ
