#MesKit

## Table of content   
- [MesKit](#idea)
    * [Introduction](#introduction)
    * [Installation](#installation)
    * [Prepare Input Data](#prepare-input-data)
        + [MAF files](#maf-files)
        + [Information of samples](#information-of-samples)   
    * [Start with the Maf object](#start-with-the-maf-object)
    * [ITH evaluation](#ith-evaluation)
    * [Clonal analysis](#clonal-analysis)
    * [Phylogenetic tree object generation](#phylogenetic-tree-object-generation)
    * [Functional exploration](#functional-exploration)
    * [Phylogenetic tree visualiztion](#phylogenetic-tree-visualiztion)
        + [A phylogenetic tree along with binary heatmap of mutations](#a-phylogenetic-tree-along-with-binary-heatmap-of-mutations)
        + [A phylogenetic tree along with CCF heatmap of mutations](#a-phylogenetic-tree-along-with-ccf-heatmap-of-mutations)
    * [Shiny app](#shiny-app)
- [Designers](#designers)
- [Credit](#credit)
- [Maintainer](#maintainer)
- [Copyright](#copyright)
- [Citation](#citation)

##Introduction
Intra-tumor heterogeneity (ITH) is now thought to be a key factor contributing to the therapeutic failures and drug resistance, which have attracted increasing attention in the cancer research field. Here, we present an R package, MesKit, for characterizing cancer genomic ITH and inferring the tumorâ€™s evolutionary history. MesKit provides a wide range of analysis including ITH evaluation, enrichment, signature, clone evolution analysis via implementation of well-established computational and statistical methods. 
The source code and documents are freely available through Github (https://github.com/Niinleslie/MesKit). We also developed a shiny application to provide easier analysis and visualization.

![overview](/vignettes/overview.png)

###Installation
Install the latest version of this package by typing the commands below in R console:

```R
install.packages("remotes")
remotes::install_github("Niinleslie/MesKit")
```

###Prepare Input Data
To analyze with MesKit, you need to provide:

+ An MAF file of multiple tumors samples from the same person `(named patientID.maf | patientID.maf.gz)`. __Required__
+ Necessary details of samples. __Required__
+ CCF results generated by PyClone `(loci.tsv & cluster.tsv)`. __Optional but recommended__

####MAF files
Mutation Annotation Format (MAF) files are tab-delimited text files with aggregated mutation information from VCF Files. The following fields are required to be contained in the MAF files with MesKit.

__Mandatory fields:__

`Hugo_Symbol, Chromosome`, `Start_Position`, `End_Position`, `Variant_Classification`, `Variant_Type`, `Reference_Allele`, `Tumor_Seq_Allele2`, `VAF`, `Tumor_Sample_Barcode`.

__Example MAF files__

```
##       Hugo_Symbol Chromosome Start_Position End_Position
## 1          CDK11B          1        1583495      1583495
## 2 MIR4417,MIR4689          1        5735104      5735104
## 3           MST1L          1       17083837     17083837
##   Variant_Classification Variant_Type Reference_Allele Tumor_Seq_Allele2
## 1                 Intron          SNP                T                 C
## 2                    IGR          SNP                C                 T
## 3      Missense_Mutation          SNP                T                 C
##      VAF Tumor_Sample_Barcode
## 1 0.1875           SRR3670033
## 2 0.2142           SRR3670034
## 3 0.1756           SRR3670035
```

####Information of samples
Below is an example of tumors sampling across multiple spatially-distinct regions. It should contain the `sampleID`, `patientID` and `lesion`.

```
##       sample patient lesion
## 1 SRR3670033      P1     T1
## 2 SRR3670034      P1     T2
## 3 SRR3670035      P1     T3
## 4 SRR3670036      P1     T4
## 5 SRR3670037      P1     T5
```


###Start with the Maf object

`readMaf` function create an MAF object by reading MAF files along with specific sample information. Parameter `refBuild` is used to set particular full genome sequences for Homo sapiens reference (`"hg19"` or `"hg38"`). Additionally, we recommend you provide cancer cell fraction (CCF) data generated by PyClone for visualizing clonal history. `ccf.mutation.id` manually specifies which columns could be joint by `ccf.mutation.sep` to match mutation id in ccf data.

```R
maf.File <- system.file("extdata/maf", "HCC6046.maf", package = "MesKit")
sampleInfo.File <- system.file("extdata", "HCC6046.sampleInfo.txt", package = "MesKit")
ccf.cluster.File <- system.file("extdata/ccf", "HCC6046.cluster.tsv", package = "MesKit")
ccf.loci.File <- system.file("extdata/ccf", "HCC6046.loci.tsv", package = "MesKit")


# Maf object with CCF information
maf <- readMaf(mafFile = maf.File, sampleInfo = sampleInfo.File, 
                ccfClusterTsvFile = ccf.cluster.File, 
                ccfLociTsvFile = ccf.loci.File,
                refBuild = "hg19")  
```

###ITH evaluation
MesKit offers several functions to estimate intratumoral heterogeneity (ITH) with bulk-tumor WES data, including `mathScore`, `vafCluster`, `mutSharedPrivate` and `JaccardSimilarity`.

`mathScore` function estimates ITH of single sample using MATH approach. Typically, the higher the MATH score, the more heterogeneous is a tumor.

```R
mathScore(maf, minvaf = 0.02, maxvaf = 1)
```

```
## $sampleLevel
##   Tumor_Sample_Barcode MATH_score
## 1           SRR3670033     90.860
## 2           SRR3670034     94.110
## 3           SRR3670035     89.225
## 4           SRR3670036     91.864
## 5           SRR3670037     84.010

```

`vafCluster` function clusters variant allele frequencies (VAFs) based on a Gaussian finite mixture model. The function produces a density distribution plot, depicting clusters of mutations with different VAFs in all/selected samples. We consider the number of clusters as a simple indicator of ITH.

Parameters `minVaf` and `maxVaf` can be set to filter mutations. In addition, we offer several options for defining the plotting styles. Parameter `plotOption` could be set as `"separate"`, `"combine"` or `"compare"`, which respectively means to output the frequency map in different documents, one documents or the ridgeline plot.

```R
vafCluster(maf, plotOption = "combine", showMATH = T)
```

![vafplot](/vignettes/vafplot.png)

For each patient, `mutSharedPrivate `function identifies mutations shared by all samples as shared mutations (trunk mutations), mutations observed in a subset of samples as partial-shared mutations, others which are unique to single sample as private mutations (branch mutations). This function provides an overview of mutations using stack plots. Parameter `show.num` determines whether to show the numbers of mutations in the plot.

```R
mutSharedPrivate(maf, show.num = FALSE)
```

![mutSharedPrivate](/vignettes/mutSharedPrivate.png)

`mutOncoTSG` function calculates the proportion of shared/partial-shared/private mutations that occurs in oncogenes and tumor suppressors genes (TSGs), separatedly. The default oncogenes list and TSGs list of pan-cancer were curated from previous publications and public databases.

```R
oncogeneListFile <- system.file("extdata/", "oncogene.list.txt", package = "MesKit")
tsgListFile <- system.file("extdata/", "TSG.list.txt", package = "MesKit")
mutOncoTSG(maf, oncogeneListFile, tsgListFile, show.percentage = TRUE)
```

![mutOncoTSG](/vignettes/mutOncoTSG.png)

`JaccardIndex` function returns a graphical display of Jaccard similarity correlation matrix. Jaccard similarity coefficients are calculated based on the ratio of shared mutations for paired-samples from single patient.

```R
JaccardIndex(maf, type = "lower")
```

![JaccardIndex](/vignettes/JaccardIndex.png)

###Clonal analysis
Subclones inference generated by PyClone can be integrated to reconstruct and interpret subclonal architecture.

`tumorClonesPlot` function can visualize the CCF distribution of subclones and mutations (`loci.tsv` and `cluster.tsv `generated by PyClone are required). Parameter `clone.min.mut` and `clone.min.aveCCF` specify the minimum mutation number of a subclone and the minimum average CCF value of a subclone, respectively.

```R
tumorClonesPlot(maf, clone.min.mut = 5, clone.min.aveCCF = 0.1)
```

![ClonePlot](/vignettes/ClonePlot.png)

`ccfDensity` function can be used to print pairwise CCF plots of mutations which are identified across samples from single patient. Recent studies have used this method to infer whether metastases are seeded in monoclonal manner or polyclonal manner from primary tumors.

```R
ccf.pairwise <- ccfDensity(maf, show.density = TRUE)
cowplot::plot_grid(ccf.pairwise[[1]],ccf.pairwise[[2]])
```

![ccfDensity](/vignettes/ccfDensity.png)

###Phylogenetic tree object generation
Based on the Maf object, `NJtree` function reconstructs phylogenetic tree via the neighbor-joining method and stored in an njtree object, which is later utilized for functional analysis, mutational signature analysis and printing phylogenetic trees.

```R
njtree <- getNJtree(maf)
```

###Functional exploration
MesKit conducts enrichment analysis of Gene ontology (GO) and pathway using __ClusterProfiler__ R package. `GO.njtree` function and `Pathway.njtree` function return enrichment results on both tree-level and branch-level. By default, `Pathway.njtree` function performs KEGG pathway enrichment analysis, besides, you can choose "reactome pathway" via parameter `pathway.type`.

```R
# Required org.Hs.eg.db
library("org.Hs.eg.db")

# GO enrichment analysis
GO.result <- GO.njtree(njtree, GO.type = "BP", pval = 0.05, 
                pAdjustMethod = "BH", qval = 0.2, plotType = "dot", showCategory = 5)
names(GO.result$GO.category)
```

```
##  [1] "SRR3670033"                                            
##  [2] "SRR3670034"                                            
##  [3] "SRR3670035"                                            
##  [4] "SRR3670036"                                            
##  [5] "SRR3670037"                                            
##  [6] "SRR3670033âˆ©SRR3670036"                                 
##  [7] "SRR3670035âˆ©SRR3670034"                                 
##  [8] "SRR3670037âˆ©SRR3670033âˆ©SRR3670036"                      
##  [9] "SRR3670037âˆ©SRR3670036âˆ©SRR3670033âˆ©SRR3670035âˆ©SRR3670034"
## [10] "All"
```

```R
names(GO.result$GO.plot)
```

```
## [1] "SRR3670036"                       "SRR3670037"                      
## [3] "SRR3670035âˆ©SRR3670034"            "SRR3670037âˆ©SRR3670033âˆ©SRR3670036"
```

```R
GO.result$GO.plot[["SRR3670036"]]
```

![GOplot](/vignettes/GOplot.png)

```R
pathway.result <- Pathway.njtree(njtree, pathway.type = "KEGG", pval = 0.05, 
                pAdjustMethod = "BH", qval = 1, plotType = "dot", showCategory = 5)
```

![pathwayplot](/vignettes/Pathwayplot.png)

`treeMutationalSig` function identifies potential signatures which could be attributed to known mutational processes for trunks/branches of phylogenetic trees. This function is based on R package deconstructSigs and you can set signatures matrix via `signature.ref` (`cosmic` or `nature2013`). Parameter `mutThreshold `specifies the minimum number of mutations (Default: 50) required to perform mutation signature analysis. Moreover, you could provide a driver genes list via `driverGenesFile` and the final result would contain an additional column recording driver genes for particular mutation intersections.

```R
# Requires BSgenome.Hsapiens.UCSC.hg19 if you set refBuild = "hg19"
library("BSgenome.Hsapiens.UCSC.hg19")
# Requires BSgenome.Hsapiens.UCSC.hg38 if you set refBuild = "hg38"
library("BSgenome.Hsapiens.UCSC.hg38")
```

```R
mutSig <- treeMutationalSig(njtree, mutThreshold = 50, driverGenesFile = NULL)
head(mutSig)
```

```
#  Branch                            Alias Mutation quantity  Signature  Signature weight Aetiology(Cosmic)
#1 SRR3670037âˆ©SRR3670036âˆ©SRR3670033âˆ©SRR3670035âˆ©SRR3670034  T  161 Signature 12   0.2103    Unknown
#2 SRR3670037âˆ©SRR3670033âˆ©SRR3670036                        B1 5   No Signature   0.0000    Unknown
#3 SRR3670035âˆ©SRR3670034                                   B2 19  No Signature   0.0000    Unknown
#4 SRR3670033âˆ©SRR3670036                                   B3 11  No Signature   0.0000    Unknown
#5 SRR3670037                                              B4 126 Signature 29   0.2048    Tobacco chewing habit
#6 SRR3670036                                              B7 104 Signature 5    0.5069    Unknown
#7 SRR3670035                                              B6 73  Signature 4    0.3161    Tobacco mutagens
#8 SRR3670034                                              B5 70  Signature 12   0.3407    Unknown
#9 SRR3670033                                              B8 102 Signature 12   0.2575    Unknown
```
### Phylogenetic tree visualization
`plotPhyloTree` function is aimed at visualizing phylogenetic trees. If `show.mutSig = TRUE` (Default), mutational signatures would be shown in different colors for each trunk and branch. Either binary mutation heatmap (indicates the presence or absence of a mutation) or ccf heatmap can be displayed by setting parameter `heatmap.type`.

####A phylogenetic tree along with binary heatmap of mutations
```R
plotPhyloTree(njtree, show.mutSig = TRUE, heatmap.type = 'binary') 
```

![tree](/vignettes/tree.png)

####A phylogenetic tree along with CCF heatmap of mutations

```R
plotPhyloTree(njtree, show.mutSig = TRUE, heatmap.type = 'CCF') 
```

![treewithccf](/vignettes/treewithccf.png)

### Shiny app

```R
library("shiny")
runApp(appDir = system.file("shiny",package = "MesKit"))
```

## Designers
* Jian Ren,renjian.sysu@gmail.com.
* Qi Zhao,zhaoqi3@mail2.sysu.edu.cn.
* Mengni Liu,liumn5@mail2.sysu.edu.cn
 
## Credit
This software were developed by:

* Mengni Liu,liumn5@mail2.sysu.edu.cn
* Jianyu Chen,chenjy327@mail2.sysu.edu.cn
* Chengwei Wang,

## Maintainer
Mengni Liu
## Copyright
Copyright Â© 2014-2018. RenLab from SYSUCC. All Rights Reserved
For more useful tools/applications, please go to renbal.org
## Citation
Citation (from within R, enter citation("MesKit")):

MesKit: a tool kit for dissecting cancer evolution from multi-region derived tumor biopsies via somatic mutations

