---
title: 'MesKit: Analyze and Visualize Multi-region Whole-exome Sequencing Data'
output:
  html_document:
    df_print: paged
    highlight: pygments
    self_contained: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
  pdf_document:
    toc: yes
    toc_depth: '5'
date: "`r Sys.Date()`"
---

<style type="text/css">
    .list-group-item.active,.list-group-item.active:focus,.list-group-item.active:hover {
    background-color: #007510;
    }
    body {
      font-family: Calibri, helvetica, sans-serif;
      font-size: 16px;
      line-height: 1.6;
    }
    h1 {
      font-size: 127%;
      color: #750400;
    }
    h2 {
      font-size: 121%;
      color: #750400;
    }    
    h3 {
      font-weight: bold;
      font-size: 18px;
      color: #750400;
      line-height: 2.4;
    }
    h4 {
      font-weight: bold;
      font-size: 17px;
      color: #750400;
      line-height: 2.1;
    }
    h5{
      font-size: 16px;
      font-weight: bold;
      color: #750400;
      line-height: 1.9;
    }
    a, a:hover {
      font-size: 16px;
      color: #5bc227;
    }

</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
  
### 1. Introduction  
  
Intra-tumor heterogeneity (ITH) is now thought to be a key factor that results in the therapeutic failures and drug resistance, which have arose increasing attention in cancer research. Here, we present an R package, MesKit, for characterizing cancer genomic ITH and inferring the history of tumor evolutionary. MesKit provides a wide range of analyses including ITH evaluation, enrichment, signature, clone evolution analysis via implementation of well-established computational and statistical methods. The source code and documents are freely available through Github (https://github.com/Niinleslie/MesKit). We also developed a shiny application to provide easier analysis and visualization.  
  

#### 1.1 Citation

In R console, enter ```citation("MesKit")```.  
  
_**MesKit: a tool kit for dissecting cancer evolution from multi-region derived tumor biopsies via somatic mutations (Submitted)**_
   
   
### 2. Prepare input Data  
  
To analyze with MesKit, you need to provide:  
  
* A MAF file of multi-region samples from patients (```*.maf / *.maf.gz```). **Required**
* Cancer cell fraction (CCF) data of somatic mutations. **Optional but recommended**
* A segmentation file. **Optional**
* The GISTIC outputs.  **Optional**

**Note:** ```Patient_ID``` and ```Tumor_Sample_Barcode``` should be consistant in all input files, respectively.
  
#### 2.1 MAF file  
  
Mutation Annotation Format (MAF) files are tab-delimited text files with aggregated mutations information from VCF Files. The input MAF file (or "*.maf.gz") of MesKit should have additional columns named ```Patient_ID``` and ```Tumor_ID``` on the basis of standard MAF files. Besides, as for the ```Variant_Classification```column, allowed values can be found at [Mutation Annotation Format Page](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/).   
The following fields are required to be contained in the MAF files with MesKit. 
   
   
**Mandatory fields:**  
  
```Hugo_Symbol```, ```Chromosome```, ```Start_Position```, ```End_Position```, ```Variant_Classification```, ```Variant_Type```, ```Reference_Allele```, ```Tumor_Seq_Allele2```, ```Ref_allele_depth```, ```Alt_allele_depth```, ```VAF```, ```Tumor_Sample_Barcode```, ```Patient_ID```, ```Tumor_ID``` 
   
   
**Note:** Multi-region samples from the a single tumor are indicated with the same ```Tumor_ID```, such as "primary", "metastasis" and "lymph". In addition, values in the ```Hugo_Symbol``` field are not necessarily from the HUGO database.

**Example MAF file**  
  
```{r echo=FALSE, paged.print=FALSE, rownames.print=FALSE}
MAFtable <- read.table(system.file("extdata", "HCC6046.maf", package = "MesKit"), header=TRUE)
extractLines <- rbind(MAFtable[1, ], MAFtable[383, ])
extractLines <- rbind(extractLines, MAFtable[739, ])
data.frame(extractLines, row.names = NULL)
```  
  
  
#### 2.2 CCF files  
  
By default, there are six mandatory fields in input CCF file: ```Patient_ID```,	```Tumor_Sample_Barcode```, ```Chromosome```, ```Start_Position```, ```CCF``` and  ```CCF_std```/```CCF_CI_High``` (required when identifying clonal/subclonal mutations). ```Chromosome``` field of mafFile and ccfFile should be in format (both in number or both start with "chr"). Notably, if CCF files contain other variants apart from SNVs, ```Reference_Allele``` and ```Tumor_Seq_Allele2``` should also be included in the input CCF files.


**Example CCF file**

```{r echo=FALSE, paged.print=FALSE, rownames.print=FALSE}
ccfInfo <- read.table(system.file("extdata", "HCC6046.ccf.tsv", package = "MesKit"), header = TRUE)
ccfInfo[1:5, ]
```
   

#### 2.3 Segmentation files

The segmentation file is a tab-delimited file with the following 6 or 7 columns:

* ```Patient_ID``` - ID of a patient
* ```Tumor_Sample_Barcode``` - Tumor sample barcode of samples
* ```Chromosome``` - chromosome name or ID
* ```Start_Position``` - genomic start position of segments (1-indexed)
* ```End_Position``` - genomic end position of segments (1-indexed)
* ```Segment_Mean/CopyNumber``` - segment mean value or absolute integer copy number

**Note:** Positions are in base pair units.
    
   
**Example Segmentation file**

```{r echo=FALSE, paged.print=FALSE, rownames.print=FALSE}
segInfo <- read.table(system.file("extdata", "HCC6046.seg.txt", package = "MesKit"), header = TRUE)
segInfo[1:5, ]
```  
 
### 3. Installation  
  
  
#### Via GitHub 

Install the latest version of this package by typing the commands below in R console:  
  
```{r eval=FALSE}
install.packages("remotes")
remotes::install_github("Niinleslie/MesKit")
```   
  
  
### 4. Start with the Maf object  
   
   
```readMaf``` function creates Maf/MafList objects by reading MAF files and cancer cell fraction (CCF) data (opitional but recommended). Parameter ```refBuild``` is used to set reference genome version for Homo sapiens reference (```"hg18"```, ```"hg19"``` or ```"hg38"```).
   
   
```{r message=FALSE}
library(MesKit)
```  

  
```{r}
maf.File <- system.file("extdata/", "HCC_LDC.maf", package = "MesKit")
ccf.File <- system.file("extdata/", "HCC_LDC.ccf.tsv", package = "MesKit")
# Maf object generation
maf <- readMaf(mafFile = maf.File, refBuild = "hg19") 

# Maf object with CCF information
maf <- readMaf(mafFile = maf.File,
               ccfFile = ccf.File,
               refBuild = "hg19")  
```  

### 5. Mutational landscape

#### 5.1 Mutational profile   

In order to explore the genomic alterations during cancer progression with multi-region sequencing approach, we provided ```classifyMut``` function to categorize mutations. The classification is based on shared pattern or clonal status (CCF data is required) of mutations, which can be specified by ```class``` option. Additionally, ```classByTumor``` can be used to reveal the mutational profile within tumors.  

```{r message=FALSE, fig.align='left', fig.width=11, fig.height=6.5}
# Driver genes of CRC collected from [IntOGen](https://www.intogen.org/search) (v.2020.2）
driverGene.file <- system.file("extdata/", "IntOGen-DriverGenes_HC.tsv", package = "MesKit")
driverGene <- as.character(read.table(driverGene.file)$V1)
mut.class <- classifyMut(maf, class =  "SP", patient.id = 'HCC8257')
head(mut.class)
```
  
```plotMutProfile``` function can visualize the mutational profile of tumor samples. 

```{r message=FALSE, fig.align='left', fig.width=11, fig.height=6.5}
plotMutProfile(maf, class =  "SP", geneList = driverGene)
```
 

#### 5.2 CNA profile

The ```plotCNA``` function can characterizes the CNA landscape across samples based on copy number data from segmentation algorithms. Besides, MesKit provides options to integrate GISTIC2 results, which can be obtained from [http://gdac.broadinstitute.org](http://gdac.broadinstitute.org). Please make sure the genome version based on these results is consistent with ```refBuild``` of the Maf/MafList object .


```{r message=FALSE}
# Read segment file
segCN <- system.file("extdata", "HCC_LDC.seg.txt", package = "MesKit")

# Reading gistic output files
all.lesions <- system.file("extdata", "LIHC_all_lesions.conf_99.txt", package = "MesKit")
amp.genes <- system.file("extdata", "LIHC_amp_genes.conf_99.txt", package = "MesKit")
del.genes <- system.file("extdata", "LIHC_del_genes.conf_99.txt", package = "MesKit")

seg <- readSegment(segFile = segCN, gisticAllLesionsFile = all.lesions,
                   gisticAmpGenesFile = amp.genes, gisticDelGenesFile = del.genes)
seg[1:5, ]
```


```{r, error=TRUE, fig.width=10, fig.align='left', fig.height=5}
plotCNA(seg, patient.id = c("HCC5647", "HCC8257", "HCC6690"))
```  

   
### 6. Measurement of intra-tumor heterogeneity  

#### 6.1 Within tumors

##### 6.1.1 MATH score calculation

The ```mathScore``` function estimates ITH per sample using mutant-allele tumor heterogeneity (MATH) approach [1](#refer). Typically, the higher the MATH score is, more heterogeneous a tumor sample is [2](#refer) [3](#refer). For MRS, this function can estimate the MATH score within tumor based on the merged VAF when ```withTumor = TRUE```. 
  

```{r warning=FALSE, fig.width=8, fig.height=5}
# calculate MATH score of each sample
MATH.sample <- mathScore(maf)
MATH.sample
```  
   
   
##### 6.1.2 AUC of CCF
   
   
The ```ccfAUC``` function estimates the tumor heterogeneity via the area under the curve (AUC) of the cumulative density from all cancer cell fractions per tumor. Tumors with higher AUC values are considered to be more heterogeneous [4](#refer). If the MAF data contains information from more than one patient, you can subset by specifying the corresponding Patient ID.     
   

```{r fig.height=5, fig.width=6, message=FALSE}
AUC <- ccfAUC(maf, patient.id = 'HCC8257')
names(AUC)
``` 

```{r fig.height=5, fig.width=6, message=FALSE}
AUC$AUC.value
```  


```{r fig.height=5, fig.width=5.5, message=FALSE, fig.align='left'}
# show cumulative density plot of CCF
AUC$CCF.density.plot
```   
   
   
##### 6.1.3 VAF clustering
   
```vafCluster``` function clusters variant allele frequencies (VAFs) based on a Gaussian finite mixture model. This function produces a density distribution plot, depicting clusters of mutations with different VAFs. We consider the number of clusters as a simple indicator of ITH. In addition, this function can read segmentation file to adjust CNA and LOH.
  
```{r message=FALSE, fig.height=5.5, fig.width=9.5}
clusterVaf <- vafCluster(maf, patient.id = 'HCC8257')
head(clusterVaf$cluster.data[, -1])
clusterPlots <- clusterVaf$cluster.plot
cowplot::plot_grid(plotlist = clusterPlots)
```  
   
   
#### 6.2 Between tumors


##### 6.2.1 Fixation index

Fixation Index (Fst) is a classical metric of population genetics, which can be used to measure the genetic divergence between regions [5](#refer) [6](#refer) [7](#refer). ```calFst``` function calculates the proportion of the total variance in allele frequency caused by frequency differences between populations.
   
   
```{r fig.align='left', fig.width=5.5, fig.height=5, fig.asp=1.0}
calFst(maf, patient.id = 'HCC8257', plot = TRUE, number.cex = 10)
```  


#### 6.2.2 Nei’s genetic distance
   
Nei’s genetic distance is used in population genetics to assess the similarity between populations, taking heterogeneity within populations into account. The function ```calNeiDist``` calculates Nei's distance of CCF among multi-region samples from the same patient for each sample-pair [8](#refer).   
   
```{r fig.align='left', fig.width=5.5, fig.height=5, fig.asp=1.0}
calNeiDist(maf, patient.id = 'HCC8257', number.cex = 10)
```  
   
   
### 7. Clonal origins inference 
   
Metastasis remains poorly understood despite its critical clinical significance, and the understanding of metastasis process can offer supplementary information for clinical treatments.   
      
   
#### 7.1 Pairwise CCF comparison

Distinct patterns of monoclonal versus polyclonal seeding based on the cancer cell fraction (CCF) of somatic mutations between sample/tumor pairs. ```compareCCF``` function prints pairwise CCF plots of mutations,  which are identified across samples from a single patient. Recently, this method has been widely used to deduce the potential metastatic route between different paired tumor lesions [9](#refer) [10](#refer).  
   
   
```{r fig.align='left', fig.width=5, fig.height=5, message=FALSE, warning=FALSE}
ccf.list <- compareCCF(maf)
ccf.pairwise <- ccf.list$HCC8257[["T3-T4"]]

# compare CCF using smoothScatter
graphics::smoothScatter(matrix(c(ccf.pairwise[,3], ccf.pairwise[,4]),ncol = 2),
              xlim = c(0,1), ylim = c(0,1),
              colramp = colorRampPalette(c("white", RColorBrewer::brewer.pal(9,"BuPu"))),
              xlab = "T3", ylab = "T4")

## show driver genes
gene.idx <- which(ccf.pairwise$Hugo_Symbol %in% driverGene) 
points(ccf.pairwise[gene.idx, 3:4], cex=0.6, col=2, pch=2)
text(ccf.pairwise[gene.idx, 3:4], cex=0.7, pos=1, ccf.pairwise$Hugo_Symbol[gene.idx])
title('HCC8257', cex.main = 1.2)
```  
   
   
#### 7.2 Jaccard similarity index
   
The Jaccard similarity index (JSI) can be used to calculate mutational similarity between regions, which is defined as the ratio of shared variants to all variants for sample pairs [11](#refer). Users can distinguish monoclonal versus polyclonal seeding in metastases (including lymph node metastases and distant metastases) via ```compareJSI``` function, and higher JSI values indicate the higher possibility of polyclonal seeding [12](#refer). 
   
   
```{r fig.align='left', fig.width=7, fig.height=7, fig.asp=1.0}
JSI.res <- compareJSI(maf, plot = FALSE, patient.id = 'HCC8257') 
names(JSI.res)
```

```{r fig.align='left', fig.width=7, fig.height=7, fig.asp=1.0}
# show the JSI result
JSI.res$JSI.multi
JSI.res$JSI.pair
```  
   
   
#### 7.3 Neutral evolution
   
The subclonal mutant allele frequencies of a tumor follow a simple power-law distribution predicted by neutral growth [13](#refer). Users can evaluate whether a tumor follows neutral evolution or not under strong selection via the ```testNeutral``` function. Tumors with R2 >= ```R2.threshold``` (Default: 0.98) are considered to follow neutral evolution. Besides, this function can also generate the model fitting plot of each sample if the argument ```plot``` is set to```TRUE```.
   
   
```{r message=FALSE, warning=FALSE, fig.height=4.5, fig.width=4.5}
neutralResult <- testNeutral(maf, min.mut.count = 10, patient.id = 'HCC8257')
neutralResult$neutrality.metrics
neutralResult$model.fitting.plot$T3
```     
   
   
   
### 8. Phylogenetic tree analysis
   
   
#### 8.1 Phylogenetic tree construction  
   
With MesKit, phylogenetic tree construction for each individual is based on the binary present/absence matrix of mutations across all tumor regions.  
Based on the Maf object, ```getPhyloTree``` function reconstructs phylogenetic tree in different methods, including "NJ" (Neibor-Joining) , "MP" (maximum parsimony),  "ML" (maximum likelihood), "FASTME.ols" and "FASTME.bal", which can be set by controlling the ```method``` parameter. The phylogenetic tree would be stored in a ```phyloTree/phyloTreeList``` object, and it can later be utilized for functional exploration, mutational signature analysis and tree visualization.  
   
   
```{r}
phyloTree <- getPhyloTree(maf, patient.id = "HCC8257", method = "NJ", min.vaf = 0.06)
```   
   
   
#### 8.2 Compare Phylogenetic trees
   
   
Comparison between phylogenetic trees can reveal consensus patterns of tumor evolution. The ```compareeTree``` function computes distances between phylogenetic trees constructed through different methods, and it returns a vector containing four distances by ```treedist``` from ```phangorn``` R package. See [treedist](http://evolution.genetics.washington.edu/phylip/doc/treedist.html) for details.
   
   
```{r, message=FALSE, warning=FALSE, fig.height=4, fig.width=6}
tree.NJ <- getPhyloTree(maf, patient.id = 'HCC8257', method = "NJ")
tree.MP <- getPhyloTree(maf, patient.id = 'HCC8257', method = "MP")

# compare phylogenetic trees constructed by two approaches
compareTree(tree.NJ, tree.MP, plot = TRUE)
```
   
   
   
#### 8.3 Functional exploration  
   
   
The ```phyloTree``` object can be further used to conduct functional exploration. Below is an example of how to perform KEGG enrichment analysis via [ClusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) by extracting genes in certain categories from a phylotree object.
  
  
```{r, message=FALSE, warning=FALSE, fig.width=5, fig.height=5}
library(org.Hs.eg.db)
library(clusterProfiler)

# Pathway enrichment analysis
HCC8257.branches <- phyloTree@mut.branches

# pathway enrichment for public mutated genes in patient HCC8257
HCC8257_Public <- HCC8257.branches[HCC8257.branches$Mutation_Type == "Public", ]
geneIDs = suppressMessages(bitr(HCC8257_Public$Hugo_Symbol, fromType="SYMBOL", 
                              toType=c("ENTREZID"), OrgDb="org.Hs.eg.db"))

KEGG_HCC8257_Public  = enrichKEGG(
  gene = geneIDs$ENTREZID,
  organism = 'hsa',
  keyType = 'kegg',
  pvalueCutoff = 0.05,
)

dotplot(KEGG_HCC8257_Public)
```  
  
  
#### 8.4 Mutational characteristics analysis

The sequence context of the base substitutions can be retrieved from the corresponding reference genome to construct a mutation matrix with counts for all 96 trinucleotide changes using “mut_matrix”. Subsequently, the 6 base substitution type spectrum can be plotted with “plot_spectrum” , which can be divided into several sample groups.

```mutTrunkBranch``` function calculates the fraction of branch/trunk mutations occurring in each of the six types of base substitution types (C>A, C>G, C>T, T>A, T>C, T>G) and conducts two-sided Fisher’s exact tests. For C>T mutations, it can be further classified into C>T at CpG sites and other sites by setting ```CT = TRUE```. This function provides option ```plot``` to print the distribution of branch/trunk mutations. Substitutions types with significant difference between trunk and branch mutations are marked with Asterisks. 

   
```{r message=FALSE}
# load the genome reference
library(BSgenome.Hsapiens.UCSC.hg19)
```
  
  
```{r, warning = FALSE, message=FALSE}
mutClass <- mutTrunkBranch(phyloTree, CT = TRUE, plot = TRUE)
names(mutClass)
```
   
   
```{r, warning = FALSE, message=FALSE, fig.height=5, fig.width=5}
mutClass$mutTrunkBranch.res
mutClass$mutTrunkBranch.plot
```
   
   
```triMatrix``` function returns a mutation count matrix of 96 trinucleotides based on somatic SNVs per sample. The ```fitSignatures``` function takes the matrix and known signature matrix as input to find the optimal nonnegative linear combination of mutation signatures to reconstruct matrix and calculate cosine similarity based on somatic SNVs. Besides, the signature matrix can be specified (```"cosmic_v2"```, ```"exome_cosmic_v3"``` or ```"nature2013"```) or provided by users via ```signaturesRef```. 

   
```{r, warning = FALSE, message=FALSE, fig.height=7, fig.width=10}
trimatrix_HCC8257 <- triMatrix(phyloTree, withinTumor = TRUE)
fit_HCC8257 <- fitSignatures(trimatrix_HCC8257)
```   

   
```plotMutSigProfile``` function can be utilized to visualize 96 trinucleodide mutational profile or reconstructed mutational profile.  
   
```{r, warning = FALSE, message=FALSE, fig.height=3, fig.width=10}
# Visualize the 96 trinucleodide mutational profile
plotMutSigProfile(trimatrix_HCC8257)[[1]]
```
  
   
```{r, fig.height=6, fig.width=10}
# Compare the reconstructed mutational profile with the original mutational profile
plotMutSigProfile(fit_HCC8257)[[1]]
```  


```{r, warning = FALSE, message=FALSE, fig.height=5, fig.width=8}
# Below plot shows cosine similarities between the mutational profile of each group and COSMIC signatures

library(ComplexHeatmap)
ComplexHeatmap::Heatmap(fit_HCC8257$HCC8257$cosine.similarity, name = "Cosine similarity")
```   
   
   
### 9. Phylogenetic tree visualization  
   
   
Users can visualize phylogenetic trees with annotations through ```plotPhyloTree``` function. By specifying ```branchCol``` parameter, the branches can be either colored according to classification of mutations or putative known signatures. Argument  ```show.bootstrap``` is provided to show the support values of internal nodes. Additionally, via ```mutHeatmap``` function, users can get a better depiction of mutational patterns mutations in tumor phylogeny based on MRS. Either binary mutation heatmap (indicates the presence or absence of mutations) or CCF heatmap can be displayed by running ```mutHeatmap``` function.


```{r, fig.align='left', fig.width=12, fig.height=5, message=FALSE, warning=FALSE}
# A phylogenetic tree along with binary and CCF heatmap of mutations 
phylotree_HCC8257 <- plotPhyloTree(phyloTree)
heatmapbinary_HCC8257 <- mutHeatmap(maf, use.ccf = FALSE, 
                                    patient.id = "HCC8257", min.vaf = 0.06)
heatmapCCF_HCC8257 <- mutHeatmap(maf, use.ccf = TRUE, 
                                 patient.id = "HCC8257", min.vaf = 0.06)
cowplot::plot_grid(phylotree_HCC8257, heatmapbinary_HCC8257, 
                   heatmapCCF_HCC8257, nrow = 1, rel_widths = c(1.3, 1, 1))
```
    
    

     
### 10. References {#refer}


1. Mroz, Edmund A, and James W Rocco. MATH, a novel measure of intratumor genetic heterogeneity, is high in poor-outcome classes of head and neck squamous cell carcinoma. Oral oncology vol. 49,3 (2013): 211-5.

2. Mroz EA , Tward AD , Pickering CR , et al. High intratumor genetic heterogeneity is related to worse outcome in patients with head and neck squamous cell carcinoma[J]. Cancer, 2013, 119(16):3034-3042. [https://doi.org/10.1002/cncr.28150](https://doi.org/10.1002/cncr.28150)

3. Mroz EA, Tward AM, Hammon RJ, Ren Y, Rocco JW (2015) Intra-tumor Genetic Heterogeneity and Mortality in Head and Neck Cancer: Analysis of Data from The Cancer Genome Atlas. PLoS Med 12(2): e1001786. [https://doi.org/10.1371/journal.pmed.1001786](https://doi.org/10.1371/journal.pmed.1001786)

4. Charoentong, P. et al. Pan-cancer immunogenomic analyses reveal genotypeimmunophenotype relationships and predictors of response to checkpoint blockade. Cell Rep. 18, 248–262 (2017)

5. Weir, B.S. and Cockerham, C.C. Estimating F-Statistics for the Analysis of Population Structure. Evolution, 1984, 38, 1358-1370. [http://dx.doi.org/10.2307/2408641](http://dx.doi.org/10.2307/2408641)

6. Bhatia G , Patterson N , Sankararaman S , et al. Estimating and interpreting FST: The impact of rare variants[J]. Genome Research, 2013, 23(9):1514-1521.

7. Holsinger K E , Weir B S . Genetics in geographically structured populations: defining, estimating and interpreting FST[J]. Nature Reviews Genetics, 2009, 10(9):639-650.

8. Puente X S, Pinyol M, Quesada V, et al. Whole-genome sequencing identifies recurrent mutations in chronic lymphocytic leukaemia[J]. Nature, 2011, 475(7354):101-105.

9. Christina Curtis et al. Quantitative evidence for early metastatic seeding in colorectal cancer, Nature Genetics (2019). [http://dx.doi.org/10.1038/s41588-019-0423-x](http://dx.doi.org/10.1038/s41588-019-0423-x)

10. Ruidong Xue et al. Genomic and Transcriptomic Profiling of Combined Hepatocellular and Intrahepatic Cholangiocarcinoma Reveals Distinct Molecular Subtypes. Cancer Cell, 2019, 35(6):932-947.

11. Makohon-Moore, A., Zhang, M., Reiter, J. et al. Limited heterogeneity of known driver gene mutations among the metastases of individual patients with pancreatic cancer. Nat Genet 49, 358–366 (2017). [https://doi.org/10.1038/ng.3764](https://doi.org/10.1038/ng.3764)

12. Hu Z, Li Z, Ma Z & Curtis C. Pan-cancer analysis of clonality and the timing of systemic spread in paired primary tumors and metastases. bioRXiv (2019). [http://dx.doi.org/10.1101/825240](http://dx.doi.org/10.1101/825240)

13. Williams, M., Werner, B., Barnes, C. et al. Identification of neutral tumor evolution across cancer types. Nat Genet 48, 238–244 (2016). [https://doi.org/10.1038/ng.3489](https://doi.org/10.1038/ng.3489)
  
  
  
### 11. Session Info
```{r}
sessionInfo()
```
  
  