---
title: 'MesKit: Analyze and visualize Multi-region Whole-exome Sequencing Data'
output:
  html_document:
    df_print: paged
    highlight: pygments
    self_contained: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
  pdf_document:
    toc: yes
    toc_depth: '5'
date: "`r Sys.Date()`"
---

<style type="text/css">
    .list-group-item.active,.list-group-item.active:focus,.list-group-item.active:hover {
    background-color: #007510;
    }
    body {
      font-family: Calibri, helvetica, sans-serif;
      font-size: 16px;
      line-height: 1.6;
    }
    h1 {
      font-size: 127%;
      color: #750400;
    }
    h2 {
      font-size: 121%;
      color: #750400;
    }    
    h3 {
      font-weight: bold;
      font-size: 18px;
      color: #750400;
      line-height: 2.4;
    }
    h4 {
      font-weight: bold;
      font-size: 17px;
      color: #750400;
      line-height: 2.1;
    }
    h5{
      font-size: 16px;
      font-weight: bold;
      color: #750400;
    }
    a, a:hover {
      font-size: 16px;
      color: #5bc227;
    }

</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
  
### 1. Introduction  
  
Intra-tumor heterogeneity (ITH) is now thought to be a key factor contributing to the therapeutic failures and drug resistance, which have attracted increasing attention in the cancer research. Here, we present an R package, MesKit, for characterizing cancer genomic ITH and inferring the history of tumor evolutionary . MesKit provides a wide range of analysis including ITH evaluation, enrichment, signature, clone evolution analysis via implementation of well-established computational and statistical methods. The source code and documents are freely available through Github (https://github.com/Niinleslie/MesKit). We also developed a shiny application to provide easier analysis and visualization.  
  

#### 1.1 Citation

In R console, enter ```citation("MesKit")```.  
  
_**MesKit: a tool kit for dissecting cancer evolution from multi-region derived tumor biopsies via somatic mutations**_
   
   
### 2. Prepare input Data  
  
To analyze with MesKit, you need to provide:  
  
* An MAF file of multiple tumors samples from the same person (named ```patientID.maf | patientID.maf.gz```). **Required**
* Cancer cell fraction (CCF) data of SSNVs. **Optional but recommended**
* A segmentation file. **Optional**
* The GISTIC output. **Optional**
  
#### 2.1 MAF files  
  
Mutation Annotation Format (MAF) files are tab-delimited text files with aggregated mutation information from VCF Files. The input MAF file (or "*.gz") should have an additional column named ```Patient_ID``` based on standard MAF file, which are allowed to include multiple patients. The following fields are required to be contained in the MAF files with MesKit. Besides, as for the ```Variant_Classification```column, allowed values can be found at [Mutation Annotation Format Page](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/).
  
**Mandatory fields:**  
  
```Hugo_Symbol```, ```Chromosome```, ```Start_Position```, ```End_Position```, ```Variant_Classification```, ```Variant_Type```, ```Reference_Allele```, ```Tumor_Seq_Allele2```, ```Ref_allele_depth```, ```Alt_allele_depth```, ```VAF```, ```Tumor_Sample_Barcode```, ```Patient_ID```.  

**Example MAF file**  
  
```{r echo=FALSE, paged.print=FALSE, rownames.print=FALSE}
MAFtable <- read.table(system.file("extdata", "HCC6046.maf", package = "MesKit"), header=TRUE)
extractLines <- rbind(MAFtable[1, ],MAFtable[383, ])
extractLines <- rbind(extractLines,MAFtable[739, ])
data.frame(extractLines, row.names = NULL)
```  
  
  
#### 2.2 CCF files  
  
There are 6 mandatory fields in for the CCF input file: ```Patient_ID```,	```Tumor_Sample_Barcode```, ```Chromosome```, ```Start_Position```, ```CCF```, ```CCF_std```/```CCF_CI_High``` (required when identify clonal/subclonal mutations), and the format used in the ```Chromosome``` field should be corresponding with the ```Chromosome``` field in the MAF file (both in number or both start with "chr").  


**Example CCF file**

```{r echo=FALSE, paged.print=FALSE, rownames.print=FALSE}
ccfInfo <- read.table(system.file("extdata", "HCC6046.CCF.txt", package = "MesKit"), header = TRUE)
ccfInfo[1:5, ]
```
   

#### 2.3 Segmentation files

The segmentation file is a tab-delimited file with the following 5 or 6 columns:

* ```Patient_ID``` - ID of a patient
* ```Tumor_Sample_Barcode``` - Tumor sample barcode of samples
* ```Chromosome``` - chromosome name or ID
* ```Start_Position``` - genomic start position of segments (1-indexed)
* ```End_Position``` - genomic end position of segments (1-indexed)
* ```CopyNumber``` - absolute integer copy number (or Segment_Mean)
Positions are in base pair units. The column names of a segmentation file should be constant.
    
   
**Example Segmentation file**

```{r echo=FALSE, paged.print=FALSE, rownames.print=FALSE}
segInfo <- read.table(system.file("extdata", "HCC6046.seg.txt", package = "MesKit"), header = TRUE)
segInfo[1:5, ]
```  
 
### 3. Installation  
  
  
#### Via GitHub 
Install the latest version of this package by typing the commands below in R console:  
  
```{r eval=FALSE}
install.packages("remotes")
remotes::install_github("Niinleslie/MesKit")
```   
  
  
### 4. Start with the Maf object  
  
```readMaf``` function creates an MAF object by reading MAF files and cancer cell fraction (CCF) data (Opitional but recommended). Parameter ```refBuild``` is used to set particular full genome sequences for Homo sapiens reference (```"hg18"```, ```"hg19"``` or ```"hg38"```). Besides, you can select the type of variant you are interested in by controlling the parameter ```mutType```, and the ```chrSilent``` parameter can be set to silent certain chromosome.
  
```{r message=FALSE}
library(MesKit)
```  

  
```{r}
maf.File <- system.file("extdata/", "HCC_LDC.maf", package = "MesKit")
ccf.File <- system.file("extdata/", "HCC_LDC.ccf.tsv", package = "MesKit")
# Maf object generation
maf <- readMaf(mafFile = maf.File, refBuild = "hg19") 

# Maf object with CCF information
maf <- readMaf(mafFile = maf.File,
               ccfFile = ccf.File,
               refBuild = "hg19")  
```  

### 5. Mutational landscape

#### 5.1 Mutational profile   

SSNVs from MRS are divided into shared (shared by all samples), partial-shared (p-shared, observed in a subset of samples), private (unique to a single sample). Additionally, ```classifyMut``` function can distinguish clonal and subclonal SSNVs in each sample if CCF data is available (Criterion: clonal – 95% CI overlaps with 1; subclonal – the upper bound of 95% CI is smaller than 1 [1](#refer)).  Option ```class``` is provided to select categorizations of variants.

```{r message=FALSE, fig.align='left', fig.width=10, fig.height=7, warning=FALSE}
driverGene.file <- system.file("extdata/", "IntOGen-DriverGenes_HC.tsv", package = "MesKit")
classifyMut(maf, class =  "SP", topGenesCount = 10, 
               driverGenesFile = driverGene.file, showColnames = FALSE)
```
   

#### 5.2 CNA profile
The ```plotCNA``` function can accept copy number information generated from segmentation programmers and plot the CNA profile. Besides, MesKit provides options to integrate GISTIC results, which can be obtained from [http://gdac.broadinstitute.org](http://gdac.broadinstitute.org). Please make sure the genome version based on these results is consistent with ```refBuild```.
   

```{r message=FALSE}
# Read segment file
segCN <- system.file("extdata", "HCC_LDC.seg.txt", package = "MesKit")

# Reading gistic output files
all.lesions <- system.file("extdata", "LICH_all_lesions.conf_99.txt", package = "MesKit")
amp.genes <- system.file("extdata", "LIHC_amp_genes.conf_99.txt", package = "MesKit")
del.genes <- system.file("extdata", "LIHC_del_genes.conf_99.txt", package = "MesKit")

seg <- readSegment(segCN.file = segCN, gisticAllLesionsFile = all.lesions,
                   gisticAmpGenesFile = amp.genes, gisticDelGenesFile = del.genes)
seg[1:5, ]
```


```{r, error=TRUE, fig.width=11, fig.align='center', fig.height=10}
plotCNA(seg)$ALL
```  

   
### 6. Measurement of intra-tumor heterogeneity  

#### 6.1 Within tumors

##### 6.1.1 MATH score calculation

The ```mathScore``` function estimates ITH of each sample of every single patient using MATH approach [2](#refer), and return both the data frame and a violin plot to better visualize the results. Typically, the higher the MATH score, the more heterogeneous a tumor sample is.   
  
```{r warning=FALSE, fig.width=10, fig.height=7}
MATH.scores <- mathScore(maf, min.vaf = 0.02, max.vaf = 1)
MATH.scores$MATH.df
MATH.scores$MATH.plot
```  

##### 6.1.2 AUC of CCF
The ```ccfAUC``` function estimates the tumor heterogeneity via the area under the curve (AUC) of the cumulative density from all cancer cell fractions per tumor. Tumors with higher AUC are considered more heterogeneous [3](#refer). If the maf data contains information from more than one patient, you can subste by specifying the corresponding Patient ID.     
   

```{r fig.height=5, fig.width=6, message=FALSE}
AUC <- ccfAUC(maf)
names(AUC)
``` 

```{r fig.height=5, fig.width=6, message=FALSE}
AUC$AUC.value
AUC$CCF.density.plot[['HCC5647']]
```  

```{r fig.height=5, fig.width=6, message=FALSE}
AUC$CCF.AUC.plot
```  

```{r fig.height=5, fig.width=6, message=FALSE}
# show cumulative density plot of CCF for patient HCC5647
AUC$CCF.density.plot[['HCC5647']]
```   
   
##### 6.1.3 VAF clustering
  
```vafCluster``` function clusters variant allele frequencies (VAFs) based on a Gaussian finite mixture model. This function produces a density distribution plot, depicting clusters of mutations with different VAFs in all/selected samples. We consider the number of clusters as a simple indicator of ITH.  Parameters ```min.vaf``` and ```max.vaf``` can be set to filter mutations. In addition, we offer several options for defining the plotting styles. Parameter `plotOption` could be set as ```"combine"``` or ```"compare"```, which means to generate a frequency map in different formats.  This function can read segment file to adjust for CNA and LOH.
  
```{r message=FALSE, fig.height=7.5, fig.width=8}
vaf.cluster <- vafCluster(maf, plotOption = "combine", showMATH = TRUE)
vaf.cluster[['HCC5647']]
```  
  

#### 6.2 Between tumors
 
##### 6.2.1 Shared patterns of variants
  
For each patient, ```mutSharedPattern``` function identifies mutations shared by all samples as shared mutations (trunk mutations), mutations observed in a subset of samples as partial-shared mutations, and others which are unique to a single sample as private mutations (branch mutations). This function provides an overview of mutations using stack plots. Parameter ```show.num``` determines whether to show the numbers of mutations in stack plots.  
  
```{r, fig.align='left', fig.width=7, fig.height=5.5}
mutShared <- mutSharedPattern(maf, show.num = FALSE)
mutShared[['HCC5647']]
```  
  

##### 6.2.2 Fixation index

Fixation Index (Fst) is a classical metric of population genetics, which can be used to measures the genetic divergence between regions [4](#refer) [5](#refer) [6](#refer). ```calFst``` function calculates the proportion of the total variance in allele frequency caused by frequency differences between populations.

```{r fig.align='left', fig.width=5.5, fig.height=5.5, fig.asp=1.0}
Fst.res <- calFst(maf, patient.id =  "HCC5647", plot = TRUE)
calFst(maf)[['HCC5647']]
```  

```{r fig.align='left', fig.width=5.5, fig.height=5.5, fig.asp=1.0}
names(Fst.res)
Fst.res$Fst.mean
Fst.res$Fst.pair
Fst.res$Fst.plot
```

### 7. Clonal origins inference 
Metastasis remains poorly understood despite its critical clinical importance. The understanding of metastasis process and its routes are a relevant factor to contribute in the advance of actual and future of clinical treatments. 

#### 7.1 Pairwise CCF comparison

Distinct patterns of monoclonal versus polyclonal seeding based on the cancer cell fraction (CCF) of SSNVs between sample pairs. ```ccfDensity``` function prints pairwise CCF plots of mutations which are identified across samples from a single patient. Recent studies have used this method to deduce the potential metastatic route between different paired tumor lesions [7](#refer) [8](#refer).  
  
```{r fig.align='left', fig.width=10, fig.height=5, message=FALSE, warning=FALSE}
ccf.pairwise <- ccfDensity(maf, show.density = TRUE)
cowplot::plot_grid(ccf.pairwise[['HCC5647']][[1]], 
                   ccf.pairwise[['HCC5647']][[2]], ncol = 2)
```  
   

#### 7.2 Nei’s genetic distance
Nei’s genetic distance is used in population genetics to assess the similarity between populations, taking into account heterogeneity within populations. The function ```calNeiDist``` calculates Nei's distance of CCF among multi-region samples from the sanme patient. for each sample-pair [9](#refer).   
   
```{r fig.align='left', fig.width=6, fig.height=5, fig.asp=1.0}
Neidist <- calNeiDist(maf)
names(Neidist)
```  

```{r fig.align='left', fig.width=6, fig.height=5, fig.asp=1.0}
# show the correlation matrix of Nei's dist of Patient HCC5647
Neidist$Nei.dist[['HCC5647']]
# show the correlation plot of Nei's dist of Patient HCC5647
Neidist$Nei.plot[['HCC5647']]
```    
   
#### 7.3 Jaccard similarity index

The Jaccard similarity index (JSI) can be used to calculate mutational similarity between lesions, which is defined as the ratio of shared variants to all variants for sample pairs (Limited heterogeneity of known driver gene mutations among the metastases of individual patients with pancreatic cancer). Users can distinguish monoclonal versus polyclonal seeding in metastases (including lymph node metastases and distant metastases) via ```calJSI``` function. Higher JSI values are more likely indicates polyclonal seeding [10](#refer). 

  
```{r fig.align='left', fig.width=6, fig.height=5, fig.asp=1.0}
JSI.res <- calJSI(maf, use.circle = TRUE)
names(JSI.res)
```  

```{r fig.align='left', fig.width=6, fig.height=5, fig.asp=1.0}
# show the correlation matrix of JSI of Patient HCC5647
JSI.res$JSI.dist[['HCC5647']]
# show the correlation plot of JSI of Patient HCC5647
JSI.res$JSI.plot[['HCC5647']]
```  

#### 7.4 Neutral evolution
The subclonal mutant allele frequencies of a follow a simple power-law distribution predicted by neutral growth [11](#refer). Users can evaluate whether a tumor follows neutral evolution or under strong selection via the ```testNeutral``` function. Tumors of R2 >= ```R2.threshold``` (Default: 0.98) were considered to follow neutral evolution. This function also prints the model fitting plot of each sample if the argument ```plot``` is ```TRUE```.

```{r, message=FALSE, warning=FALSE, fig.height=8, fig.width=10}
neutralResult <- testNeutral(maf, plot = TRUE)

rbind(neutralResult[['HCC5647']]$HCC5647_T1$model.fitting.out, 
      neutralResult[['HCC5647']]$HCC5647_T3$model.fitting.out, 
      neutralResult[['HCC5647']]$HCC5647_T4$model.fitting.out, 
      neutralResult[['HCC5647']]$HCC5647_T5$model.fitting.out)

cowplot::plot_grid(neutralResult[['HCC5647']]$HCC5647_T1$model.fitting.plot, 
                   neutralResult[['HCC5647']]$HCC5647_T3$model.fitting.plot, 
                   neutralResult[['HCC5647']]$HCC5647_T4$model.fitting.plot, 
                   neutralResult[['HCC5647']]$HCC5647_T5$model.fitting.plot, ncol= 2, scale = 0.95)
```  


### 8. Phylogenetic tree analysis

#### 8.1 Phylogenetic tree construction  
  
Based on the Maf object, ```getPhyloTree``` function reconstructs phylogenetic tree via different methods, including "NJ" (Neibor-Joining) , "MP" (maximum parsimony) and "ML" (maximum likelihood) , which can be set by controlling the ```method``` parameter, and stored in a ```phyloTree``` object, which is later utilized for functional analysis, mutational signature analysis and printing phylogenetic trees.  

```{r}
phylotree.list <- getPhyloTree(maf, method = "NJ")
```

#### 8.2 Compare Phylogenetic trees
Comparison between phylogenetic trees can reveal concensus patterns of tumor evolution. The ```compareeTree``` function computes distances between phylogenetic trees constructed through different methods. This function returns a vector containing four distance by ```treedist``` from ```phangorn``` R package. See [treedist](http://evolution.genetics.washington.edu/phylip/doc/treedist.html) for details.

```{r, message=FALSE, results=FALSE}
tree.NJ <- getPhyloTree(maf, method = "NJ")
tree.MP <- getPhyloTree(maf, method = "MP")
```

```{r, warning = FALSE, message=FALSE}
# Requires BSgenome.Hsapiens.UCSC.hg19 if you set refBuild = "hg19"
library("BSgenome.Hsapiens.UCSC.hg19")
```

```{r, fig.width=9}
compareTree(tree.NJ[['HCC5647']], tree.MP[['HCC5647']], plot = TRUE)
```  

#### 8.3 Functional exploration  
  
MesKit conducts enrichment analysis of Gene ontology (GO) and pathway using [ClusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) R package. ```treeGO``` function and ```treePathway``` function returns enrichment results on both tree-level and branch-level. By default, ```treePathway``` function performs enrichment analysis of KEGG pathway and ```Reactome``` pathway is also available. Users can conduct enrichment of functional driver gene mutations specifically by providing a driver gene list.
  
  
##### 8.3.1 GO enrichment analysis
```{r, message=FALSE, warning=FALSE}
# Required org.Hs.eg.db
library("org.Hs.eg.db")

# GO enrichment analysis
GO.result <- treeGO(phylotree.list, driverGenesFile = driverGene.file, 
                    GO.type = "BP", pval = 0.05, pAdjustMethod = "BH", 
                    qval = 0.2, plotType = "dot", showCategory = 5)
names(GO.result[['HCC5647']]$GO.category)
names(GO.result[['HCC5647']]$GO.plot)
``` 
  
```{r, message=FALSE}
GO.result[['HCC5647']]$GO.category[['All']]
```
  
```{r, fig.align='left', fig.height=4.5, fig.width=7, message=FALSE}
GO.result[['HCC5647']]$GO.plot[['All']]
```
  
  
##### 8.3.2 Pathway enrichment analysis
```{r, message=FALSE, warning=FALSE}
# Pathway enrichment analysis
pathway.result <- treePathway(phylotree.list, driverGenesFile = driverGene.file, 
                              pathway.type = "KEGG", pval = 0.05, pAdjustMethod = "BH", 
                              qval = 1, plotType = "dot", showCategory = 5)
```  
  
  
```{r message=FALSE}
pathway.result[['HCC5647']]$pathway.category[['All']]
```   
  
  
#### 8.4 Mutational signature analysis

```mutTrunkBranch``` function calculates the fraction of branch/trunk mutations occurring in each of the six types of base substitution types (C>A, C>G, C>T, T>A, T>C, T>G) and conducts paired Wilcoxon test. The confidence level can be tuned by parameter ```conf.level```. And the results of Wilcoxon rank-sum test are shown in the ```Significance``` column, where ```*``` indicates the significant differences exists between trunk and branch. Besides, this function provides option ```plot``` to print the distribution of branch/trunk mutations.   


```{r, warning = FALSE, message=FALSE, fig.height=6, fig.width=8}
mutClass <- mutTrunkBranch(phylotree.list, conf.level = 0.95, plot = TRUE)
names(mutClass)
```
```{r, warning = FALSE, message=FALSE, fig.height=6, fig.width=8}
mutClass[[1]][["HCC5647"]]
mutClass[[2]][["HCC5647"]]
```

```treeMutSig``` function identifies potential signatures for trunks/branches of phylogenetic trees, which is implemented based on [deconstructSigs](https://github.com/raerose01/deconstructSigs) R package.  Users can set signatures matrix via ```signaturesRef``` (```cosmic``` or ```nature2013```). Parameter ```mutThreshold``` specifies the minimum number of mutations (Default: 15) required to perform mutation signature analysis. 
   
```{r, warning = FALSE, message=FALSE, fig.height=7, fig.width=10}
tree.mutSig <- treeMutSig(phylotree.list,
                          driverGenesFile = driverGene.file,
                          plot = TRUE)
names(tree.mutSig)
```   
   
   
   
```{r, warning = FALSE, message=FALSE, fig.height=7, fig.width=10}
# summary of mutational signature of Patient HCC5647
tree.mutSig$mutSig.summary[["HCC5647"]]

# plot of mutational signature of Patient HCC5647
tree.mutSig$mutSig.plot[["HCC5647"]]
```

      
### 9. Phylogenetic tree visualization  
   
Users can visualize phylogenetic trees with annotations by ```plotPhyloTree``` function. The trunks/branches will be colored according to corresponding mutational signatures when ```show.mutSig = TRUE``` (Default). Either binary mutation heatmap (indicates the presence or absence of mutations) or CCF heatmap can be displayed by setting parameter `heatmap.type`. Argument  ```show.bootstrap``` is provided to show how much support for internal nodes.


Numbers above branches are bootstrap values.
Note that some nodes receive bootstrap values of 100%, indicating
strong support for these nodes, whereas other nodes receive much
weaker support
  
 
```{r, fig.align='left', fig.width=8.5, fig.height=5, message=FALSE, warning=FALSE}
# A phylogenetic tree along with binary heatmap of mutations 
phyloTree.binary <-  plotPhyloTree(phylotree.list, show.mutSig = TRUE, heatmap.type = "binary") 
phyloTree.binary[['HCC5647']]
```
  
```{r, fig.align='left', fig.width=8, fig.height=5, message=FALSE, warning=FALSE}
# 9.2 A phylogenetic tree along with CCF heatmap of mutations 
phyloTree.CCF <-  plotPhyloTree(phylotree.list, show.mutSig = TRUE, heatmap.type = "CCF") 
phyloTree.CCF[['HCC5647']]
```
     
### 10. References {#refer}

1. Mcgranahan N , Favero F , De Bruin E C , et al. Clonal status of actionable driver events and the timing of mutational processes in cancer evolution[J]. Science Translational Medicine, 2015, 7(283):283ra54-283ra54.

2. Mroz, Edmund A, and James W Rocco. MATH, a novel measure of intratumor genetic heterogeneity, is high in poor-outcome classes of head and neck squamous cell carcinoma. Oral oncology vol. 49,3 (2013): 211-5.

3. Charoentong, P. et al. Pan-cancer immunogenomic analyses reveal genotypeimmunophenotype relationships and predictors of response to checkpoint blockade. Cell Rep. 18, 248–262 (2017)

4. Weir, B.S. and Cockerham, C.C. Estimating F-Statistics for the Analysis of Population Structure. Evolution, 1984, 38, 1358-1370. [http://dx.doi.org/10.2307/2408641](http://dx.doi.org/10.2307/2408641)

5. Bhatia G , Patterson N , Sankararaman S , et al. Estimating and interpreting FST: The impact of rare variants[J]. Genome Research, 2013, 23(9):1514-1521.

6. Holsinger K E , Weir B S . Genetics in geographically structured populations: defining, estimating and interpreting FST[J]. Nature Reviews Genetics, 2009, 10(9):639-650.

7. Christina Curtis et al. Quantitative evidence for early metastatic seeding in colorectal cancer, Nature Genetics (2019). [http://dx.doi.org/10.1038/s41588-019-0423-x](http://dx.doi.org/10.1038/s41588-019-0423-x)

8. Ruidong Xue et al. Genomic and Transcriptomic Profiling of Combined Hepatocellular and Intrahepatic Cholangiocarcinoma Reveals Distinct Molecular Subtypes. Cancer Cell, 2019, 35(6):932-947.

9. Puente X S, Pinyol M, Quesada V, et al. Whole-genome sequencing identifies recurrent mutations in chronic lymphocytic leukaemia[J]. Nature, 2011, 475(7354):101-105.

10. Hu Z, Li Z, Ma Z & Curtis C. Pan-cancer analysis of clonality and the timing of systemic spread in paired primary tumors and metastases. bioRXiv (2019). [http://dx.doi.org/10.1101/825240](http://dx.doi.org/10.1101/825240)

11. Williams, M., Werner, B., Barnes, C. et al. Identification of neutral tumor evolution across cancer types. Nat Genet 48, 238–244 (2016). [https://doi.org/10.1038/ng.3489](https://doi.org/10.1038/ng.3489)
  
  
  
### 11. Session Info
```{r}
sessionInfo()
```
  
  