---
title: 'MesKit: Analyze and visualize Multi-region Whole-exome Sequencing Data'
output:
  html_document:
    df_print: paged
    highlight: pygments
    self_contained: yes
    theme: united
    toc: yes
    toc_depth: 4
    toc_float: yes
  pdf_document:
    toc: yes
    toc_depth: '4'
date: "`r Sys.Date()`"
---

<style type="text/css">
    .list-group-item.active,.list-group-item.active:focus,.list-group-item.active:hover {
    background-color: #007510;
    }
    body {
      font-family: Calibri, helvetica, sans-serif;
      font-size: 14px;
      line-height: 1.8;
    }
    h1 {
      font-size: 127%;
      color: #750400;
    }
    h2 {
      font-size: 121%;
      color: #750400;
    }    
    h3 {
      font-weight: bold;
      font-size: 109%;
      color: #750400;
      line-height: 2.4;
    }
    h4 {
      font-weight: bold;
      font-size: 100%;
      color: #750400;
      line-height: 2.1;
    }
    a, a:hover {
      color: #007510;
    }
</style>


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```  
  
### 1. Introduction  
  
Intra-tumor heterogeneity (ITH) is now thought to be a key factor contributing to the therapeutic failures and drug resistance, which have attracted increasing attention in the cancer research field. Here, we present an R package, MesKit, for characterizing cancer genomic ITH and inferring the tumorâ€™s evolutionary history. MesKit provides a wide range of analysis including ITH evaluation, enrichment, signature, clone evolution analysis via implementation of well-established computational and statistical methods. The source code and documents are freely available through Github (https://github.com/Niinleslie/MesKit). We also developed a shiny application to provide easier analysis and visualization.  
  

#### 1.1 Citation

In R console, enter ```citation("MesKit")```.  
  
_**MesKit: a tool kit for dissecting cancer evolution from multi-region derived tumor biopsies via somatic mutations**_
  
### 2. Prepare input Data  
  
To analyze with MesKit, you need to provide:  
  
* An MAF file of multiple tumors samples from the same person (named ```patientID.maf | patientID.maf.gz```). **Required**
* Necessary details of samples. **Required**
* CCF results generated by [PyClone](https://github.com/aroth85/pyclone) (```loci.tsv & cluster.tsv```). **Optional but recommended**

  
#### 2.1 MAF files  
  
Mutation Annotation Format (MAF) files are tab-delimited text files with aggregated mutation information from VCF Files. The following fields are required to be contained in the MAF files with MesKit.  
  
**Mandatory fields:**  
  
```Hugo_Symbol```, ```Chromosome```, ```Start_Position```, ```End_Position```, ```Variant_Classification```, ```Variant_Type```, ```Reference_Allele```, ```Tumor_Seq_Allele2```, ```VAF```, ```Tumor_Sample_Barcode```.  

**Example MAF file**  
  
```{r echo=FALSE, paged.print=FALSE, rownames.print=FALSE}
MAFtable <- read.table(system.file("extdata/maf", "HCC6046.maf", package = "MesKit"), header=TRUE)
extractLines <- rbind(MAFtable[1,],MAFtable[383,])
extractLines <- rbind(extractLines,MAFtable[739,])
data.frame(extractLines, row.names = NULL)
```  
  
  
#### 2.2 Information of samples  
  
Below is an example of tumors sampling across multiple spatially-distinct regions. It should contain the ```sampleID```, ```patientID``` and ```lesion```.  
  
```{r echo=FALSE, paged.print=FALSE, rownames.print=FALSE}
sampleInfo <- read.table(system.file("extdata", "HCC6046.sampleInfo.txt", package = "MesKit"), header = TRUE)
data.frame(sampleInfo, row.names = NULL)
maf.File <- system.file("extdata/maf", "HCC6046.maf", package = "MesKit")
```  

 
### 3. Installation  
  
  
#### Via GitHub 
Install the latest version of this package by typing the commands below in R console:  
  
```{r eval=FALSE}
install.packages("remotes")
remotes::install_github("Niinleslie/MesKit")
```  
  
  
  
### 4. Start with the Maf object  
  
```readMaf``` function create an MAF object by reading MAF files along with specific sample information. Parameter ```refBuild``` is used to set particular full genome sequences for Homo sapiens reference (```"hg19"``` or ```"hg38"```). Additionally, we recommend you provide cancer cell fraction (CCF) data generated by [PyClone](https://github.com/aroth85/pyclone) for visualizing clonal history. ```ccf.mutation.id``` manually specifies which columns could be joint by ```ccf.mutation.sep``` to match mutation id in ccf data.  
  
```{r message=FALSE}
library(MesKit)
```  
  
  
```{r}
maf.File <- system.file("extdata/maf", "HCC6046.maf", package = "MesKit")
sampleInfo.File <- system.file("extdata", "HCC6046.sampleInfo.txt", package = "MesKit")
ccf.cluster.File <- system.file("extdata/ccf", "HCC6046.cluster.tsv", package = "MesKit")
ccf.loci.File <- system.file("extdata/ccf", "HCC6046.loci.tsv", package = "MesKit")

# Maf object generation
maf <- readMaf(mafFile = maf.File, sampleInfo = sampleInfo.File, 
                refBuild = "hg19") 

# Maf object with CCF information
maf <- readMaf(mafFile = maf.File, sampleInfo = sampleInfo.File, 
                ccfClusterTsvFile = ccf.cluster.File, 
                ccfLociTsvFile = ccf.loci.File,
                refBuild = "hg19")  
```  
  
  
  
### 5. ITH evaluation  
  
MesKit offers several functions to estimate intratumoral heterogeneity (ITH) with bulk-tumor WES data, including  ```mathScore```, ```vafCluster```, ```mutSharedPrivate``` and ```JaccardSimilarity```.  
  
```mathScore``` function estimates ITH of single sample using MATH approach. Typically, the higher the MATH score, the more heterogeneous is a tumor.   
  
```{r}
mathScore(maf, minvaf = 0.02, maxvaf = 1)
```  
  
```vafCluster``` function clusters variant allele frequencies (VAFs) based on a Gaussian finite mixture model. The function produces a density distribution plot, depicting clusters of mutations with different VAFs in all/selected samples. We consider the number of clusters as a simple indicator of ITH.  
Parameters ```minVaf``` and ```maxVaf``` can be set to filter mutations. In addition, we offer several options for defining the plotting styles. Parameter `plotOption` could be set as ```"separate"```, ```"combine"``` or ```"compare"```, which respectively means to output the frequency map in different documents, one documents or the ridgeline plot.  
  
```{r message=FALSE}
vafCluster(maf, plotOption = "combine", showMATH = T)
```  
  
For each patient, ```mutSharedPrivate``` function identifies mutations shared by all samples as shared mutations (trunk mutations), mutations observed in a subset of samples as partial-shared mutations, others which are unique to single sample as private mutations (branch mutations). This function provides an overview of mutations using stack plots. Parameter ```show.num``` determines whether to show the numbers of mutations in the plot.  
  
```{r,fig.align='left',fig.width=7, fig.height=5}
mutSharedPrivate(maf, show.num = FALSE)
```  
  
```mutOncoTSG``` function calculates the proportion of shared/partial-shared/private mutations that occurs in oncogenes and tumor suppressors genes (TSGs), separatedly. The default oncogenes list and TSGs list of pan-cancer were curated from previous publications and public databases.  
  
```{r, fig.align='left',fig.width=6, fig.height=4.5}
oncogeneListFile <- system.file("extdata/", "oncogene.list.txt", package = "MesKit")
tsgListFile <- system.file("extdata/", "TSG.list.txt", package = "MesKit")
mutOncoTSG(maf, oncogeneListFile, tsgListFile, show.percentage = TRUE)
```  
  
```JaccardIndex``` function returns a graphical display of Jaccard similarity correlation matrix. Jaccard similarity coefficients are calculated based on the ratio of shared mutations for paired-samples from single patient.   
```{r fig.align='left',fig.width=5.5, fig.height=5.5, fig.asp=1.0}
JaccardIndex(maf, type = "lower")
```  
  
  
  
### 6. Clonal analysis  
  
Subclones inference generated by [PyClone](https://github.com/aroth85/pyclone) can be integrated to reconstruct and interpret subclonal architecture.  
  
```tumorClonesPlot``` function can visualize the CCF distribution of subclones and mutations (```loci.tsv``` and ```cluster.tsv``` generated by PyClone are required). Parameter ```clone.min.mut``` and ```clone.min.aveCCF``` specify the minimum mutation number of a subclone and the minimum average CCF value of a subclone, respectively.  
  
```{r fig.align='left',fig.width=6.5, fig.height=4.5}
tumorClonesPlot(maf, clone.min.mut = 5, clone.min.aveCCF = 0.1)
```  
  
```ccfDensity``` function can be used to print pairwise CCF plots of mutations which are identified across samples from single patient. Recent studies have used this method to infer whether metastases are seeded in monoclonal manner or polyclonal manner from primary tumors.  
  
```{r fig.align='left', fig.width=8, fig.height=3.5, message=FALSE}
ccf.pairwise <- ccfDensity(maf, show.density = TRUE)
#ccf.pairwise[[1]]
cowplot::plot_grid(ccf.pairwise[[1]], ccf.pairwise[[2]])
```  

### 7. Phylogenetic tree object generation  
  
Based on the Maf object, ```NJtree``` function reconstructs phylogenetic tree via the neighbor-joining method and stored in an njtree object, which is later utilized for functional analysis, mutational signature analysis and printing phylogenetic trees.  
  
```{r}
njtree <- getNJtree(maf)
```  
  
  
  
### 8. Functional exploration  
  
MesKit conducts enrichment analysis of Gene ontology (GO) and pathway using [ClusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) R package. ```GO.njtree``` function and ```Pathway.njtree``` function  return enrichment results on both tree-level and branch-level. By default, ```Pathway.njtree``` function performs KEGG pathway enrichment analysis, besides, you can choose ```"reactome pathway"``` via parameter ```pathway.type```.  
  
  
#### 8.1 GO enrichment analysis
```{r, message=FALSE}
# Required org.Hs.eg.db
library("org.Hs.eg.db")

# GO enrichment analysis
GO.result <- GO.njtree(njtree, GO.type = "BP", pval = 0.05, 
                pAdjustMethod = "BH", qval = 0.2, plotType = "dot", showCategory = 5)
names(GO.result$GO.category)
names(GO.result$GO.plot)
```
  
```{r, message=FALSE}
head(GO.result$GO.category[["SRR3670033"]])
```
  
```{r, fig.align='left', fig.height=4.5, fig.width=8.5, message=FALSE}
GO.result$GO.plot[["SRR3670036"]]
```
  
  
#### 8.2 Pathway enrichment analysis
```{r, message=FALSE, warning=FALSE}
# Pathway enrichment analysis
pathway.result <- Pathway.njtree(njtree, pathway.type = "KEGG", pval = 0.05, 
                pAdjustMethod = "BH", qval = 1, plotType = "dot", showCategory = 5)
names(pathway.result$pathway.category)
names(pathway.result$pathway.plot)
```  
  
  
```{r message=FALSE}
head(pathway.result$pathway.category[["All"]])
```
  
  
#### 8.3 Mutational signature analysis
  
```treeMutationalSig``` function identifies potential signatures which could be attributed to known mutational processes for trunks/branches of phylogenetic trees. 
This function is based on R package [deconstructSigs](https://github.com/raerose01/deconstructSigs) and you can set signatures matrix via ```signaturesRef``` (```cosmic``` or ```nature2013```). Parameter ```mutThreshold``` specifies the minimum number of mutations (Default: 50) required to perform mutation signature analysis. Moreover, you could provide a driver genes list via ```driverGenesFile``` and the final result would contain an additional column recording driver genes for particular mutation intersections.  
  
  
```{r, warning = FALSE, message=FALSE}
# Requires BSgenome.Hsapiens.UCSC.hg19 if you set refBuild = "hg19"
library("BSgenome.Hsapiens.UCSC.hg19")
# Requires BSgenome.Hsapiens.UCSC.hg38 if you set refBuild = "hg38"
library("BSgenome.Hsapiens.UCSC.hg38")
```

```{r, warning = FALSE, message=FALSE}
driverGenesFile <- system.file("extdata", "putative_driver_genes.txt", package = "MesKit")
treeMSOutput <- treeMutationalSig(njtree, driverGenesFile = driverGenesFile, mutThreshold = 50)
```

With the output from ```treeMutationalSig```, function ```mutSigSummary``` summarise signature information as a data frame and ```plotMutationalSig``` visualises signature information above with function ```plotMutationalSig```.

```{r, warning = FALSE, message=FALSE}
mutSig <- mutSigSummary(treeMSOutput)
head(mutSig)
plotMutationalSig(treeMSOutput)
```

Function ```mutBranchTrunk``` conludes the relationship and categories of muttations. When plotting the boxplot by ```plotBranchTrunk```, mutBranchTrunk shows the distribution of branch or trunk mutations, and the confidence level for both functions above can be tuned by parameter ```conf.level```.

```{r, warning = FALSE, message=FALSE}
mutBT <- mutBranchTrunk(treeMSOutput, conf.level = 0.95)
head(mutBT)
plotBranchTrunk(treeMSOutput, conf.level = 0.95)

```
  
### 9. Phylogenetic tree visualization  
  
```plotPhyloTree``` function is aimed at visualizing phylogenetic trees. If ```show.mutSig = TRUE``` (Default), mutational signatures would be shown in different colors for each trunk and branch. Either binary mutation heatmap (indicates the presence or absence of a mutation) or ccf heatmap can be displayed by setting parameter `heatmap.type`.
  
  
#### 9.1 A phylogenetic tree along with binary heatmap of mutations  
  
  
```{r, fig.align='left',fig.width=9, fig.height=5, message=FALSE, warning=FALSE, error=TRUE}
plotPhyloTree(njtree, show.mutSig = TRUE, heatmap.type = 'binary') 
```
  
  
#### 9.2 A phylogenetic tree along with CCF heatmap of mutations  
  
```{r, fig.align='left',fig.width=9, fig.height=5, message=FALSE, warning=FALSE, error=TRUE}
plotPhyloTree(njtree, show.mutSig = TRUE, heatmap.type = 'CCF') 
```
  
  
### 10. Session Info
```{r}
sessionInfo()
```
  
  
