#' calFst
#' @description genetic divergence between regions of subclonal sSNVs using the Weir and Cockerham method 
#' @references Sun R, Hu Z, Sottoriva A, et al. Between-region genetic divergence reflects the mode and tempo of tumor evolution. 
#' Nat Genet. 2017;49(7):1015–1024. doi:10.1038/ng.3891
#' 
#' Bhatia G, Patterson N, Sankararaman S, Price AL. Estimating and interpreting FST: the impact of rare variants. 
#' Genome Res. 2013;23(9):1514–1521. doi:10.1101/gr.154831.113
#' 
#' @param maf an Maf object generated by readMaf function.
#' @param patient.id select the specific patients. Default: NULL, all patients are included
#' @param use.adjVAF whether to use adjusted VAF values rather than original VAF values, default FALSE
#' @param min.vaf specify the minimum CCF value of subclonal mutations, default is 0.08
#' 
#' @return a list contains Fst value of MRS and Hudson estimator of each sample-pair, respectively.
#'
#' @examples
#' calFst(maf)
#' @import dplyr
#' @export calFst




fst.hudson.pair <- function(maf.pair) {
	
	vafCol <- which(grepl("vaf", colnames(maf.pair)))

	#maf.filter <- maf.pair %>%
		#dplyr::filter_at(
			#vars(dplyr::starts_with("vaf")),  all_vars(. > min.vaf)
			#)

	## calculate covarian and standard deviations
	maf.pair <- mutate(
		maf.pair, 
		covariance = (vaf1-vaf2)^2-(vaf1*(1-vaf1))/(depth1-1)-(vaf2*(1-vaf2))/(depth2-1), 
		sd = vaf1*(1-vaf2)+vaf2*(1-vaf1)
		)

	Fst.h = sum(maf.pair$covariance)/sum(maf.pair$sd)
  	return(Fst.h)
}

fst.hudson.patient <- function(df) {
	## pairwise heterogeneity
	samples <- as.character(unique(df$Tumor_Sample_Barcode))
	pairs <- combn(length(samples), 2, simplify = FALSE)
	fstHudson <-c()
	pairIDs <- c()

  	for (pair in pairs){
  		maf.pair <- subset(df, Tumor_Sample_Barcode %in% c(samples[pair[1]],samples[pair[2]])) %>%
  			tidyr::pivot_wider(
  				names_from = Tumor_Sample_Barcode, 
  				values_from = c(VAF, totalDepth),
  				values_fill = c(VAF = 0, totalDepth = 0)
  				) %>%
  			dplyr::select(-Patient_ID)

  		colnames(maf.pair) <- c("mutation_id", "vaf1", "vaf2", "depth1", "depth2")
  		pairIDs <- c(pairIDs, 
  			paste0("\"", samples[pair[1]], "\"", "-", "\"", samples[pair[2]], "\"")
  			)
  		fstHudson <- c(fstHudson, fst.hudson.pair(maf.pair))
  	}
  	Fst <- mean(fstHudson)
  	return(list(Fst.mean = Fst, Fst.pair = data.frame(sample.pair=pairIDs, fst.hudson=fstHudson)))

  }


calFst <- function(maf, patient.id = NULL, use.adjVAF = FALSE, min.vaf=0.08){
	mafData <- maf@data

    if(is.null(patient.id)){
        patient.id = unique(mafData$Patient_ID)
    }else{
        patient.setdiff <- setdiff(patient.id, unique(mafData$Patient_ID))
        if(length(patient.setdiff) > 0){
            stop(paste0("Patient ", patient.setdiff, " can not be found in your data"))
        }
    }

	if(use.adjVAF){
		if(! "CCF" %in% colnames(mafData)){
			stop("Adjust VAF was not found in maf object. Check if CCF data was provided")
		}
		vaf.col = "VAF_adj" 
	}else{
		vaf.col = "VAF"
	}

	Fst.out <- mafData %>%
		dplyr::filter(Patient_ID %in% patient.id) %>%
    	tidyr::unite(
        	"mutation_id", 
        	c(
            	"Chromosome", 
            	"Start_Position", 
            	"Reference_Allele", 
            	"Tumor_Seq_Allele2"), 
        	sep = ":", 
        	remove = FALSE
    	) %>% 
    	#dplyr::filter(Variant_Type == "SNP") %>%
    	dplyr::mutate(
        	totalDepth = Ref_allele_depth + Alt_allele_depth) %>%
    	dplyr::select(        	
        	mutation_id,
        	Tumor_Sample_Barcode,
        	Patient_ID,
        	!!vaf.col,
        	totalDepth) %>% 
    	dplyr::rename("VAF" = !!vaf.col) %>%
    	dplyr::filter(VAF > min.vaf) %>%
    	dplyr::group_by(Patient_ID) %>%
        dplyr::group_map(~fst.hudson.patient(.), keep = TRUE) %>%
        rlang::set_names(patient.id)

  	return(Fst.out)
}



