#' calFst
#' @description genetic divergence between regions of subclonal sSNVs using the Weir and Cockerham method 
#' @references Sun R, Hu Z, Sottoriva A, et al. Between-region genetic divergence reflects the mode and tempo of tumor evolution. 
#' Nat Genet. 2017;49(7):1015–1024. doi:10.1038/ng.3891
#' 
#' Bhatia G, Patterson N, Sankararaman S, Price AL. Estimating and interpreting FST: the impact of rare variants. 
#' Genome Res. 2013;23(9):1514–1521. doi:10.1101/gr.154831.113
#' 
#' @param maf an Maf object generated by readMaf function.
#' @param min.clonal.CCF a cut-off of CCF that distinguishes clonal and subclonal SSNVs. Clonal mutations are excluded in FST calculations.
#' @param min.VAF specify the minimum CCF value of subclonal mutations.
#' @return a list contains Fst value of MRS and Hudson estimator of each sample-pair, respectively.
#'
#' @examples
#' calFst(maf)
#' @export calFst



fst.hudson <- function(maf.pair, minVAF) {
	vafCol <- which(grepl("vaf", colnames(maf.pair)))
	maf.filter <- dplyr::filter(maf.pair, apply(maf.pair[,vafCol],1,max) > minVAF) 

	## calculate covarian and standard deviations
	maf.filter <- mutate(maf.filter, covariance = (vaf1-vaf2)^2-(vaf1*(1-vaf1))/(depth1-1)-(vaf2*(1-vaf2))/(depth2-1), 
		sd = vaf1*(1-vaf2)+vaf2*(1-vaf1)) 
	Fst.h = mean(maf.filter$covariance)/mean(maf.filter$sd)
  	return(Fst.h)
}



calFst <- function(maf, min.clonal.CCF = 0.6, minVAF=0.08){
	maf.dat <- maf@data
	if(! "CCF" %in% colnames(maf.data){
		stop("Calculation of Fst requires CCF data. No CCF data was found when readMaf")
	}
	else {
		mergedCCF <- getMergedCCF(maf.dat)
		clonalSNV <- as.vector(mergedCCF[mergedCCF$mergedCCF >= min.clonal.CCF, "mutation_id"])
		
		## filter clonal SNVs which does not contribute to ITH
		subMaf <- maf.dat %>%
			tidyr::unite("mutation_id", c("Chromosome", "Start_Position", "Reference_Allele", "Tumor_Seq_Allele2"), sep = ":", remove = FALSE) %>% 
			dplyr::filter(Variant_Type == "SNP" & !mutation_id %in% clonalSNV) %>%
			dplyr::mutate(totalDepth = Ref_allele_depth + Alt_allele_depth)

		## pairwise heterogeneity
		samples <- as.character(unique(subMaf$Tumor_Sample_Barcode))
		pairs <- combn(length(samples), 2, simplify = FALSE)

		fstHudson <-c()
		pairIDs <- c()
  		for (pair in pairs){
  			maf.pair <- subset(subMaf, Tumor_Sample_Barcode %in% c(samples[pair[1]],samples[pair[2]])) %>%
  			dplyr::select(Tumor_Sample_Barcode, mutation_id, VAF, totalDepth ) %>%
  			tidyr::pivot_wider(names_from = Tumor_Sample_Barcode, values_from = c(VAF, totalDepth))
  			colnames(maf.pair) <- c("mutation_id", "vaf1", "vaf2", "depth1", "depth2")

  			pairIDs <- c(pairIDs, paste(samples[pair[1]], samples[pair[2]], sep = "_"))
  			fstHudson <- c(fstHudson, fst.hudson(maf.pair, minVAF))
  		}
  		Fst <- mean(fstHudson)
  		Fst.out <- list(Fst, data.frame(paired.samples=pairIDs, fst.hudson = fstHudson))
  		names(Fst.out) <- c("Fst", "fst.hudson")
  		return(Fst.out)
  	}
}



