#' calFst
#' @description genetic divergence between regions of subclonal sSNVs using the Weir and Cockerham method 
#' @references Sun R, Hu Z, Sottoriva A, et al. Between-region genetic divergence reflects the mode and tempo of tumor evolution. 
#' Nat Genet. 2017;49(7):1015–1024. doi:10.1038/ng.3891
#' 
#' Bhatia G, Patterson N, Sankararaman S, Price AL. Estimating and interpreting FST: the impact of rare variants. 
#' Genome Res. 2013;23(9):1514–1521. doi:10.1101/gr.154831.113
#' 
#' @param maf an Maf object generated by readMaf function.
#' @param patient.id select the specific patients. Default: NULL, all patients are included
#' @param min.vaf specify the minimum VAF, default is 0.08
#' @param withinType calculate fst within types in each patients,default is FALSE.
#'  
#' @return a list contains Fst value of MRS and Hudson estimator of each sample-pair, respectively.
#'
#' @examples
#' calFst(maf)
#' @import dplyr
#' @export calFst




fst.hudson.pair <- function(maf.pair) {
	
	vafCol <- which(grepl("vaf", colnames(maf.pair)))

	maf.pair <- mutate(
		maf.pair, 
		covariance = (vaf1-vaf2)^2-(vaf1*(1-vaf1))/(depth1-1)-(vaf2*(1-vaf2))/(depth2-1), 
		sd = vaf1*(1-vaf2)+vaf2*(1-vaf1)
		)

	Fst.h = mean(maf.pair$covariance)/mean(maf.pair$sd)
  return(Fst.h)
}

fst.hudson.patient <- function(df, min.vaf, plot = TRUE, use.circle = TRUE, title = NULL, withinType = FALSE) {
  
  df <- as.data.frame(df)
  patient <- unique(df$Patient_ID)
  ## filter by min.vaf
    if(withinType){

        type <- unique(df$Tumor_Type)
        
        if(length(unique(df$Tumor_Sample_Barcode))  < 2 ){
            message(paste0("Warnings: Only one sample was found of ", type,
                           " in ", patient, ". It you want to compare CCF between regions, withinType should be set as FALSE"))
            return(NA)
        }
        
        df <- df %>%
            dplyr::select(-Tumor_Type)
        # ## filter by min.vaf
        # idx <- (df %>% 
        #             dplyr::group_by(mutation_id) %>% 
        #             dplyr::summarise(max = max(VAF)) %>% 
        #             dplyr::filter(max > min.vaf) %>% 
        #             as.data.frame())$mutation_id  
        # 
        # df <- df[df$mutation_id %in% idx,] %>% 
        #     dplyr::select(-Tumor_Type)
        # 
        # if(length(unique(df$Tumor_Sample_Barcode))  < 2 ){
        #     message(paste0("Warnings: Only one sample was found of ", type,
        #                    " in ", patient, " after filter by min.vaf."))
        #     return(NA)
        # }

    }
  else{
      if(length(unique(df$Tumor_Sample_Barcode))  < 2 ){
          message(paste0("Warnings: Only one sample was found in ", patient, "."))
          return(NA)
      } 
  }

  
	## pairwise heterogeneity
  patientID <- as.character(unique(df$Patient_ID))
	samples <- as.character(unique(df$Tumor_Sample_Barcode))
	pairs <- combn(length(samples), 2, simplify = FALSE)

  Fst.dist <- diag(1, nrow = length(samples), ncol = length(samples))
  rownames(Fst.dist) <- samples
  colnames(Fst.dist) <- samples
  
  Fst.df <- list()
  
  for (pair in pairs){
    maf.pair <- subset(df, Tumor_Sample_Barcode %in% c(samples[pair[1]],samples[pair[2]])) %>%
      tidyr::pivot_wider(
        names_from = Tumor_Sample_Barcode,       
        values_from = c(VAF, totalDepth),
        values_fill = c(VAF = 0, totalDepth = 0)
        ) %>%
      dplyr::ungroup() %>%
      dplyr::select(-Patient_ID)
    colnames(maf.pair) <- c("mutation_id", "vaf1", "vaf2", "depth1", "depth2")
    
    name <- paste(samples[pair[1]],samples[pair[2]],sep = "_")

    Fst.dist[pair[1],pair[2]] <- Fst.dist[pair[2],pair[1]] <- fst.hudson.pair(maf.pair)
    
    sub <- data.frame(patientID, name, fst.hudson.pair(maf.pair))
    Fst.df <- rbind(Fst.df,sub)
    
  }
  colnames(Fst.df) <- c("Patient_ID","Pair", "Fst")
  Fst.avg <- mean(Fst.dist)
  if(is.null(title)){
      if(withinType){
          title <- paste0("Fst of ",type," in ", patientID, ": ",round(Fst.avg,2))
      }
      else{
          title <- paste0("Fst of patient ", patientID, ": ",round(Fst.avg,2))
      }
  }
  
  return(list(
      Fst.df = Fst.df,
      Fst.avg = Fst.avg, 
      Fst.pair = Fst.dist,
      Fst.plot = if(plot) plotCorr(
        Fst.dist, 
        use.circle, 
        title = if(!is.null(title)) title else{paste0("Fst of patient ", patientID, ": ",round(Fst.avg,2))}) else{NA} 
    )
  )
 }


groupByType <- function(df, min.vaf, plot = TRUE, use.circle = TRUE, title = NULL, withinType = FALSE){
    
    types <- unique(df$Tumor_Type)
    
    Fst.type.out <- df %>% 
        group_by(Tumor_Type) %>% 
        dplyr::group_map(
            ~fst.hudson.patient(.,
                                min.vaf = min.vaf,
                                plot = plot,
                                use.circle = use.circle,
                                title = title,
                                withinType = withinType),keep = TRUE) %>% 
        rlang::set_names(types)
    
    Fst.type.out <- Fst.type.out[!is.na(Fst.type.out)]
    
    return(Fst.type.out)
    
}

calFst <- function(
  maf, 
  patient.id = NULL, 
  #use.adjVAF = FALSE, 
  min.vaf = 0.08,
  #max.vaf = 1,
  plot = TRUE,
  use.circle = TRUE,
  title = NULL,
  withinType = FALSE){

	mafData <- maf@data

    if(is.null(patient.id)){
        patient.id = unique(mafData$Patient_ID)
    }else{
        patient.setdiff <- setdiff(patient.id, unique(mafData$Patient_ID))
        if(length(patient.setdiff) > 0){
            stop(paste0("Patient ", patient.setdiff, " can not be found in your data"))
        }
        mafData <- mafData[Patient_ID %in% patient.id,]
    }
	

	#if(use.adjVAF){
		#if(! "CCF" %in% colnames(mafData)){
			#stop("Adjust VAF was not found in maf object. Check if CCF data was provided")
		#}
		#vaf.col = "VAF_adj" 
	#}else{
		#vaf.col = "VAF"
	#}
	mafData <-  mafData %>%
	    dplyr::filter(VAF > min.vaf,
	                  Patient_ID %in% patient.id) %>% 
	    tidyr::unite(
	        "mutation_id", 
	        c(
	            "Chromosome", 
	            "Start_Position", 
	            "Reference_Allele", 
	            "Tumor_Seq_Allele2"), 
	        sep = ":", 
	        remove = FALSE
	    ) %>% 
	    #dplyr::filter(Variant_Type == "SNP") %>%
	    dplyr::mutate(
	        totalDepth = Ref_allele_depth + Alt_allele_depth)
	
    if(withinType){
        filter.out <- mafData %>%
            dplyr::filter(Clonal_Status == "Subclonal") %>%
            dplyr::select(        	
                mutation_id,
                Tumor_Type,
                Tumor_Sample_Barcode,
                Patient_ID,
                VAF,
                #!!vaf.col,
                totalDepth)
        
        ## fresh patient.id
        patient.id <- unique(filter.out$Patient_ID)
        
        Fst.out <- filter.out %>% 
            dplyr::group_by(Patient_ID) %>%
            dplyr::group_map(
                ~groupByType(.,
                             min.vaf = min.vaf,
                             plot = plot,
                             use.circle = use.circle,
                             title = title,
                             withinType = withinType), 
                keep = TRUE) %>%
            rlang::set_names(patient.id)
    }
	else{
	    filter.out <- mafData %>%
	        dplyr::filter(Clonal_Status == "Subclonal") %>%
	        dplyr::select(        	
	            mutation_id,
	            Tumor_Sample_Barcode,
	            Patient_ID,
	            VAF,
	            #!!vaf.col,
	            totalDepth)
	    ## fresh patient.id
	    patient.id <- unique(filter.out$Patient_ID)
	    
	    Fst.out <- filter.out %>% 
	        dplyr::group_by(Patient_ID) %>%
	        dplyr::group_map(~fst.hudson.patient(.,
	                         min.vaf = min.vaf,
	                         plot = plot,
	                         use.circle = use.circle,
	                         title = title,
	                         withinType = withinType), 
	            keep = TRUE) %>%
	        rlang::set_names(patient.id)
	    
	    Fst.out <- Fst.out[!is.na(Fst.out)]
	}
        
  df.fst <- list()
  if(withinType){
      patient.list <- c()
      type.list <- c()
      avg.list <- c()
      name.list <- c()
      Fst.pair <- list()
      Fst.plot <- list()
      Fst.df <- list()
      for(patient in names(Fst.out)){
          
          ## get result for each patient
          Fst.out.p <- Fst.out[[patient]]
          
          for(type in names(Fst.out.p)){
              ## get result for each patient in each type
              Fst.out.p.t <- Fst.out.p[[type]]
               
              name <- paste0(patient,"_",type) 
              patient.list <- c(patient.list,patient)
              type.list <- c(type.list,type)
              avg.list <- c(avg.list,Fst.out.p.t$Fst.avg)
              
              Fst.pair[[name]] <- Fst.out.p.t$Fst.pair
              Fst.plot[[name]] <- Fst.out.p.t$Fst.plot
              df <- Fst.out.p.t$Fst.df
              df$type <- name
              Fst.df[[name]] <- df
          }
      }
      Fst.df <- plyr::rbind.fill(Fst.df)
      colnames(Fst.df) <- c("Patient_ID","Pair", "Fst", "Tumor_Type")
      
      Fst.df<- dplyr::select(Fst.df,Patient_ID,Tumor_Type,Pair,Fst)
      
      Fst.avg = data.frame(Patient_ID = patient.list,
                           Tumor_Type = type.list,
                           Fst.avg = avg.list)
      
      if(plot){
          Fst.out=list(
              Fst.avg = Fst.avg,
              Fst.pair = Fst.df,
              Fst.plot = Fst.plot
          )
      }
      else{
          Fst.out=list(
              Fst.avg = Fst.avg,
              Fst.pair = Fst.df
          )  
      }
      
      
  }
  else{
      Fst.df <- plyr::rbind.fill(lapply(Fst.out, function(x) x$Fst.df))
      if(plot){
          Fst.out=list(
              Fst.avg = data.frame(Patient_ID = names(Fst.out), Fst.avg = unlist(lapply(Fst.out, function(x) x$Fst.avg))),
              Fst.pair = Fst.df,
              Fst.plot = lapply(Fst.out, function(x) x$Fst.plot)
          ) 
      }
      else{
          Fst.out=list(
              Fst.avg = data.frame(Patient_ID = names(Fst.out), Fst.avg = unlist(lapply(Fst.out, function(x) x$Fst.avg))),
              Fst.pair = Fst.df
          )
      }
  }

  
  return(Fst.out)
}
