#' calFst
#' @description genetic divergence between regions of subclonal sSNVs using the Weir and Cockerham method 
#' @references Sun R, Hu Z, Sottoriva A, et al. Between-region genetic divergence reflects the mode and tempo of tumor evolution. Nat Genet. 2017;49(7):1015–1024. doi:10.1038/ng.3891
#' 
#' Bhatia G, Patterson N, Sankararaman S, Price AL. Estimating and interpreting FST: the impact of rare variants. 
#' Genome Res. 2013;23(9):1514–1521. doi:10.1101/gr.154831.113
#' 
#' @param maf an Maf object generated by readMaf function.
#' @param patient.id select the specific patients. Default: NULL, all patients are included.
#' @param chrSilent Select chromosomes needed to be dismissed. Default NULL.
#' @param mutType select Proper variant classification you need. Default "All". Option: "nonSilent".
#' @param use.indel Logical value. Whether to use INDELs besides somatic SNVs. Default: TRUE.
#' @param min.vaf specify the minimum VAF_adj, default is 0.02
#' @param min.total.depth The minimum total allele depth for filtering variants. Default: 2.
#' @param withinTumor calculate fst within types in each patients,default is FALSE.
#'  
#' @return a list contains Fst value of MRS and Hudson estimator of each sample-pair, respectively.
#'
#' @examples
#' calFst(maf)
#' @import dplyr
#' @export calFst

calFst <- function(
   maf, 
   patient.id = NULL, 
   chrSilent = NULL,
   mutType = "All",
   use.indel = TRUE,
   min.vaf = 0.02,
   min.total.depth = 2,
   plot = TRUE,
   use.circle = TRUE,
   title = NULL,
   withinTumor = FALSE,
   number.cex = 8, 
   number.col = "#C77960" ){
   
    
    if(min.total.depth <= 1){
        stop("min.total.depth should be greater than 1")
    }
    if(withinTumor){
        clonalStatus <- "Subclonal"
    }else{
        clonalStatus <- NULL
    }
    maf <- subsetMaf(maf,
                    patient.id = patient.id,
                    chrSilent = chrSilent,
                    mutType = mutType,
                    use.indel = use.indel,
                    min.vaf = min.vaf,
                    min.total.depth = min.total.depth,
                    clonalStatus = clonalStatus)
   mafData <- maf@data
   
   if(!"VAF_adj" %in% colnames(mafData)){
      stop("Adjust VAF was not found in maf object.
		     Check if CCF data was provided or let adjusted.VAF be TRUE in function readMaf when VAF have been adjusted")
   }

   Fst.input <- mafData %>% 
       dplyr::mutate(totalDepth = Ref_allele_depth + Alt_allele_depth) %>% 
      dplyr::select(        	
         Mut_ID,
         Tumor_ID,
         Tumor_Sample_Barcode,
         Patient_ID,
         VAF_adj,
         totalDepth)
   
   ## fresh patient.id
   patient.id <- unique(Fst.input$Patient_ID)
   
   Fst.avg <- data.frame()
   Fst.pair <- list()
   if(plot){
       Fst.plot <- list()
   }
   for(patient in patient.id){
      patient.data <- subset(Fst.input, Patient_ID == patient)
      
      if(!withinTumor){
         patient.data$Tumor_ID <- "All"
      }
      ids <- unique(patient.data$Tumor_ID)
      for(id in ids){
         subdata <- subset(patient.data, Tumor_ID == id)
         if(withinTumor){
            if(length(unique(subdata$Tumor_Sample_Barcode))  < 2 ){
               message(paste0("Warnings: Only one sample was found of ", id,
                              " in ", patient, ". It you want to compare CCF between regions, withinTumor should be set as FALSE"))
               next
            }
         }
         else{
            if(length(unique(subdata$Tumor_Sample_Barcode))  < 2 ){
               message(paste0("Warnings: Only one sample was found in ", patient, "."))
               next
            } 
         }
         
         ## pairwise heterogeneity
         samples <- as.character(unique(subdata$Tumor_Sample_Barcode))
         pairs <- combn(length(samples), 2, simplify = FALSE)
         
         dist_mat <- diag(1, nrow = length(samples), ncol = length(samples))
         rownames(dist_mat) <- samples
         colnames(dist_mat) <- samples
         
         for (pair in pairs){
            maf.pair <- subset(subdata, Tumor_Sample_Barcode %in% c(samples[pair[1]],samples[pair[2]])) %>%
               dplyr::select(-Tumor_ID) %>% 
               tidyr::pivot_wider(
                  names_from = Tumor_Sample_Barcode,       
                  values_from = c(VAF_adj, totalDepth),
                  values_fill = c(VAF_adj = 0, totalDepth = 0)
               ) %>%
               dplyr::select(-Patient_ID)
            colnames(maf.pair) <- c("Mut_ID", "vaf1", "vaf2", "depth1", "depth2")
            
            name <- paste(samples[pair[1]],samples[pair[2]],sep = "_")
            vafCol <- which(grepl("vaf", colnames(maf.pair)))
            maf.pair <- dplyr::mutate(
               maf.pair, 
               covariance = (vaf1-vaf2)^2-(vaf1*(1-vaf1))/(depth1-1)-(vaf2*(1-vaf2))/(depth2-1), 
               sd = vaf1*(1-vaf2)+vaf2*(1-vaf1)
            )
            fst <- mean(maf.pair$covariance)/mean(maf.pair$sd)
            dist_mat[pair[1],pair[2]] <- dist_mat[pair[2],pair[1]] <- fst
            # if(fst == -Inf){
            #     maf.pair <- maf.pair %>% 
            #         dplyr::filter(depth1 == 1)
            #     print(maf.pair)
            # }
         }
         
         ## average fst
         avg <- mean(dist_mat)
         if(withinTumor){
            df_avg <- data.frame(Patient_ID = patient, Tumor_ID = id, Fst.avg = avg)
            Fst.avg <- rbind(Fst.avg, df_avg)
         }else{
            df_avg <- data.frame(Patient_ID = patient, Fst.avg = avg)
            Fst.avg <- rbind(Fst.avg, df_avg)
         }
         
         if(withinTumor){
            name <- paste(patient,id,sep = "_")
            Fst.pair[[name]] <- dist_mat
         }else{
            Fst.pair[[patient]] <- dist_mat
         }
         
         if(is.null(title)){
            if(withinTumor){
               title <- paste0("Fst of ",id," in ", patient, ": ",round(avg,2))
            }else{
               title <- paste0("Fst of patient ", patient, ": ",round(avg,2))
            }
         }
         
         if(plot){
            p <- plotCorr(
               dist_mat, 
               use.circle, 
               number.cex = number.cex,
               number.col = number.col,
               title = if(!is.null(title)) title else{NA} 
            )
            if(withinTumor){
               name <- paste(patient,id,sep = "_")
               Fst.plot[[name]] <- p
            }else{
               Fst.plot[[patient]] <- p
            }
         }
         
      }
   }
   
   if(plot){
      return(list(Fst.avg = Fst.avg,Fst.pair = Fst.pair,Fst.plot = Fst.plot))
   }else{
      return(list(Fst.avg = Fst.avg,Fst.pair = Fst.pair))
   }
   
   return(Fst.out)
}