<<<<<<< HEAD
#' calFst
#' @description genetic divergence between regions of subclonal sSNVs using the Weir and Cockerham method 
#' @references Sun R, Hu Z, Sottoriva A, et al. Between-region genetic divergence reflects the mode and tempo of tumor evolution. 
#' Nat Genet. 2017;49(7):1015–1024. doi:10.1038/ng.3891
#' 
#' Bhatia G, Patterson N, Sankararaman S, Price AL. Estimating and interpreting FST: the impact of rare variants. 
#' Genome Res. 2013;23(9):1514–1521. doi:10.1101/gr.154831.113
#' 
#' @param maf an Maf object generated by readMaf function.
#' @param patient.id select the specific patients. Default: NULL, all patients are included
#' @param min.vaf specify the minimum VAF, default is 0.08
#' @param withinType calculate fst within types in each patients,default is FALSE.
#'  
#' @return a list contains Fst value of MRS and Hudson estimator of each sample-pair, respectively.
#'
#' @examples
#' calFst(maf)
#' @import dplyr
#' @export calFst




fst.hudson.pair <- function(maf.pair) {
	
	vafCol <- which(grepl("vaf", colnames(maf.pair)))
    
	maf.pair <- mutate(
		maf.pair, 
		covariance = (vaf1-vaf2)^2-(vaf1*(1-vaf1))/(depth1-1)-(vaf2*(1-vaf2))/(depth2-1), 
		sd = vaf1*(1-vaf2)+vaf2*(1-vaf1)
		)

	Fst.h = mean(maf.pair$covariance)/mean(maf.pair$sd)
  return(Fst.h)
}

fst.hudson.patient <- function(df, min.vaf, plot = TRUE, use.circle = TRUE, title = NULL, withinType = FALSE) {
  
  df <- as.data.frame(df)
  patient <- unique(df$Patient_ID)
  ## filter by min.vaf
    if(withinType){

        type <- unique(df$Tumor_Type)
        
        if(length(unique(df$Tumor_Sample_Barcode))  < 2 ){
            message(paste0("Warnings: Only one sample was found of ", type,
                           " in ", patient, ". It you want to compare CCF between regions, withinType should be set as FALSE"))
            return(NA)
        }
        
        df <- df %>%
            dplyr::select(-Tumor_Type)
        # ## filter by min.vaf
        # idx <- (df %>% 
        #             dplyr::group_by(mutation_id) %>% 
        #             dplyr::summarise(max = max(VAF)) %>% 
        #             dplyr::filter(max > min.vaf) %>% 
        #             as.data.frame())$mutation_id  
        # 
        # df <- df[df$mutation_id %in% idx,] %>% 
        #     dplyr::select(-Tumor_Type)
        # 
        # if(length(unique(df$Tumor_Sample_Barcode))  < 2 ){
        #     message(paste0("Warnings: Only one sample was found of ", type,
        #                    " in ", patient, " after filter by min.vaf."))
        #     return(NA)
        # }

    }
  else{
      if(length(unique(df$Tumor_Sample_Barcode))  < 2 ){
          message(paste0("Warnings: Only one sample was found in ", patient, "."))
          return(NA)
      } 
  }

  
	## pairwise heterogeneity
  patientID <- as.character(unique(df$Patient_ID))
	samples <- as.character(unique(df$Tumor_Sample_Barcode))
	pairs <- combn(length(samples), 2, simplify = FALSE)

  Fst.dist <- diag(1, nrow = length(samples), ncol = length(samples))
  rownames(Fst.dist) <- samples
  colnames(Fst.dist) <- samples

  for (pair in pairs){
    maf.pair <- subset(df, Tumor_Sample_Barcode %in% c(samples[pair[1]],samples[pair[2]])) %>%
      tidyr::pivot_wider(
        names_from = Tumor_Sample_Barcode,       
        values_from = c(VAF, totalDepth),
        values_fill = c(VAF = 0, totalDepth = 0)
        ) %>%
      dplyr::ungroup() %>%
      dplyr::select(-Patient_ID)
    colnames(maf.pair) <- c("mutation_id", "vaf1", "vaf2", "depth1", "depth2")

    #pairIDs <- c(pairIDs, 
    #  paste0("\"", samples[pair[1]], "\"", "-", "\"", samples[pair[2]], "\"")
    #  )
    #fstHudson <- c(fstHudson, fst.hudson.pair(maf.pair))
    Fst.dist[pair[1],pair[2]] <- Fst.dist[pair[2],pair[1]] <- fst.hudson.pair(maf.pair)
  }
  Fst.avg <- mean(Fst.dist)
  if(is.null(title)){
      if(withinType){
          title <- paste0("Fst of ",type," in ", patientID, ": ",round(Fst.avg,2))
      }
      else{
          title <- paste0("Fst of patient ", patientID, ": ",round(Fst.avg,2))
      }
  }
  
  return(list(
      Fst.avg = Fst.avg, 
      Fst.pair = Fst.dist,
      Fst.plot = if(plot) plotCorr(
        Fst.dist, 
        use.circle, 
        title = if(!is.null(title)) title else{paste0("Fst of patient ", patientID, ": ",round(Fst.avg,2))}) else{NA} 
    )
  )
 }


groupByType <- function(df, min.vaf, plot = TRUE, use.circle = TRUE, title = NULL, withinType = FALSE){
    
    types <- unique(df$Tumor_Type)
    
    Fst.type.out <- df %>% 
        group_by(Tumor_Type) %>% 
        dplyr::group_map(
            ~fst.hudson.patient(.,
                                min.vaf = min.vaf,
                                plot = plot,
                                use.circle = use.circle,
                                title = title,
                                withinType = withinType),keep = TRUE) %>% 
        rlang::set_names(types)
    
    Fst.type.out <- Fst.type.out[!is.na(Fst.type.out)]
    
    return(Fst.type.out)
    
}

calFst <- function(
  maf, 
  patient.id = NULL, 
  #use.adjVAF = FALSE, 
  min.vaf = 0.08,
  #max.vaf = 1,
  plot = TRUE,
  use.circle = TRUE,
  title = NULL,
  withinType = FALSE){

	mafData <- maf@data

    if(is.null(patient.id)){
        patient.id = unique(mafData$Patient_ID)
    }else{
        patient.setdiff <- setdiff(patient.id, unique(mafData$Patient_ID))
        if(length(patient.setdiff) > 0){
            stop(paste0("Patient ", patient.setdiff, " can not be found in your data"))
        }
    }

	#if(use.adjVAF){
		#if(! "CCF" %in% colnames(mafData)){
			#stop("Adjust VAF was not found in maf object. Check if CCF data was provided")
		#}
		#vaf.col = "VAF_adj" 
	#}else{
		#vaf.col = "VAF"
	#}
	mafData <-  mafData %>%
	    dplyr::filter(VAF > min.vaf,
	                  Patient_ID %in% patient.id) %>% 
	    tidyr::unite(
	        "mutation_id", 
	        c(
	            "Chromosome", 
	            "Start_Position", 
	            "Reference_Allele", 
	            "Tumor_Seq_Allele2"), 
	        sep = ":", 
	        remove = FALSE
	    ) %>% 
	    #dplyr::filter(Variant_Type == "SNP") %>%
	    dplyr::mutate(
	        totalDepth = Ref_allele_depth + Alt_allele_depth)
	
    if(withinType){
        filter.out <- mafData %>%
            dplyr::filter(Clonal_Status == "Subclonal") %>%
            dplyr::select(        	
                mutation_id,
                Tumor_Type,
                Tumor_Sample_Barcode,
                Patient_ID,
                VAF,
                #!!vaf.col,
                totalDepth)
        
        ## fresh patient.id
        patient.id <- unique(filter.out$Patient_ID)
        
        Fst.out <- filter.out %>% 
            dplyr::group_by(Patient_ID) %>%
            dplyr::group_map(
                ~groupByType(.,
                             min.vaf = min.vaf,
                             plot = plot,
                             use.circle = use.circle,
                             title = title,
                             withinType = withinType), 
                keep = TRUE) %>%
            rlang::set_names(patient.id)
    }
	else{
	    filter.out <- mafData %>%
	        dplyr::filter(Clonal_Status == "Subclonal") %>%
	        dplyr::select(        	
	            mutation_id,
	            Tumor_Sample_Barcode,
	            Patient_ID,
	            VAF,
	            #!!vaf.col,
	            totalDepth)
	    ## fresh patient.id
	    patient.id <- unique(filter.out$Patient_ID)
	    
	    Fst.out <- filter.out %>% 
	        dplyr::group_by(Patient_ID) %>%
	        dplyr::group_map(~fst.hudson.patient(.,
	                         min.vaf = min.vaf,
	                         plot = plot,
	                         use.circle = use.circle,
	                         title = title,
	                         withinType = withinType), 
	            keep = TRUE) %>%
	        rlang::set_names(patient.id)
	    
	    Fst.out <- Fst.out[!is.na(Fst.out)]
	}
        
  
  if(withinType){
      patient.list <- c()
      type.list <- c()
      avg.list <- c()
      name.list <- c()
      Fst.pair <- list()
      Fst.plot <- list()
      for(patient in names(Fst.out)){
          
          ## get result for each patient
          Fst.out.p <- Fst.out[[patient]]
          
          for(type in names(Fst.out.p)){
              ## get result for each patient in each type
              Fst.out.p.t <- Fst.out.p[[type]]
               
              name <- paste0(patient,"_",type) 
              patient.list <- c(patient.list,patient)
              type.list <- c(type.list,type)
              avg.list <- c(avg.list,Fst.out.p.t$Fst.avg)
              
              Fst.pair[[name]] <- Fst.out.p.t$Fst.pair
              Fst.plot[[name]] <- Fst.out.p.t$Fst.plot
          }
      }
      Fst.avg = data.frame(Patient_ID = patient.list,
                           Tumor_Type = type.list,
                           Fst.avg = avg.list)
      
      Fst.out=list(
          Fst.avg = Fst.avg,
          Fst.pair = Fst.pair,
          Fst.plot = Fst.plot
      )
      
  }
  else{
      Fst.out=list(
          Fst.avg = data.frame(Patient_ID = names(Fst.out), Fst.avg = unlist(lapply(Fst.out, function(x) x$Fst.avg))),
          Fst.pair = lapply(Fst.out, function(x) x$Fst.pair),
          Fst.plot = lapply(Fst.out, function(x) x$Fst.plot)
      )  
  }

  
  return(Fst.out)
}
=======
#' calFst
#' @description Genetic divergence between regions of subclonal sSNVs using the Weir and Cockerham method 
#' @references Sun R, Hu Z, Sottoriva A, et al. Between-region genetic divergence reflects the mode and tempo of tumor evolution. Nat Genet. 2017;49(7):1015-1024.
#' 
#' Bhatia G, Patterson N, Sankararaman S, Price AL. Estimating and interpreting FST: the impact of rare variants. Genomic Res. 2013;23(9):1514-1521.
#' 
#' @param maf Maf or MafList object generated by readMaf function.
#' @param patient.id Select the specific patients. Default: NULL, all patients are included.
#' @param chrSilent Chromosomes excluded in the analysis. e.g, chrX, chrY. Default NULL.
#' @param mutType Select proper variant classification you need. Default "All". Option: "nonSyn".
#' @param use.indel Logical value. Whether to use INDELs besides somatic SNVs. Default: TRUE.
#' @param min.vaf Specify the minimum VAF_adj, default is 0.02
#' @param use.adjVAF Calculate Fst by adjusted VAF.Default: FALSE. 
#' @param min.total.depth The minimum total allele depth for filtering variants. Default: 2.
#' @param withinTumor Calculate fst within types in each patients,default is FALSE.
#'  
#' @return A list contains Fst value of MRS and Hudson estimator of each sample-pair, respectively.
#'
#' @examples
#' calFst(maf)
#' @import dplyr
#' @export calFst

calFst <- function(
   maf, 
   patient.id = NULL, 
   chrSilent = NULL,
   mutType = "All",
   use.indel = TRUE,
   min.vaf = 0.02,
   min.total.depth = 2,
   use.adjVAF = FALSE,
   plot = TRUE,
   use.circle = TRUE,
   title = NULL,
   withinTumor = FALSE,
   number.cex = 8, 
   number.col = "#C77960"){
   
    
    if(min.total.depth <= 1){
        stop("Error: min.total.depth should be greater than 1")
    }
    if(withinTumor){
        clonalStatus <- "Subclonal"
    }else{
        clonalStatus <- NULL
    }
    
    if(class(maf) == "Maf"){
        maf_list <- list(maf)
    }else if(class(maf) == "MafList"){
        ## patient filter
        if(!is.null(patient.id)){
            maf_list <- subsetMafList(maf, patient.id = patient.id)
        }else{
            maf_list <- maf
        }
        
    }else{
        stop("Error: input should be either Maf or MafList object")
    }
    
    result <- list()
    for(m in maf_list){
        maf_data <- subsetMaf(m,
                         chrSilent = chrSilent,
                         mutType = mutType,
                         use.indel = use.indel,
                         min.vaf = min.vaf,
                         min.total.depth = min.total.depth,
                         clonalStatus = clonalStatus,
                         use.adjVAF = use.adjVAF)
            
        patient <- unique(maf_data$Patient_ID)
        if(!"VAF_adj" %in% colnames(maf_data)){
            stop("Error: adjust_VAF was not found in maf object.
		     Check if CCF data was provided or let adjusted.VAF be TRUE in function readMaf when VAF have been adjusted")
<<<<<<< HEAD
   }

   Fst.input <- mafData %>% 
       dplyr::mutate(totalDepth = Ref_allele_depth + Alt_allele_depth) %>% 
      dplyr::select(        	
         Mut_ID,
         Tumor_ID,
         Tumor_Sample_Barcode,
         Patient_ID,
         VAF_adj,
         totalDepth)
   
   ## fresh patient.id
   patient.id <- unique(Fst.input$Patient_ID)
   
   Fst.avg <- data.frame()
   Fst.pair <- list()
   if(plot){
       Fst.plot <- list()
   }
   for(patient in patient.id){
      patient.data <- subset(Fst.input, Patient_ID == patient)
      
      if(!withinTumor){
         patient.data$Tumor_ID <- "All"
      }
<<<<<<< HEAD
    }
    Fst.avg = data.frame(Patient_ID = patient.list,
                         Tumor_Type = type.list,
                         Fst.avg = avg.list)
    
    Fst.out=list(
      Fst.avg = Fst.avg,
      Fst.pair = Fst.pair,
      Fst.plot = Fst.plot
    )
    
  }
  else{
    Fst.out=list(
      Fst.avg = data.frame(Patient_ID = names(Fst.out), Fst.avg = unlist(lapply(Fst.out, function(x) x$Fst.avg))),
      Fst.pair = lapply(Fst.out, function(x) x$Fst.pair),
      Fst.plot = lapply(Fst.out, function(x) x$Fst.plot)
    )  
  }
  
  
  return(Fst.out)
<<<<<<< HEAD
}
>>>>>>> 76c155305e8a78ba390291810c7b38dad221073d
=======
}
>>>>>>> ac0983680247db451afb237ddccd3104fda3c2b9
=======
      ids <- unique(patient.data$Tumor_ID)
      for(id in ids){
         subdata <- subset(patient.data, Tumor_ID == id)
         if(withinTumor){
            if(length(unique(subdata$Tumor_Sample_Barcode))  < 2 ){
               message(paste0("Warnings: Only one sample was found of ", id,
                              " in ", patient, ". It you want to compare CCF between regions, withinTumor should be set as FALSE"))
               next
=======
        }
        
        Fst_input <- maf_data %>% 
            tidyr::unite(
                "Mut_ID",
                c(
                    "Chromosome",
                    "Start_Position",
                    "Reference_Allele",
                    "Tumor_Seq_Allele2"
                ),
                sep = ":",
                remove = FALSE
            ) %>% 
            dplyr::mutate(totalDepth = Ref_allele_depth + Alt_allele_depth) %>% 
            dplyr::select(        	
                Mut_ID,
                Tumor_ID,
                Tumor_Sample_Barcode,
                VAF_adj,
                totalDepth) %>% 
            ## remove NA(it may be caused by NA of CCF)
            dplyr::filter(!is.na(VAF_adj))
            
        if(!withinTumor){
            Fst_input$Tumor_ID <- "All"
        }
        ids <- unique(Fst_input$Tumor_ID)
        for(id in ids){
            subdata <- subset(Fst_input, Tumor_ID == id)
            if(withinTumor){
                if(length(unique(subdata$Tumor_Sample_Barcode))  < 2 ){
                    message(paste0("Warnings: Only one sample was found of ", id,
                                   " in ", patient, ". It you want to compare CCF between regions, withinTumor should be set as FALSE"))
                    next
                }
            }
            else{
                if(length(unique(subdata$Tumor_Sample_Barcode))  < 2 ){
                    message(paste0("Warnings: Only one sample was found in ", patient, "."))
                    next
                } 
            }
            
            ## pairwise heterogeneity
            samples <- as.character(unique(subdata$Tumor_Sample_Barcode))
            pairs <- combn(length(samples), 2, simplify = FALSE)
            
            dist_mat <- diag(1, nrow = length(samples), ncol = length(samples))
            rownames(dist_mat) <- samples
            colnames(dist_mat) <- samples
            
            for (pair in pairs){
                s  <- subset(subdata, Tumor_Sample_Barcode %in% c(samples[pair[1]],samples[pair[2]])) %>%
                    dplyr::select(-Tumor_ID) 
                maf.pair <- as.data.frame(s) %>% 
                    tidyr::pivot_wider(
                        names_from = Tumor_Sample_Barcode,       
                        values_from = c(VAF_adj, totalDepth),
                        values_fill = c(VAF_adj = 0, totalDepth = 0)
                    )
                colnames(maf.pair) <- c("Mut_ID", "vaf1", "vaf2", "depth1", "depth2")
                
                name <- paste(samples[pair[1]],samples[pair[2]],sep = "_")
                vafCol <- which(grepl("vaf", colnames(maf.pair)))
                maf.pair <- dplyr::mutate(
                    maf.pair, 
                    covariance = (vaf1-vaf2)^2-(vaf1*(1-vaf1))/(depth1-1)-(vaf2*(1-vaf2))/(depth2-1), 
                    sd = vaf1*(1-vaf2)+vaf2*(1-vaf1)
                )
                fst <- mean(maf.pair$covariance)/mean(maf.pair$sd)
                dist_mat[pair[1],pair[2]] <- dist_mat[pair[2],pair[1]] <- fst
            }
            ## average fst
            Fst.avg <- mean(dist_mat[upper.tri(dist_mat, diag = F)])
            Fst.pair <- dist_mat
            if(plot){
                ## get significant number
                if(is.null(title)){
                    min_value <- min(dist_mat[dist_mat!=1 & dist_mat!=0])
                    significant_digit <- gsub(pattern =  "0\\.0*","",as.character(min_value))
                    digits <- nchar(as.character(min_value)) - nchar(significant_digit) 
                    if(withinTumor){
                        title_id <- paste0("Fst of ",id," in ", patient, ": ",round(Fst.avg,digits))
                    }else{
                        title_id <- paste0("Fst of patient ", patient, ": ",round(Fst.avg,digits))
                    }
                }else{
                    title_id <- title
                }
                Fst.plot <- plotCorr(
                    dist_mat, 
                    use.circle, 
                    number.cex = number.cex,
                    number.col = number.col,
                    title = if(!is.null(title_id)) title_id else{NA} 
                )
>>>>>>> 02140d70bdd5a1cb0f873a66ce1fefbd1cee5ccd
            }
            
            if(withinTumor){
                name <- paste(patient,id,sep = "_")
            }else{
                name <- patient
            }
            
            if(plot){
                result[[name]] <- list(Fst.avg = Fst.avg,Fst.pair = Fst.pair,Fst.plot = Fst.plot)
            }else{
                result[[name]] <- list(Fst.avg = Fst.avg,Fst.pair = Fst.pair)
            }
<<<<<<< HEAD
         }
         
      }
   }
   
   if(plot){
      return(list(Fst.avg = Fst.avg,Fst.pair = Fst.pair,Fst.plot = Fst.plot))
   }else{
      return(list(Fst.avg = Fst.avg,Fst.pair = Fst.pair))
   }
   
   return(Fst.out)
}
>>>>>>> 39e9f6634fd04ac5527fd4780956390939c46f72
=======
            
        }
    }
    
    if(length(result) > 1){
        return(result)
    }else if(length(result) == 0){
        return(NA)
    }else{
        return(result[[1]])
    }
    
    return(result)
}
   
>>>>>>> 02140d70bdd5a1cb0f873a66ce1fefbd1cee5ccd
