################################################################################
#' calFst
#' @description Genetic divergence between regions of subclonal sSNVs using 
#' the Weir and Cockerham method 
#' @references Sun R, Hu Z, Sottoriva A, et al. Between-region 
#' genetic divergence reflects the mode and tempo of tumor evolution. 
#' Nat Genet. 2017;49(7):1015-1024.
#' 
#' Bhatia G, Patterson N, Sankararaman S, Price AL. 
#' Estimating and interpreting FST: the impact of rare variants. 
#' Genomic Res. 2013;23(9):1514-1521.
#' 
#' @param maf Maf or MafList object generated by \code{\link{readMaf}} function.
#' @param patient.id Select the specific patients. 
#' Default: NULL, all patients are included.
#' @param min.vaf Specify the minimum VAF_adj, default is 0.02
#' @param min.total.depth The minimum total allele depth for filtering variants. 
#' Default: 2.
#' @param plot Logical (Default:TRUE). Whether to show the plot, default TRUE
#' @param withinTumor Calculate fst within types in each patients,default is FALSE.
#' @param use.circle Logical (Default:TRUE). Whether to use "circle" 
#' as visualization method of correlation matrix
#' @param title The title of the plot. Default is "Nei's distance"
#' @param number.cex The size of text shown in correlation plot. Default 8.
#' @param number.col The color of text shown in correlation plot. 
#' Default "#C77960".
#' @param ... Other options passed to \code{\link{subMaf}}
#'  
#' @return A list contains Fst value of MRS and Hudson estimator of each sample-pair, respectively.
#'
#' @examples
#' maf.File <- system.file("extdata", "HCC6046.maf", package = "MesKit")
#' ccf.File <- system.file("extdata", "HCC6046.ccf.tsv", package = "MesKit")
#' maf <- readMaf(mafFile=maf.File, ccfFile = ccf.File, refBuild="hg19")
#' calFst(maf)
#' @import dplyr
#' @export calFst

calFst <- function(
  maf, 
  patient.id = NULL, 
  min.vaf = 0.02,
  min.total.depth = 2,
  plot = TRUE,
  withinTumor = FALSE,
  use.circle = TRUE,
  title = NULL,
  number.cex = 8, 
  number.col = "#C77960",...){
  
  
  if(min.total.depth <= 1){
    stop("Error: min.total.depth should be greater than 1")
  }
  if(withinTumor){
    clonalStatus <- "Subclonal"
  }else{
    clonalStatus <- NULL
  }
  
  ## check input data
  maf_list <- checkMafInput(maf, patient.id = patient.id)
  
  result <- list()
  for(m in maf_list){
    maf_data <- subMaf(m,
                          min.vaf=min.vaf,
                          min.total.depth=min.total.depth,
                          clonalStatus=clonalStatus,
                          ...)
    
    patient <- getMafPatient(m)
    if(nrow(maf_data) == 0){
      message("Warning :there was no mutation in ", patient, " after filtering.")
      next
    }
    if(!"VAF_adj" %in% colnames(maf_data)){
      stop("Error: adjust_VAF was not found in maf object.
		     Check if CCF data was provided or let adjusted.VAF be TRUE in function readMaf when VAF have been adjusted")
    }
    
    Fst_input <- maf_data %>% 
      tidyr::unite(
        "Mut_ID",
        c(
          "Chromosome",
          "Start_Position",
          "Reference_Allele",
          "Tumor_Seq_Allele2"
        ),
        sep = ":",
        remove = FALSE
      ) %>% 
      dplyr::mutate(totalDepth=Ref_allele_depth + Alt_allele_depth) %>% 
      dplyr::select(        	
        Mut_ID,
        Tumor_ID,
        Tumor_Sample_Barcode,
        VAF_adj,
        totalDepth) %>% 
      ## remove NA(it may be caused by NA of CCF)
      dplyr::filter(!is.na(VAF_adj))
    
    if(!withinTumor){
      Fst_input$Tumor_ID <- "All"
    }
    ids <- unique(Fst_input$Tumor_ID)
    for(id in ids){
      subdata <- subset(Fst_input, Tumor_ID == id)
      if(withinTumor){
        if(length(unique(subdata$Tumor_Sample_Barcode))  < 2 ){
          message(paste0("Warnings: Only one sample was found of ", id,
                         " in ", patient, ". If you want to compare CCF between regions, withinTumor should be set as FALSE"))
          next
        }
      }
      else{
        if(length(unique(subdata$Tumor_Sample_Barcode))  < 2 ){
          message(paste0("Warnings: Only one sample was found in ", patient, "."))
          next
        } 
      }
      
      ## pairwise heterogeneity
      samples <- as.character(unique(subdata$Tumor_Sample_Barcode))
      pairs <- combn(length(samples), 2, simplify=FALSE)
      
      dist_mat <- diag(1, nrow=length(samples), ncol=length(samples))
      rownames(dist_mat) <- samples
      colnames(dist_mat) <- samples
      
      for (pair in pairs){
        s  <- subset(subdata, Tumor_Sample_Barcode %in% c(samples[pair[1]],samples[pair[2]])) %>%
          dplyr::select(-Tumor_ID) 
        maf.pair <- as.data.frame(s) %>% 
          tidyr::pivot_wider(
            names_from = Tumor_Sample_Barcode,       
            values_from = c(VAF_adj, totalDepth),
            values_fill = list(VAF_adj=0, totalDepth=0)
          )
        colnames(maf.pair) <- c("Mut_ID", "vaf1", "vaf2", "depth1", "depth2")
        
        name <- paste(samples[pair[1]], samples[pair[2]], sep = "_")
        vafCol <- which(grepl("vaf", colnames(maf.pair)))
        maf.pair <- dplyr::mutate(
          maf.pair, 
          covariance = (vaf1-vaf2)^2-(vaf1*(1-vaf1))/(depth1-1)-(vaf2*(1-vaf2))/(depth2-1), 
          sd = vaf1*(1-vaf2)+vaf2*(1-vaf1)
        )
        fst <- mean(maf.pair$covariance)/mean(maf.pair$sd)
        dist_mat[pair[1], pair[2]] <- dist_mat[pair[2], pair[1]] <- fst
      }
      ## average fst
      Fst.avg <- mean(dist_mat[upper.tri(dist_mat, diag = FALSE)])
      Fst.pair <- dist_mat
      if(plot){
        ## get significant number
        if(is.null(title)){
          min_value <- min(dist_mat[dist_mat != 1 & dist_mat != 0])
          significant_digit <- gsub(pattern =  "0\\.0*", "", as.character(min_value))
          digits <- nchar(as.character(min_value)) - nchar(significant_digit) 
          if(withinTumor){
            title_id <- paste0("Fst of ", id, " in ", patient, ": ", 
                               round(Fst.avg, digits))
          }else{
            title_id <- paste0("Fst of patient ", patient, ": ", round(Fst.avg, digits))
          }
        }else{
          title_id <- title
        }
        Fst.plot <- plotCorr(
          dist_mat, 
          use.circle, 
          number.cex=number.cex,
          number.col=number.col,
          title = if(!is.null(title_id)) title_id else{NA} 
        )
      }
      
      if(withinTumor){
        name <- paste(patient, id, sep = "_")
      }else{
        name <- patient
      }
      
      if(plot){
        result[[name]] <- list(Fst.avg=Fst.avg, Fst.pair=Fst.pair, Fst.plot=Fst.plot)
      }else{
        result[[name]] <- list(Fst.avg=Fst.avg, Fst.pair=Fst.pair)
      }
      
    }
  }
  
  if(length(result) > 1){
    return(result)
  }else if(length(result) == 0){
    return(NA)
  }else{
    return(result[[1]])
  }
}