#' calNeiDist
#' @description Nei's distance of CCF for each sample-pair
#' @references Lee JK, Wang J, Sa JK, et al. Spatiotemporal genomic architecture informs precision oncology in glioblastoma. Nat Genet. 2017;49(4):594–599. doi:10.1038/ng.3806
#' 
#' @param maf an Maf object generated by readMaf function
#' @param patient.id select the specific patients. Default: NULL, all patients are included.
#' @param chrSilent Select chromosomes needed to be dismissed. Default NULL.
#' @param mutType select Proper variant classification you need. Default "All". Option: "nonSilent".
#' @param use.indel Logical value. Whether to use INDELs besides somatic SNVs. Default: TRUE.
#' @param min.vaf specify the minimum VAF, default is 0.08
#' @param plot logical (Default:TRUE). Whether to show the plot, default TRUE
#' @param use.circle logical (Default:TRUE). Whether to use "circle" as visualization method of correlation matrix
#' default TURE
#' @param title the title of the plot, default is "Nei's distance"
#' @param withinTumor calculate fst within types in each patients,default is FALSE.
#' 
#' @return Nei's genetic distance matrix and heatmap of sample-pairs from the same patient
#'
#' @examples
#' calNeiDist(maf)
#' @export calNeiDist

calNeiDist <- function(maf, 
    patient.id = NULL, 
    chrSilent = NULL,
    mutType = "All",
    use.indel = TRUE,
    min.vaf = 0.02,
    plot = TRUE, 
    use.circle = TRUE, 
    title = NULL,
    withinTumor = FALSE,
    number.cex = 8, 
    number.col = "#C77960") {
    
    
    if(withinTumor){
        clonalStatus <- "Subclonal"
    }else{
        clonalStatus <- NULL
    }
    maf <- subsetMaf(maf,
                     patient.id = patient.id,
                     chrSilent = chrSilent,
                     mutType = mutType,
                     use.indel = use.indel,
                     min.vaf = min.vaf,
                     clonalStatus = clonalStatus)
    mafData <- maf@data

    if(! "CCF" %in% colnames(mafData)){
        stop(paste0("Calculation of Nei’s distance requires CCF data." ,
            "No CCF data was found when generate Maf object with readMaf function"))
    }
    
    Nei_input <- mafData %>%
        dplyr::mutate(
            CCF = dplyr::if_else(is.na(CCF), 0, CCF)
        ) %>% 
        dplyr::select(
                Mut_ID,
                Tumor_ID,
                Tumor_Sample_Barcode,
                Patient_ID,
                CCF)
    
    ## fresh patient id
    patient.id <- unique(Nei_input$Patient_ID)
    
    Nei.dist.avg <- data.frame()
    Nei.dist <- list()
    if(plot){
        Nei.plot <- list()
    }
    for(patient in patient.id){
        patient.data <- subset(Nei_input, Patient_ID == patient)
        if(!withinTumor){
            patient.data$Tumor_ID <- "All"
        }
        ids <- unique(patient.data$Tumor_ID)
        for(id in ids){
            subdata <- subset(patient.data, Tumor_ID == id)
            if(withinTumor){
                if(length(unique(subdata$Tumor_Sample_Barcode)) < 2){
                    message(paste0("Warnings: Only one sample was found of ", id,
                                   " in ", patient, ". It you want to compare CCF between regions, withinTumor should be set as FALSE\n"))
                    next
                }
            }
            
            if(length(unique(subdata$Tumor_Sample_Barcode)) < 2){
                message(paste0("Warnings: Only one sample was found of ",patient,"."))
                return(NA)
            }
            
            subdata <- tidyr::pivot_wider(subdata,
                                     names_from = Tumor_Sample_Barcode,
                                     values_from = CCF,
                                     values_fill = list(CCF = 0)) %>%
                dplyr::select(-Patient_ID, -Mut_ID, -Tumor_ID)
            
            dist_mat <- diag(0, nrow = ncol(subdata), ncol = ncol(subdata))
            for (i in 1:(ncol(subdata) - 1)) {
                s1 <- colnames(subdata)[i]
                for (j in (i + 1):ncol(subdata)) {
                    s2 <- colnames(subdata)[j]
                    x <- as.vector(subdata[, i])
                    y <- as.vector(subdata[, j])
                    
                    x_ <- sum(x ^ 2 + (1 - x) ^ 2)
                    y_ <- sum(y ^ 2 + (1 - y) ^ 2)
                    
                    xy <- sum(x * y + (1 - x) * (1 - y))
                    
                    nei_dist <- -log(xy / sqrt(x_ * y_))
                    dist_mat[i, j] <- dist_mat[j, i] <- nei_dist
                }
            }
            rownames(dist_mat) <- colnames(subdata)
            colnames(dist_mat) <- colnames(subdata)
            
            avg <- mean(dist_mat)
            if(withinTumor){
                df_avg <- data.frame(Patient_ID = patient, Tumor_ID = id, Nei.dist.avg = avg)
                Nei.dist.avg <- rbind(Nei.dist.avg, df_avg)
            }else{
                df_avg <- data.frame(Patient_ID = patient, Nei.dist.avg = avg)
                Nei.dist.avg <- rbind(Nei.dist.avg, df_avg)
            }
            
            if(withinTumor){
                name <- paste(patient,id,sep = "_")
                Nei.dist[[name]] <- dist_mat
            }else{
                Nei.dist[[patient]] <- dist_mat
            }
            
            if(is.null(title)){
                if(withinTumor){
                    title <- paste0("Nei distance of ",id," in ", patient, ": ",round(avg,2))
                }else{
                    title <- paste0("Nei distance of patient ", patient, ": ",round(avg,2))
                }
            }
            
            if(plot){
                p <- plotCorr(
                    dist_mat, 
                    use.circle, 
                    number.cex = number.cex,
                    number.col = number.col,
                    title = if(!is.null(title)) title else{NA} 
                )
                if(withinTumor){
                    name <- paste(patient,id,sep = "_")
                    Nei.plot[[name]] <- p
                }else{
                    Nei.plot[[patient]] <- p
                }
            }

        }
    }
    
    if(plot){
        return(list(Nei.dist.avg = Nei.dist.avg, Nei.dist = Nei.dist, Nei.plot =  Nei.plot))           
    }else{
        return(list(Nei.dist.avg = Nei.dist.avg, Nei.dist = Nei.dist))           
    }

}
