#' classifyMut
#'
#' @param maf  the Maf object generated by readMaf
#' Classify SSNVs/Indels into Shared/P-shared/Private, Clonal/Subclonl
#' or Shared-Clonal/P-shared-Clonal/Private-Clonal/Shared-Subclonal/P-shared-SubClonal/Private-SubClonal 
#' @param class  the class which would be represented, default is "SP" (Shared/P-shared/Private),
#' other options: "CS" (Clonal/Subclonl) and "SPCS".
#' @param patient.id  select the specific patients. Default: NULL, all patients are included
#' @param geneList list of genes to restrict the analysis. Default NULL.
#' @param plot  TRUE(Default).
#' @param topGenesCount the number of genes print, default is 10
#' @param remove_empty_columns  whether remove the samples without alterations. Only works when plot is TRUE
#' @param remove_empty_rows  whether remove the genes without alterations. Only works when plot is TRUE
#' @param showColnames TRUE(Default). Show sample names of columns.
#' @param patientsCol  a list containing customized colors for distinct patients. Default: NULL.
#' @param bgCol  background grid color. Default: "#f0f0f0"
#' @examples
#' classifyMut(maf, class = "SP")
#' @export classifyMut


mutType_SP <- function(unique_barcode_count, total_barcode_count) {
    dplyr::case_when(
        unique_barcode_count == 1 ~ "Private",
        unique_barcode_count == total_barcode_count ~ "Shared",
        (1 < unique_barcode_count) &
            (unique_barcode_count < total_barcode_count) ~ "P_shared"
    )
}

mutType_SPCS <- function(unique_barcode_count, total_barcode_count, Status) {
    dplyr::case_when(
        unique_barcode_count == 1 ~ paste0("Private_", Status),
        unique_barcode_count == total_barcode_count ~ paste0("Shared_", Status),
        (1 < unique_barcode_count) &
            (unique_barcode_count < total_barcode_count) ~ paste0("P_shared_", Status)
    )    
}

multiHits <- function(x) {
    if (length(x) > 1) {
        return(paste0(paste0(x, collapse = ";"), ";Multi_hits"))
    }
    else {
        return(x)
    }
}

classifyMut <- function(maf,
     class = "SP",
     patient.id = NULL, 
     geneList = NULL, 
     plot =  TRUE,
     topGenesCount = 10,
     remove_empty_columns = TRUE,
     remove_empty_rows = TRUE,
     showColnames = TRUE,
     patientsCol = NULL,
     bgCol = "#f0f0f0") {
     
  class.options = c('SP', 'CS', 'SPCS')
  if(!class %in% class.options){
    stop("class can only be either 'SP', 'CS' or 'SPCS'")
  }
  
  
  if(is.null(patient.id)){
    patient.id = unique(maf@data$Patient_ID)
  }else{
    patient.setdiff <- setdiff(patient.id, unique(maf@data$Patient_ID))
    if(length(patient.setdiff) > 0){
      stop(paste0(patient.setdiff, " can not be found in your data"))
    }
  }
  
  maf_data <- maf@data %>%
    tidyr::unite(
      "mutation_id",
      c("Chromosome",
        "Start_Position",
        "Reference_Allele",
        "Tumor_Seq_Allele2"
      ),
      sep = ":",
      remove = FALSE
    ) %>%
    dplyr::filter(Patient_ID %in% patient.id)
  
  mutation_count <-
    maf_data %>%
    dplyr::group_by(Patient_ID, mutation_id) %>%
    dplyr::summarise(unique_barcode_count = dplyr::n_distinct(Tumor_Sample_Barcode))
  
  barcode_count <-
    maf_data %>%
    dplyr::group_by(Patient_ID) %>%
    dplyr::summarise(total_barcode_count = dplyr::n_distinct(Tumor_Sample_Barcode))
  
  maf_data <-
    suppressMessages(maf_data %>%
                       dplyr::left_join(mutation_count) %>%
                       dplyr::left_join(barcode_count))
  
  if(class == "SP"){
    maf_data <- maf_data %>%
      dplyr::mutate(
        Mutation_Type = mutType_SP(unique_barcode_count, total_barcode_count)
      )
  }else{
    if(! "Status" %in% colnames(maf_data)){
      stop(paste0("CCF data is necessary when class = ", class))
    }
    else{
      if(class == "CS"){
        maf_data <- maf_data %>%
          dplyr::mutate(
            Mutation_Type = Status
          )
      }
      else if(class == "SPCS"){
        maf_data <- maf_data %>%
          dplyr::filter(!is.na(Status)) %>%
          dplyr::mutate(Mutation_Type =
                          mutType_SPCS(unique_barcode_count, total_barcode_count, Status)
          )
      }
    }
  }
  
  maf_data <- maf_data %>%
    dplyr::select(Hugo_Symbol, Chromosome, 
                  Start_Position, End_Position,
                  Reference_Allele, Tumor_Seq_Allele2, 
                  Tumor_Sample_Barcode, Mutation_Type,
                  unique_barcode_count, Patient_ID
    )
  
  if (!is.null(geneList)) {
    geneSelect <- geneList
    
    maf_data <- maf_data %>%
      dplyr::rowwise() %>%
      dplyr::mutate(Driver_Mut = dplyr::if_else(
        any(strsplit(Hugo_Symbol, ",|;")[[1]] %in% geneSelect),
        TRUE,
        FALSE)) %>%
      dplyr::mutate(Hugo_Symbol = dplyr::if_else(
        Driver_Mut == TRUE,
        geneSelect[geneSelect %in% strsplit(Hugo_Symbol, ",|;")[[1]]][1],
        Hugo_Symbol))
  }
      
    if(plot){
        plotMutProfile(maf_data,
                        class = class,
                        topGenesCount = topGenesCount,
                        bgCol = bgCol,
                        patientsCol = patientsCol,                        
                        remove_empty_columns = remove_empty_columns,
                        remove_empty_rows = remove_empty_rows, 
                        showColnames = showColnames
        )

    }

    return(maf_data %>% dplyr::select(-unique_barcode_count))
}
