#' classifyMut
#'
#' @param maf the Maf object generated by readMaf
#' Classify SSNVs/Indels into Shared/P-shared/Private, Clonal/Subclonl
#' or Shared-Clonal/P-shared-Clonal/Private-Clonal/Shared-Subclonal/P-shared-SubClonal/Private-SubClonal 
#' @param driverGenesFile driver gene list. Default NULL
#' @examples
#' classifyMut(maf, class = "SP")
#' @export classifyMut


mutType_SP <- function(unique_sample_count, total_count) {
    dplyr::case_when(
        unique_sample_count == 1 ~ "Private",
        unique_sample_count == total_count ~ "Shared",
        (1 < unique_sample_count) &
            (unique_sample_count < total_count) ~ "P_shared"
    )
}


mutType_SPCS <- function(unique_sample_count, Status, total_count) {
    dplyr::case_when(
        unique_sample_count == 1 ~ paste0("Private_", Status),
        unique_sample_count == total_count ~ paste0("Shared_", Status),
        (1 < unique_sample_count) &
            (unique_sample_count < total_count) ~ paste0("P_shared_", Status)
    )    
}

multiHits <- function(x) {
    if (length(x) > 1) {
        return(paste0(paste0(x, collapse = ";"), ";Multi_hits"))
    }
    else {
        return(x)
    }
}

classifyMut <- function(maf, class = "SP", driverGenesFile = NULL) {
    
    class.options = c('SP', 'CS', 'SPCS')
        if(!class %in% class.options){
            stop("class can only be either 'SP', 'CS' or 'SPCS'")
    }

    samples_count <- length(unique(maf@data$Tumor_Sample_Barcode))      
    maf_data <- maf@data %>%
        tidyr::unite(
            "mutation_id",
            c("Chromosome",
              "Start_Position",
              "Reference_Allele",
              "Tumor_Seq_Allele2"
            ),
            sep = ":",
            remove = FALSE
        )        
    mutation_count <- maf_data %>%
        dplyr::group_by(mutation_id) %>%
        dplyr::summarise(unique_sample_count = dplyr::n_distinct(Tumor_Sample_Barcode))      
    maf_data <-
        suppressMessages(maf_data %>% dplyr::left_join(mutation_count))        
    
    if(class == "SP"){
        maf_data <- maf_data %>%
            dplyr::mutate(
                Mutation_Type = mutType_SP(unique_sample_count, samples_count)
                )
    }else{
        if(! "Status" %in% colnames(maf_data)){
            stop(paste0("CCF data is necessary when class = ", class))
        }
        else{
            if(class == "CS"){
                maf_data <- maf_data %>%
                    dplyr::mutate(
                    Mutation_Type = Status
                )
            }
            else if(class == "SPCS"){
                maf_data <- maf_data %>%
                    dplyr::filter(!is.na(Status)) %>%
                    dplyr::mutate(Mutation_Type =
                                mutType_SPCS(unique_sample_count, Status, samples_count)
                                )
            }
        }
    }
    maf_data <- maf_data %>%
        dplyr::select(Hugo_Symbol, Chromosome, 
                      Start_Position, End_Position,
                      Reference_Allele, Tumor_Seq_Allele2, 
                      Tumor_Sample_Barcode, Mutation_Type,
                      unique_sample_count
                      )
    
    if (!is.null(driverGenesFile)) {
        geneSelect <-
            read.table(driverGenesFile, col.names = "gene_Name", stringsAsFactors = FALSE)$gene_Name
        maf_data <- maf_data %>%
            dplyr::rowwise() %>%
            dplyr::mutate(Driver_Mut = dplyr::if_else(
                any(strsplit(Hugo_Symbol, ",|;")[[1]] %in% geneSelect),
                TRUE,
                FALSE)) %>%
            dplyr::mutate(Hugo_Symbol = dplyr::if_else(
                Driver_Mut == TRUE,
                geneSelect[geneSelect %in% strsplit(Hugo_Symbol, ",|;")[[1]]][1],
                Hugo_Symbol))
    }

    return(maf_data)
}