## Main Function for generating fishplot/timescape plot accoding to 
#' Get outputs from SCHISM and draw the fishplot
#' @description infers subclonal relationship based on clonevol and SCHISM. 
#' generate metastatic evolvogram by fishplot and timescape.s
#' 
#' @import htmlwidgets fishplot timescape clonevol
#' 
#' @param maf the classMaf object generated by readMaf.
#' @param inferMethod choose a method to infer suboclonal relationship. Default "clonevol".
#' @param plotOption choose the function to generate different tyoe of fishplot. Default "fishplot".
#' @param schismCellularityFile specify the directory of cluster.cellularity document. Default NULL. If inferMethod == "SCHISM", the directory is essential and should be one of the outputs generated by  SCHISM.
#' @param schismConsensusTree specify the directory of GA.consensusTree document. Default NULL. However, if inferMethod == "SCHISM", the parameter should be one of the outputs generated by  SCHISM.
#' @return the fishplot showing the evolution relationship predicted by SCHISM
#' 
#' @examples
#' \dontrun{
#' ## clonevol method
#' cloneFishPlot(maf, inferMethod="clonevol", plotOption="fishplot")
#' cloneFishPlot(maf, inferMethod="clonevol", plotOption="timescape")
#' ## SCHISM method
#' cloneFishPlot(maf, inferMethod="SCHISM", plotOption="fishplot", schismCellularityFile, schismConsensusTree)
#' cloneFishPlot(maf, inferMethod="SCHISM", plotOption="timescape", schismCellularityFile, schismConsensusTree)
#'}
#'
#' @export schism2Fishplot
#' @export clonevol2Fishplot
#' @export cloneFishPlot

## main function
cloneFishPlot <- function(maf, inferMethod="clonevol", plotOption="fishplot", 
                          schismCellularityFile=NULL, schismConsensusTree=NULL){
    if (inferMethod == "clonevol"){
        clusterTsvFile <- maf@ccf.cluster
        clonevol2fishplot(clusterTsvFile, plotOption)
    } else if (inferMethod == "SCHISM"){
        if (!is.null(schismCellularityFile) & !is.null(schismConsensusTree)){
            schism2Fishplot(schismCellularityFile, schismConsensusTree, plotOption)
        } else {
            warning("Missing schismCellularityFile or schismConsensusTree. The subclonal hierarchy results from SCHISM should be provided.")
        }
        
    }
}

## clonevol method
clonevol2Fishplot <- function(clusterTsvFile, plotOption){
    ## read mutations in cluster.tsv from PyClone and get targeted clusters
    cluster.tsv=read.table(clusterTsvFile, sep="\t",
                           stringsAsFactors=FALSE, header=TRUE)
    cluster_filter=cluster.tsv[!is.na(cluster.tsv$cluster_id) &
                                   cluster.tsv$size >= 5 &
                                   cluster.tsv$mean >= 0.1,]
    cluster_ls =unique(cluster_filter$cluster_id)
    
    ## read mutations within targeted clusters in loc.tsv file from PyClone
    loci.tsv=read.table(dir.loci.tsv, sep="\t",
                        stringsAsFactors=FALSE, header=TRUE)
    loci_filtered <- loci.tsv[which(loci.tsv$cluster_id %in% cluster_ls),]
    
    # loci_filtered <- na.omit(loci_filtered)
    sample.names <- unique(loci_filtered$sample_id)
    
    
    ## make sure the cluster id is continuous integer.
    for (cluster_newid in seq_along(ls.cluster_id)){
        loci_filtered$cluster_id[loci_filtered$cluster_id == ls.cluster_id[cluster_newid]] <- cluster_newid
    }
    
    ls.cluster_id <- unique(loci_filtered$cluster_id)
    
    ## separate each sample column
    dat.final <- data.frame()
    for (cluster in ls.cluster_id){
        dat.cluster <- loci_filtered[which(loci_filtered$cluster_id == cluster),]
        dat.samples <- data.frame(row.names=seq_along(dat.cluster[which(dat.cluster$sample_id == sample.names[1]),]$cellular_prevalence))
        for (sample in sample.names){
            col.sample <- data.frame(dat.cluster[which(dat.cluster$sample_id == sample),]$cellular_prevalence)
            names(col.sample) <- sample
            dat.samples <- cbind(dat.samples, col.sample)
        }
        dat.samples <- cbind(cluster, dat.samples)
        dat.final <- rbind(dat.final, dat.samples)
    }
    
    ##################### must be used ##################################
    ## infer subclonal relationship by clonevol
    consensusTree=infer.clonal.models(variants=dat.final,
                                      cluster.col.name='cluster',
                                      ccf.col.names=sample.names,
                                      cancer.initiation.model='monoclonal',
                                      subclonal.test='bootstrap',
                                      subclonal.test.model='non-parametric',
                                      founding.cluster=1,
                                      num.boots=1000,
                                      cluster.center='mean',
                                      min.cluster.vaf=0.01,
                                      # min probability of CCF is non-negative
                                      sum.p=0.05,
                                      # confidence interval estimate for CCF
                                      alpha=0.05)
    
    ## generate fishplot ouput which is used by fishplot and timescape
    fishPlotInput <- generateFishplotInputs(results=consensusTree)
    
    if (plotOption == "fishplot"){
        fishes = createFishPlotObjects(fishPlotInput)
        for (i in seq_along(fishes)){
            # pdf(paste('FISH', i, '.pdf', sep=""), width=8, height=5)
            fish = layoutClones(fishes[[i]])
            fish = setCol(fish,fishPlotInput$clonevol.clone.colors)
            if (length(sample.names) > 10) {
                fish=setCol(fish, as.character(sample.names))
            }
            fishPlot(fish,shape="spline", 
                     title.btm="Patient", 
                     cex.title=0.5,
                     vlines=seq(1, length(sample.names)), 
                     vlab=sample.names, pad.left=0.5)
            # dev.off()
        }
        message("FishPlot Done!")
    } else if (plotOption == "timescape"){
        ## preparation for timscape input: treeEdges
        fishPlotTreeEdges <- fishPlotInput$parents[[1]]
        fishPlotTreeEdges <- 
            c(fishPlotTreeEdges[which(fishPlotTreeEdges == 0)], 
              fishPlotTreeEdges[which(fishPlotTreeEdges > 0)]-1)
        branchEdge <- data.frame(matrix(ncol = 2, nrow = 0))
        treeEdges <- data.frame(matrix(ncol = 2, nrow = 0))
        cloneList <- 0:(length(fishPlotTreeEdges)-1)
        for (counter in seq_along(fishPlotTreeEdges)){
            if (cloneList[counter] == 0){
                next()
            }
            branchEdge <- data.frame(fishPlotTreeEdges[counter], 
                                     cloneList[counter])
            treeEdges <- rbind(treeEdges, branchEdge)
        }
        colnames(treeEdges) <- c("source", "target")
        
        ## preparation for timscape input: clonalPrev
        fishPlotClonalPrev <- as.data.frame(fishPlotInput$cell.fractions[[1]])
        clonalPrev <- data.frame(matrix(ncol = 3, nrow = 0))
        clonalPrevTemp <- data.frame(matrix(ncol = 3, nrow = 0))
        for (timepoint in colnames(fishPlotClonalPrev)){
            for (cloneIdCounter in seq_along(cloneList)){
                clonalPrevTemp <- 
                    data.frame(timepoint, cloneList[cloneIdCounter], 
                               fishPlotClonalPrev[, which(colnames(fishPlotClonalPrev) == timepoint)][cloneIdCounter])
                clonalPrev <- rbind(clonalPrev, clonalPrevTemp)
            }
        }
        colnames(clonalPrev) <- c("timepoint", "clone_id", "clonal_prev")
        ## print timescape
        timescapeobject <- timescape(clonalPrev, treeEdges, mutations="NA", clone_colours="NA",
                                     xaxis_title="Time Points", yaxis_title="Clonal Prevalence",
                                     phylogeny_title="Clonal Phylogeny", alpha=50,
                                     genotype_position="stack", perturbations="NA", sort=FALSE,
                                     show_warnings=TRUE, width=900, height=NULL)
        show(timescapeobject)
        saveWidget(timescapeobject, file="timescape.html")
        message("Timescape Plot Done!")
    }
}

## SCHISM method
schism2Fishplot <- function(schismCellularityFile, schismConsensusTree, plotOption){
    ## get cellularity infomation
    clusterCellularity=read.table(schismCellularityFile, sep="\t", 
                                  stringsAsFactors=FALSE, header=TRUE)
    
    if (plotOption == "fishplot"){
        ## get the list of samples and clusters in this function
        sampleLs=unique(clusterCellularity$sampleID)
        clusterLs=unique(clusterCellularity$clusterID)
        
        ## build the cancer celluarity/fraction table for fishplot
        frac.c=c()
        for (sampleName in sampleLs) {
            sample <- clusterCellularity[which(
                clusterCellularity$sampleID == sampleName), ]$cellularity
            frac.c <- c(frac.c,sample)
        }
        fracTable=matrix(frac.c*100, ncol=length(sampleLs))
        rownames(fracTable) <- clusterLs
        colnames(fracTable) <- sampleLs
        
        ## read the evolution relationship of different subclones
        GAconsensusTree=read.table(schismConsensusTree, sep="\t", 
                                   stringsAsFactors=FALSE, header=TRUE)
        
        ## figure out clusters which will be first nodes of the evolution tree
        endsLs=unique(GAconsensusTree[which(
            !GAconsensusTree$parent %in% GAconsensusTree$child), 1])
        
        ## make sure the first node of the tree would be first in parents
        if (!rownames(fracTable)[1] %in% endsLs){
            fracEnd=fracTable[which(
                rownames(fracTable) == endsLs[1]),]
            fracTable <- fracTable[-which(
                rownames(fracTable) == endsLs[1]),]
            fracTable <- rbind2(fracEnd, fracTable)
            rownames(fracTable)[1] <- endsLs[1]
        }
        
        ## rearrange subclonal evolution relationship as a vector(parents)
        parents <- c(seq_along(clusterLs))
        for (cluster_name in clusterLs){
            parentCluster <- GAconsensusTree[which(
                GAconsensusTree$child == cluster_name), 1]
            if (cluster_name %in% endsLs){
                parents[which(rownames(fracTable) == cluster_name)] <- 0
            } else{
                parents[which(
                    rownames(fracTable) == cluster_name)] <- which(
                        rownames(fracTable) == min(parentCluster))
            }
        }
        
        ## fishplot printing
        pdf('fish.pdf', width=8, height=5)
        fish=createFishObject(fracTable, parents, 
                              timepoints=c(seq_along(sampleLs)), 
                              fix.missing.clones=TRUE)
        if (length(clusterLs) > 10) {
            fish=setCol(fish, as.character(clusterLs))
        }
        fish=layoutClones(fish)
        fishPlot(fish, shape="spline", title.btm="PatientID", cex.title=0.5,
                 vlines=seq(1, length(sampleLs)), vlab=sampleLs, pad.left=0.5)
        dev <- dev.off()
        message("Fishplot Done!")
    } else if (plotOption == "timescape"){
        names(clusterCellularity) <- c("timepoint", "clone_id", 
                                       "clonal_prev", "sd")
        clonalPrev <- clusterCellularity[c("timepoint", "clone_id", 
                                           "clonal_prev")]
        
        ## make sure the timepoint is specific for the real data.
        # if (!is.null(SampleInfo)){
        #     ## read info file
        #     sampleInfoInput <- read.table(SampleInfo, quot="", 
        #                                   header=TRUE, fill=TRUE, 
        #                                   sep='', stringsAsFactors=FALSE)
        #     sampleInfoInput <- sampleInfoInput[order(sampleInfoInput$time),]
        #     timepoints <- sampleInfoInput$sample
        #     
        #     clonalPrevTemp <- data.frame()
        #     for (timepoint in timepoints){
        #         clonalPrevTimepoint <- clonalPrev[which(
        #             clonalPrev$timepoint == timepoint), ]
        #         clonalPrevTemp <- rbind(clonalPrevTemp, clonalPrevTimepoint)
        #     }
        #     clonalPrev <- clonalPrevTemp
        # } 
        
        ## read the evolution relationship of different subclones
        GAconsensusTree=read.table(schismConsensusTree, sep="\t", 
                                   stringsAsFactors=FALSE, header=TRUE)
        names(GAconsensusTree) <- c("source", "target")
        treeEdges <- GAconsensusTree
        
        timescapeobject <- timescape(clonalPrev, treeEdges, mutations="NA", 
                                     clone_colours="NA",
                                     xaxis_title="Time Point", 
                                     yaxis_title="Clonal Prevalence",
                                     phylogeny_title="Clonal Phylogeny", 
                                     alpha=50,
                                     genotype_position="stack", 
                                     perturbations="NA", 
                                     sort=FALSE,
                                     show_warnings=TRUE, 
                                     width=900, height=NULL)
        show(timescapeobject)
        saveWidget(timescapeobject, file="timescape.html")
        message("Timescape Plot Done!")
    }
}


