#' @title compareCCF
#' @description Density plot of CCFs in paired samples of the patient
#' This function requires CCF for clustering
#' @param maf a Maf object generated by readMaf function.
#' @param patient.id select the specific patients. Default: NULL, all patients are included.
#' @param chrSilent Select chromosomes needed to be dismissed. Default NULL.
#' @param mutType select Proper variant classification you need. Default "All". Option: "nonSyn".
#' @param use.indel Logical value. Whether to use INDELs besides somatic SNVs. Default: TRUE.
#' @param min.ccf the minimum value of CCF. Default: 0
#' @param pairByTumor pair by tumor types in each patients,default is FALSE.
#' 
#' 
#' @return
#' @examples
#' compareCCF(maf)
#' @export compareCCF


compareCCF <- function(maf,
                    patient.id = NULL,
                    chrSilent = NULL,
                    mutType = "All",
                    use.indel = TRUE,
                    pairByTumor = FALSE,
                    min.ccf = 0){
    
  
  maf <- subsetMaf(maf,
                   patient.id = patient.id,
                   chrSilent = chrSilent,
                   mutType = mutType,
                   use.indel = use.indel,
                   min.ccf = min.ccf)
  mafData <- maf@data
  ## check if ccf data is provided
  if(! "CCF" %in% colnames(mafData)){
    stop(paste0("ccfDensity function requires CCF data.\n",
    "No CCF data was found when generate Maf object."))
  }

  patient.id <- unique(mafData$Patient_ID)
  
  ccf.pair.list <- list()
  for(patient in patient.id){
      patient.data <- subset(mafData, Patient_ID == patient)
      
      if(pairByTumor){
          types <- unique(patient.data$Tumor_ID)
          if(length(types) < 2){
              message(paste0("Warnings: Only one tumor type was found in ",patient,". It you want to compare CCF between regions, pairByTumor should be set as FALSE"))
              next
          }
          ## get average CCF
          patient.data <- patient.data %>% 
              dplyr::mutate(CCF = Tumor_Average_CCF) %>%
              dplyr::filter(!is.na(CCF))
          pairs <- combn(length(types), 2, simplify = FALSE)  
      }else{
          samples <- unique(patient.data$Tumor_Sample_Barcode)
          if(length(samples) < 2){
              message(paste0("Warnings: Only one sample was found in",patient))
              next
          }
          pairs <- combn(length(samples), 2, simplify = FALSE)  
      }
      pair.name <- c()
      i <- 1
      for (pair in pairs){
          if(pairByTumor){
              S1 <- types[pair[1]]
              S2 <- types[pair[2]]  
          }else{
              S1 <- samples[pair[1]]
              S2 <- samples[pair[2]]
          }
          if(pairByTumor){
              ccf.pair <- patient.data %>% dplyr::filter(Tumor_ID %in% c(S1, S2)) %>%
                  tidyr::unite("Mut_ID2", c("Tumor_ID","Chromosome", "Start_Position", "Reference_Allele", "Tumor_Seq_Allele2"), 
                               sep = ":", remove = FALSE) %>% 
                  dplyr::distinct(Mut_ID2, .keep_all = TRUE) %>%
                  dplyr::select(Tumor_ID, Hugo_Symbol, Mut_ID, CCF) %>% 
                  tidyr::spread(key = Tumor_ID, value = CCF)
              
          }else{
              ccf.pair <- patient.data %>% dplyr::filter(Tumor_Sample_Barcode %in% c(S1, S2)) %>%
                  dplyr::select(Tumor_Sample_Barcode, Hugo_Symbol, Mut_ID, CCF) %>% 
                  tidyr::spread(key = Tumor_Sample_Barcode, value = CCF)
          }
          if(nrow(ccf.pair) == 0){
              message(paste0("Warning: No shared mutaions were detected between ",S1, " and ", S2) )
              next
          }
          v1 <- as.vector(ccf.pair[,3])
          v2 <- as.vector(ccf.pair[,4])
          ccf.pair <- ccf.pair[(!is.na(v1)&!is.na(v2)),]
          pair.name <- paste(S1,"-",S2, sep = "")
          ccf.pair.list[[patient]][[pair.name]] <- ccf.pair
      }
      
  }
  

  return(ccf.pair.list)      
}
