#' compareJSI
#'
#' @description The Jaccard similarity index (JSI) is applied to distinguish monoclonal versus polyclonal seeding in metastases
#' @references Hu Z, Li Z, Ma Z, Curtis C: Pan-cancer analysis of clonality and the timing of systemic spread in paired primary tumors and metastases. bioRxiv 2019.
#' @param maf a Maf object generated by readMaf function.
#' @param patient.id select the specific patients. Default: NULL, all patients are included
#' @param chrSilent Select chromosomes needed to be dismissed. Default NULL.
#' @param mutType select Proper variant classification you need. Default "All". Option: "nonSyn".
#' @param use.indel Logical value. Whether to use INDELs besides somatic SNVs. Default: TRUE.
#' @param min.vaf specify the minimum VAF_adj, default is 0.08
#' @param plot logical (Default:TRUE). Whether to show the plot, default TRUE
#' @param use.circle logical (Default:TRUE). Whether to use "circle" as visualization method of correlation matrix
#' @param title the title of the plot, default is "Jaccard similarity"
#'
#' @examples
#' compareJSI(maf)
#' @return correlation matrix and heatmap via Jaccard similarity coefficient method
#' @export compareJSI

compareJSI <- function(
   maf, 
   patient.id = NULL,
   pairByTumor = FALSE,
   min.vaf = 0.02,
   plot = TRUE, 
   use.circle = TRUE, 
   title = NULL,
   number.cex = 8, 
   number.col = "#C77960") {
   
   mafData <- maf@data
   
   if(! "CCF" %in% colnames(mafData)){
      stop(paste0("Error: calculation of Jaccard similarity requires CCF data." ,
                  "No CCF data was found when generate Maf object."))
   }
   
   if(min.vaf < 0){
      stop("Error: min.vaf must be greater than 0")
   }
   
   if(is.null(patient.id)){
      patient.id = unique(mafData$Patient_ID)
   }else{
      patient.setdiff <- setdiff(patient.id, unique(mafData$Patient_ID))
      if(length(patient.setdiff) > 0){
         stop(paste0("Patient ", patient.setdiff, " can not be found in your data"))
      }
      mafData <- mafData[Patient_ID %in% patient.id,]
   }
   
   JSI_input <-  mafData %>%
      dplyr::select(
         Mut_ID,
         Tumor_ID,
         Tumor_Sample_Barcode,
         Patient_ID,
         Clonal_Status,
         VAF_adj) %>%
      dplyr::filter(VAF_adj > min.vaf)
   
   ## fresh patient.id
   patient.id <- unique(JSI_input$Patient_ID)
   
   JSI.multi <- data.frame()
   JSI.pair <- list()
   if(plot){
       JSI.plot <- list()
   }

   for(patient in patient.id){
      patient.data <- subset(JSI_input, Patient_ID == patient)
      
      if(pairByTumor){
         if(length(unique(patient.data$Tumor_ID))  < 2 ){
            message(paste0("Warnings: Only one tumor type was found of ",patient, ". It you want to compare CCF between regions, pairByTumor should be set as FALSE"))
            next
         }
      }else{
         if(length(unique(patient.data$Tumor_Sample_Barcode))  < 2 ){
            message(paste0("Warnings: Only one sample was found in ", patient, "."))
            next
         } 
      }
      
      ## pairwise heterogeneity
      if(pairByTumor){
         tumors <- as.character(unique(patient.data$Tumor_ID))
         pairs <- combn(length(tumors), 2, simplify = FALSE)
         dist_mat <- diag(1, nrow = length(tumors), ncol = length(tumors))
         rownames(dist_mat) <- tumors
         colnames(dist_mat) <- tumors
      }else{
         samples <- as.character(unique(patient.data$Tumor_Sample_Barcode))
         pairs <- combn(length(samples), 2, simplify = FALSE)
         dist_mat <- diag(1, nrow = length(samples), ncol = length(samples))
         rownames(dist_mat) <- samples
         colnames(dist_mat) <- samples
      }
      
      PC_1.list <- c()
      PC_2.list <- c()
      SS_12.list <- c()
      for (pair in pairs){
         
         if(pairByTumor){
            name <- paste(tumors[pair[1]],tumors[pair[2]], sep = "_")
            vaf.pair <- subset(patient.data, Tumor_ID %in% c(tumors[pair[1]],tumors[pair[2]])) %>%
               tidyr::unite("Mut_ID2",
                            c("Mut_ID",
                              "Tumor_ID"),
                            sep = ":",
                            remove = FALSE
               ) %>%
               dplyr::distinct(Mut_ID2, .keep_all = T) %>%
               dplyr::select(Mut_ID, Tumor_ID, Clonal_Status, VAF_adj) %>% 
               tidyr::pivot_wider(
                  names_from = Tumor_ID,       
                  values_from = c(VAF_adj, Clonal_Status),
                  values_fill = c(VAF_adj = 0, Clonal_Status = 'NA')
               ) %>%
               dplyr::ungroup()
            colnames(vaf.pair) <- c("Mut_ID", "vaf1", "vaf2", "status1", "status2")
         }
         else{
            name <- paste(samples[pair[1]],samples[pair[2]], sep = "_")
            vaf.pair <- subset(patient.data, Tumor_Sample_Barcode %in% c(samples[pair[1]],samples[pair[2]])) %>%
               dplyr::select(Mut_ID, Tumor_Sample_Barcode, Clonal_Status, VAF_adj) %>% 
               tidyr::pivot_wider(
                  names_from = Tumor_Sample_Barcode,       
                  values_from = c(VAF_adj, Clonal_Status),
                  values_fill = c(VAF_adj = 0, Clonal_Status = 'NA')
               ) %>%
               dplyr::ungroup()
            colnames(vaf.pair) <- c("Mut_ID", "vaf1", "vaf2", "status1", "status2")
         }
         
         vaf.pair <- vaf.pair %>% 
            dplyr::filter(vaf1 + vaf2 !=0) %>% 
            data.table::setDT()
         PC_1 <- nrow(vaf.pair[status1 == "Clonal" & vaf1>0 & vaf2==0])
         PC_1.list <- c(PC_1.list, PC_1)
         PC_2 <- nrow(vaf.pair[status2 == "Clonal" & vaf1==0 & vaf2>0])
         PC_2.list <- c(PC_2.list, PC_2)
         SS_12 = nrow(vaf.pair[status2=="Subclonal" & status1 == "Subclonal" & vaf1>0 & vaf2>0 ])
         SS_12.list <- c(SS_12.list, SS_12)
         jsi <- SS_12/(PC_1+PC_2+SS_12)
         if(is.nan(jsi)){
            dist_mat[pair[1],pair[2]] <- dist_mat[pair[2],pair[1]] <- 0
         }
         else{
            dist_mat[pair[1],pair[2]] <- dist_mat[pair[2],pair[1]] <- jsi
            
         }
      }
      
      multi <- mean(SS_12.list)/(mean(PC_1.list) + mean(PC_2.list) + mean(SS_12.list))
      multi <- ifelse(is.nan(multi),0,multi)
      df_multi <- data.frame(Patient_ID = patient, JSI.multi = multi)
      JSI.multi <- rbind(JSI.multi, df_multi)
      
      JSI.pair[[patient]] <- dist_mat
      
      if(is.null(title)){
         title <- paste0("JSI of patient ", patient, ": ",round(multi,2))
      }
      if(plot){
         values <- sort(unique(as.numeric(dist_mat))) 
         if(length(values[values!=0 &values !=1]) == 0){
            message(paste0("Warnings: there is no JSI within (0,1),can not plot JSI for ", patient, "."))
            next
         }
         p <- plotCorr(
            dist_mat, 
            use.circle, 
            number.cex = number.cex,
            number.col = number.col,
            title = if(!is.null(title)) title else{NA} 
         )
         JSI.plot[[patient]] <- p
      }
         
   }
   
   
   ## none result
   if(nrow(JSI.multi) == 0){
       return(NA)
   }
   
   if(plot){
      return(list(JSI.multi = JSI.multi, JSI.pair = JSI.pair, JSI.plot = JSI.plot))
   }else{
      return(list(JSI.multi = JSI.multi, JSI.pair = JSI.pair))
   }
   
   
}