#' @title getPhyloTree
#' 
#' @param a Maf object generated by readMaf function
#' @param method Approach to construct phylogenetic trees.Choose one of "NJ"(Neibor-Joining),"MP"(maximum parsimony),"ML"(maximum likelihood),""FASTME.ols" or "FASTME.bal". 
#' @param min.vaf the minimum value of vaf
#' @param min.ccf the minimum value of CCF
#' @param bootstrap.rep.num bootstrap iterations.
#' @param patient.id select the specific patients. Default: NULL, all patients are included.
#' @param chrSilent Select chromosomes needed to be dismissed. Default NULL.
#' @param mutType select Proper variant classification you need. Default "All". Option: "nonSilent".
#' @param use.indel Logical value. Whether to use INDELs besides somatic SNVs. Default: TRUE.
#' 
#' 
#' @examples
#' maf.File <- system.file("extdata", "HCC6046.maf", package = "MesKit")
#' ccf.File <- system.file("extdata", "HCC6046.CCF.txt", package = "MesKit")
#' maf <- readMaf(mafFile=maf.File, refBuild="hg19")
#' phyloTree <- getPhyloTree(maf)
#' @return  a list of class PhyloTree
#' @import phangorn ape
#' @exportClass phyloTree
#' @export getPhyloTree

getPhyloTree <- function(maf, 
                      method = "NJ",
                      min.vaf = 0.02, 
                      min.ccf = 0,
                      bootstrap.rep.num = 100,
                      patient.id = NULL,
                      chrSilent = NULL,
                      mutType = "All",
                      use.indel = TRUE){
  
  method <- match.arg(method, choices = c("NJ","MP","ML","FASTME.ols","FASTME.bal"), several.ok = FALSE)
  maf <- subsetMaf(maf,
                   patient.id = patient.id,
                   chrSilent = chrSilent,
                   mutType = mutType,
                   use.indel = use.indel,
                   min.vaf = min.vaf,
                   min.ccf = min.ccf)

  mafData <- maf@data
  refBuild <- maf@ref.build
  mafData$Patient_ID <- as.character(mafData$Patient_ID)
  dat.list <- split(mafData, mafData$Patient_ID)
  phyloTree.list <- list()
  phyloTree.list <- lapply(dat.list, doGetPhyloTree,
                           refBuild = refBuild,
                           method = method,
                           bootstrap.rep.num = bootstrap.rep.num)
  return(phyloTree.list)
}