#' @title getPhyloTree
#' 
#' @param a Maf object generated by readMaf function
#' @param method Approach to construct phylogenetic trees.Choose one of "NJ"(Neibor-Joining),"MP"(maximum parsimony),"ML"(maximum likelihood),""FASTME.ols" or "FASTME.bal". 
#' @param min.vaf the minimum value of vaf
#' @param min.ccf the minimum value of CCF
#' @param bootstrap.rep.num bootstrap iterations.
#' @param patient.id select the specific patients. Default: NULL, all patients are included.
#' @param chrSilent Select chromosomes needed to be dismissed. Default NULL.
#' @param mutType select Proper variant classification you need. Default "All". Option: "nonSyn".
#' @param use.indel Logical value. Whether to use INDELs besides somatic SNVs. Default: TRUE.
#' 
#' 
#' @examples
#' maf.File <- system.file("extdata", "HCC6046.maf", package = "MesKit")
#' ccf.File <- system.file("extdata", "HCC6046.CCF.txt", package = "MesKit")
#' maf <- readMaf(mafFile=maf.File, refBuild="hg19")
#' phyloTree <- getPhyloTree(maf)
#' @return  a list of class PhyloTree
#' @import phangorn ape
#' @exportClass phyloTree_list
#' @exportClass phyloTree
#' @export getPhyloTree

getPhyloTree <- function(maf, 
                      method = "NJ",
                      min.vaf = 0.02, 
                      min.ccf = NULL,
                      bootstrap.rep.num = 100,
                      patient.id = NULL,
                      chrSilent = NULL,
                      mutType = "All",
                      use.indel = TRUE){
  
  method <- match.arg(method, choices = c("NJ","MP","ML","FASTME.ols","FASTME.bal"), several.ok = FALSE)
  
  if(class(maf) == "classMaf"){
      maf_list <- list(maf)
  }else if(class(maf) == "classMaf_list"){
      ## patient filter
      if(!is.null(patient.id)){
          maf_list <- subsetMaf_list(maf, patient.id = patient.id)
      }else{
          maf_list <- maf@patient.list
      }
  }else{
      stop("maf should be either classMaf or classMaf_list")
  }
  
  phyloTree_patient_list <- list()
  for(m in maf_list){
      maf_data <- subsetMaf(m,
                     chrSilent = chrSilent,
                     mutType = mutType,
                     use.indel = use.indel,
                     min.vaf = min.vaf,
                     min.ccf = min.ccf)
      
      refBuild <- m@ref.build
      patient <- unique(maf_data$Patient_ID)
      
      ## information input
      binary.matrix <- getMutMatrix(maf_data, use.ccf = FALSE)
      if("CCF" %in% colnames(maf_data)){
          ccf.matrix <- getMutMatrix(maf_data, use.ccf = TRUE)
      }else{
          ccf.matrix <- matrix() 
      }
      mut_dat <- t(binary.matrix)
      if(method == "NJ"){
          matTree <- nj(dist.gene(mut_dat))
          bootstrap.value <- ape::boot.phylo(matTree, mut_dat, function(e)nj(dist.gene(e)),B = bootstrap.rep.num,quiet = T)/(bootstrap.rep.num)*100
      }else if(method == "MP"){
          matTree <- byMP(mut_dat)
          bootstrap.value <- ape::boot.phylo(matTree, mut_dat, function(e){byMP(e)},B = bootstrap.rep.num,quiet = T)/(bootstrap.rep.num)*100 
      }else if(method == "ML"){
          matTree <- byML(mut_dat)
          bootstrap.value <- ape::boot.phylo(matTree, mut_dat, function(e)byML(e),B = bootstrap.rep.num,quiet = T)/(bootstrap.rep.num)*100
      }else if(method == "FASTME.bal"){
          matTree <- ape::fastme.bal(dist.gene(mut_dat))
          bootstrap.value <- ape::boot.phylo(matTree, mut_dat, function(e) ape::fastme.bal(dist.gene(e)),B = bootstrap.rep.num,quiet = T)/(bootstrap.rep.num)*100
      }else if(method == "FASTME.ols"){
          matTree <- ape::fastme.ols(dist.gene(mut_dat))
          bootstrap.value <- ape::boot.phylo(matTree, mut_dat, function(e) ape::fastme.ols(dist.gene(e)),B = bootstrap.rep.num,quiet = T)/(bootstrap.rep.num)*100
      }
      branch.id <- readPhyloTree(matTree)
      
      mut.branches_types <- treeMutationalBranches(maf_data, branch.id, binary.matrix)
      mut.branches <- mut.branches_types$mut.branches
      branch.type <- mut.branches_types$branch.type
      
      phylo.tree <- new('phyloTree',
                        patientID = patient, tree = matTree, 
                        binary.matrix = binary.matrix, ccf.matrix = ccf.matrix, 
                        mut.branches = mut.branches, branch.type = branch.type,
                        refBuild = refBuild,
                        bootstrap.value = bootstrap.value, method = method)
      phyloTree_patient_list[[patient]] <- phylo.tree
  }
  
  if(length(phyloTree_patient_list) > 1){
      phyloTree_list <- new(
          'phyloTree_list',
          patient.list = phyloTree_patient_list
      )
      return(phyloTree_list)
  }else if(length(phyloTree_patient_list) == 1){
      return(phyloTree_patient_list[[1]])
  }else{
      return(NA)
  }
}

## Prevent class 'phylo' from not existing
setClass('phylo')
## set phyloTree class
setClass('phyloTree', slots = c(
    patientID = 'character', 
    tree = 'phylo',
    bootstrap.value = 'numeric',    
    method = 'character', 
    binary.matrix = 'matrix', 
    ccf.matrix = 'matrix', 
    mut.branches = 'data.frame', 
    branch.type = 'data.frame',
    refBuild = 'character'
    
))

## set class for multiple phyloTree
setClass('phyloTree_list', slots = c(
    patient.list = 'list'
))