#' @title getPhyloTree
#' 
#' @import phangorn
#' @param a Maf object generated by readMaf function
#' @param method Approach to construct phylogenetic trees.Choose one of "NJ"(Neibor-Joining),"MP"(maximum parsimony),"ML"(maximum likelihood),""FASTME.ols" or "FASTME.bal". 
#' @param minVaf the minimum value of vaf
#' @param maxVaf the maximum value of vaf
#' @param minCCF the minimum value of CCF
#' @param bootstrap.rep.num bootstrap iterations.
#' 
#' 
#' @examples
#' maf.File <- system.file("extdata", "HCC6046.maf", package = "MesKit")
#' ccf.File <- system.file("extdata", "HCC6046.CCF.txt", package = "MesKit")
#' maf <- readMaf(mafFile=maf.File, refBuild="hg19")
#' njtree <- getPhyloTree(maf)
#' @return  an object of class PhyloTree
#' @exportClass phyloTree
#' @export getPhyloTree


# set NJtree object
getPhyloTree <- function(maf, 
                      method = "NJ",
                      minVaf=0.02, 
                      maxVaf=1,
                      minCCF= NULL,
                      bootstrap.rep.num = 100){
  
  method.options <- c("NJ","MP","ML","FASTME.ols","FASTME.bal")
  if(!method %in% method.options){
      stop("method can only be either 'NJ','ML' or 'MP','FASTME.ols','FASTME.bal'")
  }
  maf.dat <- maf@data

  if (max(maf.dat$VAF, na.rm=TRUE) > 1){
    maf.dat$VAF <- maf.dat$VAF/100
  }
  if(!is.null(minCCF)){
      if("CCF" %in% colnames(maf)){
          maf.dat <- maf.dat[CCF > minCCF, ]
      }
      else{
          stop("CCF information not found")
      }
  }
  maf.dat <- maf.dat[which(maf.dat$VAF > minVaf & maf.dat$VAF < maxVaf), ]
  ## information input
  patientID <- maf@patientID
  refBuild <- paste("BSgenome.Hsapiens.UCSC.", maf@ref.build, sep = "")
  binary.matrix <- getMutMatrix(maf.dat, use.ccf = FALSE)
  if("CCF" %in% colnames(maf.dat)){
      ccf.matrix <- getMutMatrix(maf.dat, use.ccf = TRUE)
  }else{
      ccf.matrix <- matrix() 
  }
  mut_dat <- t(binary.matrix)
  if(method == "NJ"){
      matTree <- nj(dist.gene(mut_dat))
      numRoot <- which(matTree$tip.label == "NORMAL")
      bootstrap.value <- ape::boot.phylo(matTree, mut_dat, function(e) root(nj(dist.gene(e)),numRoot),B = bootstrap.rep.num,quiet = T)
  }
  else if(method == "MP"){
      matTree <- byMP(mut_dat)
      numRoot <- which(matTree$tip.label == "NORMAL")
      bootstrap.value <- ape::boot.phylo(matTree, mut_dat, function(e) root(byMP(e),numRoot),B = bootstrap.rep.num,quiet = T) 
  }
  else if(method == "ML"){
      matTree <- byML(mut_dat)
      numRoot <- which(matTree$tip.label == "NORMAL")
      bootstrap.value <- ape::boot.phylo(matTree, mut_dat, function(e) root(byML(e),numRoot),B = bootstrap.rep.num,quiet = T)
  }
  else if(method == "FASTME.bal"){
      matTree <- ape::fastme.bal(dist.gene(mut_dat))
      numRoot <- which(matTree$tip.label == "NORMAL")
      bootstrap.value <- ape::boot.phylo(matTree, mut_dat, function(e) root(ape::fastme.bal(dist.gene(e)),numRoot),B = bootstrap.rep.num,quiet = T)
  }
  else if(method == "FASTME.ols"){
      matTree <- ape::fastme.ols(dist.gene(mut_dat))
      numRoot <- which(matTree$tip.label == "NORMAL")
      bootstrap.value <- ape::boot.phylo(matTree, mut_dat, function(e) root(ape::fastme.ols(dist.gene(e)),numRoot),B = bootstrap.rep.num,quiet = T)
  }
  branchAlias <- readPhyloTree(matTree)
  mut.branches <- .treeMutationalBranches(maf.dat, branchAlias, binary.matrix)
  phylo.tree <- new('phyloTree', patientID = patientID, tree = matTree, 
                    binary.matrix = binary.matrix, ccf.matrix = ccf.matrix, 
                    mut.branches = mut.branches, refBuild = maf@ref.build,
                    bootstrap.value = bootstrap.value)
  return(phylo.tree)
}

byMP <- function(mut_dat){
    matTree <- nj(dist.gene(mut_dat))
    tree_dat <- phangorn::as.phyDat(mut_dat, type="USER", levels = c(0, 1))
    tree_pars <- suppressMessages(phangorn::optim.parsimony(matTree, tree_dat,trace = F)) 
    matTree <- phangorn::acctran(tree_pars, tree_dat)
    return(matTree)
}

byML <- function(mut_dat){
    matTree <- nj(dist.gene(mut_dat))
    tree_dat <- phangorn::as.phyDat(mut_dat, type="USER", levels = c(0, 1))
    fitJC <- phangorn::pml(matTree, tree_dat)
    fitJC <- phangorn::optim.pml(fitJC,control = pml.control(trace = F))
    matTree <- fitJC$tree
    return(matTree)
}
#Prevent class 'phylo' from not existing
setClass('phylo')
setClass('phyloTree', slots = c(tree = 'phylo', patientID = 'character', binary.matrix = 'matrix', mut.branches = 'list', ccf.matrix = 'matrix', refBuild = 'character',
                                bootstrap.value = 'numeric'))
