#' Check Mutational Signature for each branch of phylogenetic tree
#' @description Figure out the intersection of the variants on different branches
#' by .treeMutationalBranches and combine the intersection information into njtree
#' object. Check the mutational signatures for each branch of phylogenetic tree
#' by using treeMutationalSig and return a data.frame about the most likely muational
#' signatures.
#' 
#' @import reshape2 grDevices graphics utils deconstructSigs cowplot dplyr
#' @import BSgenome BSgenome.Hsapiens.UCSC.hg19
#' @import GenomeInfoDbData
#' @importFrom GenomeInfoDb seqnames
#' @importFrom dplyr filter_at
#' 
#' @param njtree the njtree object generated by NJtree function.
#' @param refBuild BSgenome.Hsapiens.UCSC reference. Default "hg19". Full genome sequences for Homo sapiens (Human) as provided by UCSC.
#' @param driverGenesFile the directory of the driver gene list. Default NULL.
#' @param mutThreshold the threshold for the variants in a branch. Default 50.
#' @return data frame of each set/branch's mutational signature.
#' 
#' @examples
#' ## data information
#' maf.File <- system.file("extdata/multi_lesion/maf", "311252.maf", package = "Meskit")
#' sampleInfo.File <- system.file("extdata/multi_lesion", "sample_info.txt", package = "Meskit")
#' pyCloneCluster <- system.file("extdata/multi_lesion/ccf", "311252.cluster.tsv", package = "Meskit")
#' pyCloneLoci <- system.file("extdata/multi_lesion/ccf", "311252.loci.tsv", package = "Meskit")
#' maf <- readMaf(patientID = "311252", mafFile = maf.File, sampleInfoFile = sampleInfo.File, refBuild = "hg19")
#' njtree <- getNJtree(maf)
#' ## general
#' signature1 <- treeMutationalSig(njtree, mutThreshold=50)
#' ## use driverGenesFile
#' putativeDriverGenes.File <- system.file("extdata/multi_lesion", "putative_driver_genes.txt", package = "Meskit")
#' signature2 <- treeMutationalSig(njtree, driverGenesFile=putativeDriverGenes.File, mutThreshold=50)
#' ## use different signature reference
#' signature3 <- treeMutationalSig(njtree, mutThreshold=50, signaturesRef="signatures.nature2013")
#' 
#' @export treeMutationalSig
#'

## Mutational Signature function
treeMutationalSig <- function(njtree, driverGenesFile=NULL, mutThreshold=50, 
                              signaturesRef="signatures.cosmic",
                              plot.Signatures=FALSE){
    ## refBuild limitation: only hg19 or hg38
    refBuild <- njtree@refBuild
    if (!((refBuild == "hg19") | (refBuild == "hg38"))){
        stop(error="Error: refBuild's value may be incorrect. 
             refBuild could be only set as \"hg19\" or \"hg38\".")
    } else {
        refBuild <- paste("BSgenome.Hsapiens.UCSC.", refBuild, sep = "")
    }
    
    ## get branches information from njtree object
    mutBranches <- njtree@mut_branches
    patientID <- njtree@patientID
    branchesName <- names(mutBranches)
    branchesNameList <- strsplit(branchesName, split='âˆ©')
    
    ## regain the data frame of all branches with Sample
    mutSigRef <- data.frame()
    for (branchName in branchesName){
        mutBranch <- mutBranches[[branchName]]
        mutSigRef <- rbind(mutBranch, mutSigRef)
    }
    colnames(mutSigRef) <- c("Sample",  "chr", "pos", "pos_end",
                             "ref", "alt", "Hugo_Symbol", "mut_id")
    ## get the mutational signature of the branch 
    mutSigsOutput <- data.frame()
    lsPicName <- c()
    for (branchCounter in length(branchesNameList):1){
        ## generate a single branch
        branch <- Filter(Negate(is.na), 
                         branchesNameList[[branchCounter]])
        branchName <- branchesName[branchCounter]
        ## get the mutational signature of the branch
        if (length(mutSigRef[which(
            mutSigRef$Sample == branchName), 1]) < mutThreshold){
            sigsMaxName <- "No.Signature"
            sigsMaxProb <- 0
            message(paste("Branch ", branchName,
                          ": Number of mutations is less than mutThreshold.",
                          sep = ""))
            # next()
        }else{
            ## deconstructSigs
            sigsInput <- suppressWarnings(
                mut.to.sigs.input(mut.ref=mutSigRef, 
                                  sample.id="Sample", 
                                  chr="chr", 
                                  pos="pos", 
                                  ref="ref", 
                                  alt="alt",
                                  bsg=get(refBuild)))
            sigsWhich <- whichSignatures(tumor.ref=sigsInput, 
                                         signatures.ref=get(signaturesRef), 
                                         sample.id=branchName,
                                         contexts.needed=TRUE)
            
            ## get mutational signature with max weight
            sigsMax <- sigsWhich[["weights"]][which.max(sigsWhich[["weights"]])]
            sigsMaxName <- colnames(sigsMax)
            sigsMaxProb <- sigsMax[,1]
        }
        
        ## vectorize branch name
        # branch <- gsub(paste(patientID,"-",sep=""), "", branch)
        
        ## figure out putative driver genes
        if (!is.null(driverGenesFile)){
            ## read putative driver genes' list
            driverGenes <- as.character(read.table(
                driverGenesFile, header = TRUE, quote="", sep="\n")[,1])
            ## filter potative driver genes of each branch
            pdgMut <- mutSigRef[which(
                mutSigRef$Sample == branchName &
                    as.character(mutSigRef$Hugo_Symbol) %in% driverGenes),]
            pdgBranch <- as.character(pdgMut$Hugo_Symbol)
            ## collect branches' mutataional signature and potative driver genes
            mutSigsBranch <- data.frame(
                branch=c(branchName), 
                sig=sigsMaxName, 
                mut.num=length(
                    mutSigRef[which(
                        mutSigRef$Sample == branchName), 1]), 
                sig.prob=sigsMaxProb, 
                putative_driver_genes=c(paste(pdgBranch, collapse = ",")))
        } else{
            mutSigsBranch <- data.frame(
                branch=c(branchName), 
                sig=sigsMaxName, 
                mut.num=length(mutSigRef[which(
                    mutSigRef$Sample == branchName), 1]), 
                sig.prob=sigsMaxProb)
        }
        ## collect branches' mutataional signature information
        mutSigsOutput <- rbind(mutSigsOutput, mutSigsBranch)
    }
    if (plot.Signatures) {
        pic <- .plotMutationalSig(sigsInput, mutSigsOutput)
        print(pic)
        message(paste(njtree@patientID, " mutaional signature plot generation done!", sep=""))
    }
    message(paste(njtree@patientID, " mutaional signature generation done!", sep=""))
    return(mutSigsOutput)
}

## plot.MutationalSigs
.plotMutationalSig <- function(sigsInput, mutSigsOutput) {
    ## calculate the Mutation Probability
    sigsInputSum <- as.data.frame(apply(sigsInput, 1, function(x) sum(x)))
    sigsInputTrans <- as.data.frame(t(sigsInput))
    ls.branchesName <- as.character(rownames(sigsInput))
    ls.mutationType <-as.character(rownames(sigsInputTrans))
    ls.mutationGroup <- c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G")
    sigsInputBranches <- data.frame()
    
    ## calculation process(maybe could be replaced by lapply)
    for (branch in ls.branchesName) {
        sigsInputTrans[branch] <- sigsInputTrans[branch]/sigsInputSum[branch, ]
        signature <- as.character(mutSigsOutput[which(mutSigsOutput$branch == branch), ]$sig)
        sigsInputBranch <- data.frame(sigsInputTrans[branch], 
                                      rep(branch, length(sigsInputTrans[branch])), 
                                      rep(signature, length(sigsInputTrans[branch])), 
                                      stringsAsFactors = FALSE)
        colnames(sigsInputBranch) <- c("Mutation_Probability", "Branch", "Signature")
        sigsInputBranches <- rbind(sigsInputBranches, sigsInputBranch)
    }
    
    df.sigsInputTrans <- data.frame(Mutational_Type=ls.mutationType, 
                                    Group=ls.mutationType, 
                                    sigsInputBranches, 
                                    stringsAsFactors = FALSE)
    
    ## generate Mutation Type for every column
    for (mutationGroup in ls.mutationGroup) {
        df.sigsInputTrans$Group[which(grepl(mutationGroup, df.sigsInputTrans$Group))] <- mutationGroup
    }
    
    ## specific the label order of x axis
    orderlist <- c(ls.mutationType)
    df.sigsInputTrans <- transform(df.sigsInputTrans, Mutational_Type = factor(Mutational_Type, levels = orderlist))
    
    ## 
    # group.colors <- c("#009AEC", "#000000", "#C10000", "#A5A5A5", "#00C491", "#FF4FB1")
    group.colors <- pal_npg("nrc", alpha=1)(6)
    pic <- ggplot(df.sigsInputTrans, aes(x=Mutational_Type, y=Mutation_Probability, group=Group, fill=Group)) + 
        geom_bar(stat="identity") + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
        geom_rect(aes(xmin=0, xmax=16.5, ymin=0, ymax=Inf),
                  fill="#fce7e4", alpha=0.15) + 
        geom_rect(aes(xmin=16.5, xmax=32.5, ymin=0, ymax=Inf),
                  fill="#ecf8fa", alpha=0.25) + 
        geom_rect(aes(xmin=32.5, xmax=48.5, ymin=0, ymax=Inf),
                  fill="#dbfff9", alpha=0.05) + 
        geom_rect(aes(xmin=48.5, xmax=64.5, ymin=0, ymax=Inf),
                  fill="#e4e8f3", alpha=0.08) + 
        geom_rect(aes(xmin=64.5, xmax=80.5, ymin=0, ymax=Inf),
                  fill="#fdefeb", alpha=0.15) + 
        geom_rect(aes(xmin=80.5, xmax=96.5, ymin=0, ymax=Inf),
                  fill="#e5e8ef", alpha=0.1) + 
        geom_bar(stat="identity") + 
        facet_grid(Branch ~ .) + 
        scale_fill_manual(values=group.colors) + 
        ylim(0, 0.2) + 
        geom_text(data = dplyr::distinct(df.sigsInputTrans, Branch, .keep_all = TRUE), 
                  aes(y=0.175, label=Signature), 
                  position=position_nudge(x=7.5), 
                  colour="#2B2B2B", , fontface = "bold", size=2.5)
    return(pic)
}

## Branches' mutation collection
.treeMutationalBranches <- function(maf, branch, mut_sort.id){
    ## get mutationalSigs-related  infomation
    maf_input <- maf@data
    patientID <- maf@patientID
    datChr <- data.frame(chr=as.character(maf_input$Chromosome), stringsAsFactors=FALSE)
    datChr$chr <- paste("chr", datChr$chr, sep="")
    datMutgene <-  maf_input$Hugo_Symbol
    mutId <- dplyr::select(tidyr::unite(maf_input, "mut.id", 
                                        Hugo_Symbol, Chromosome, 
                                        Start_Position, 
                                        Reference_Allele, Tumor_Seq_Allele2, 
                                        sep=":"), mut.id)
    mutSigRef <- data.frame(as.character(maf_input$Tumor_Sample_Barcode), 
                            datChr, maf_input$Start_Position, 
                            maf_input$End_Position, maf_input$Reference_Allele, 
                            maf_input$Tumor_Seq_Allele2, datMutgene, 
                            mutId, stringsAsFactors=FALSE)
    colnames(mutSigRef) <- c("Sample", 
                             "chr", "pos", 
                             "pos_end", "ref", 
                             "alt", "Hugo_Symbol", 
                             "mut_id")
    
    ## get branch infomation
    ls.branch <- branch[order(nchar(branch), branch)]
    branches <- strsplit(ls.branch, split='âˆ©')
    
    ## generate mutational intersections for each branch
    mutBranchesOutput <- list()
    for (branch in branches){
        
        ## generate intersection's mut.id and get the mutation information in mutSigRef
        branch <- unlist(branch)
        branch.id <- append(branch, "mut.id")
        unbranch <- names(mut_sort.id)[which(!(names(mut_sort.id) %in% branch.id))]
        branch.intersection <- intersect(
            mut_sort.id %>% dplyr::filter_at(branch, all_vars(. == 1)), 
            mut_sort.id %>% dplyr::filter_at(unbranch, all_vars(. == 0))) 
        branch.mut.id <- branch.intersection$mut.id 
        
        ## generate the branch name
        branchName <- paste(branch, collapse="âˆ©")
        ## data duplication
        branch.mut <- mutSigRef[which(mutSigRef$mut_id %in% branch.mut.id), ]
        branch.mut$Sample <- branchName
        branch.mut <- branch.mut[!duplicated(branch.mut),]
        ## generate branch mutation list
        mutBranchesOutput[[branchName]] <- branch.mut
    }
    return(mutBranchesOutput)
}
