#' mutTrunkBranch
#' @description Summarize and conduct paired Wilcoxon test of mutations of trunk/branches in a phylogenetic tree.
#' 
#' @param phyloTree a phyloTree object generated by phyloTree function.
#' @param geneList list of genes to restrict the analysis. Default NULL.
#' @param signaturesRef The parameter used for deconstructSig. Default "cosmic_v2". Option: "genome_cosmic_v3","exome_cosmic_v3","nature2013". 
#' @param plot output a list of diagrams that visualize mutational signature of trunk/branches in a phylogenetic tree.Default FALSE. 
#' @param conf.level confidence level of the interval for wilcox.test. Default: 0.95. Option: on the scale of 0 to 1.
#' 
#' @examples
#' mutTrunkBranch(phyloTree)
#' mutTrunkBranch(phyloTree,plot = T)
#' 
#' @return  a list of box plots based on mutational categories
#' @export mutTrunkBranch

mutTrunkBranch <- function(phyloTree,
                           geneList=NULL,
                           pvalue = 0.05,
                           plot = TRUE){
    treeMSOutput <- lapply(phyloTree, doTreeMutSig,
                           geneList = geneList,
                           MTB = TRUE)
    mutTrunkBranch.list <- lapply(treeMSOutput, doMutTrunkBranch)
    mutTrunkBranch.list <- mutTrunkBranch.list[!is.na(mutTrunkBranch.list)]
    if(length(mutTrunkBranch.list) == 0){
        return(NA)
    }
    if(plot){
        # TB.plot <- lapply(treeMSOutput, doPlotTrunkBranch)
        TB.plot <- doPlotTrunkBranch(mutTrunkBranch.list,pvalue = pvalue)
        return(list(mutTrunkBranch.res = mutTrunkBranch.list,
               mutTrunkBranch.plot = TB.plot))
    }
    return(mutTrunkBranch.list = mutTrunkBranch.list)
}