#' mutTrunkBranch
#' @description Summarize and conduct paired Wilcoxon test of mutations of trunk/branches in a phylogenetic tree.
#' 
#' @param phyloTree_list a list of phyloTree generated by phyloTree function.
#' @param CT Distinction between C>T at CpG and C>T at other sites, Default FALSE
#' @param geneList list of genes to restrict the analysis. Default NULL.
#' @param signaturesRef The parameter used for deconstructSig. Default "cosmic_v2". Option: "genome_cosmic_v3","exome_cosmic_v3","nature2013". 
#' @param plot output a list of diagrams that visualize mutational signature of trunk/branches in a phylogenetic tree.Default FALSE. 
#' @param pvalue confidence level of the interval for Fisher test. Default: 0.05.
#' 
#' @examples
#' mutTrunkBranch(phyloTree_list)
#' mutTrunkBranch(phyloTree_list,plot = T)
#' 
#' @return  a list of box plots based on mutational categories
#' @export mutTrunkBranch

mutTrunkBranch <- function(phyloTree_list,
                           CT = FALSE,
                           geneList=NULL,
                           pvalue = 0.05,
                           plot = TRUE,
                           patient.id = NULL){
    
    if(!is.null(patient.id)){
        patient.setdiff <- setdiff(patient.id, names(phyloTree_list))
        if(length(patient.setdiff) > 0){
            stop(paste0(patient.setdiff, " can not be found in your data"))
        }
        phyloTree_list <- phyloTree_list[names(phyloTree_list) %in% patient.id] 
    }
    
    mtb_input_list <- lapply(phyloTree_list,function(x, CT, geneList){
        mutSigRef <- x@mut.branches
        refBuild <- x@refBuild
        patientID <- x@patientID
        refBuild <- paste("BSgenome.Hsapiens.UCSC.", x@refBuild, sep = "")
        if(!is.null(geneList)){
            mutSigRef <- mutSigRef %>% 
                dplyr::filter(Hugo_Symbol %in% geneList)
        }
        sigsInput <- countTriplet(mutSigRef = mutSigRef,
                                  withinTumor = FALSE,
                                  refBuild = refBuild,
                                  patientID = patientID,
                                  CT = CT)
        branch_ids <- unique(mutSigRef$Branch_ID)
        branch_sample_num <- lapply(branch_ids,function(x){
            s <- strsplit(x,"âˆ©")[[1]]
            num <- length(s)
        })
        trunkName <- branch_ids[which.max(branch_sample_num)]
        return(list(
            sigsInput = sigsInput, 
            trunkName = trunkName,
            patientID = patientID)
        )}, CT = CT, geneList = geneList)
    
    mtb_output_list <- lapply(mtb_input_list, doMutTrunkBranch,CT = CT)
    mtb_output_list <- mtb_output_list[!is.na(mtb_output_list)]
    if(length(mtb_output_list) == 0){
        return(NA)
    }
    if(plot){
        plot.list <- lapply(mtb_output_list, doPlotTrunkBranch, pvalue = pvalue, CT = CT)
        return(list(mutTrunkBranch.res = mtb_output_list,
               mutTrunkBranch.plot = plot.list))
    }
    return(mutTrunkBranch.list = mtb_output_list)
}