#' mutTrunkBranch
#' @description Summarize and conduct paired Wilcoxon test of mutations of trunk/branches in a phylogenetic tree.
#' 
#' @param phyloTree object generated by phyloTree function.
#' @param CT Distinction between C>T at CpG and C>T at other sites, Default FALSE
#' @param geneList list of genes to restrict the analysis. Default NULL.
#' @param signaturesRef The parameter used for deconstructSig. Default "cosmic_v2". Option: "genome_cosmic_v3","exome_cosmic_v3","nature2013". 
#' @param plot output a list of diagrams that visualize mutational signature of trunk/branches in a phylogenetic tree.Default FALSE. 
#' @param pvalue confidence level of the interval for Fisher test. Default: 0.05.
#' 
#' @examples
#' mutTrunkBranch(phyloTree)
#' mutTrunkBranch(phyloTree,plot = T)
#' 
#' @return  a list of box plots based on mutational categories
#' @export mutTrunkBranch

mutTrunkBranch <- function(phyloTree,
                           CT = FALSE,
                           geneList=NULL,
                           pvalue = 0.05,
                           plot = TRUE,
                           patient.id = NULL){
    
    if(class(phyloTree) == "phyloTree"){
        phyloTree_list <- list(phyloTree)
    }else if(class(phyloTree) == "phyloTree_list"){
        if(!is.null(patient.id)){
            phyloTree_list <- subset_phyloTree_list(phyloTree,patient.id = patient.id)
        }else{
            phyloTree_list <- phyloTree@patient.list
        }
    }else{
        stop("phyloTree should be class 'phyloTree' or class 'phyloTree_list' generated by function getPhyloTree")
    }
    
    result <- list()
    for(phyloTree in phyloTree_list){
        mutSigRef <- phyloTree@mut.branches
        refBuild <- phyloTree@refBuild
        patient <- phyloTree@patientID
        refBuild <- paste("BSgenome.Hsapiens.UCSC.", phyloTree@refBuild, sep = "")
        if(!is.null(geneList)){
            mutSigRef <- mutSigRef %>% 
                dplyr::filter(Hugo_Symbol %in% geneList)
        }
        sigsInput <- countTriplet(mutSigRef = mutSigRef,
                                  withinTumor = FALSE,
                                  refBuild = refBuild,
                                  patientID = patientID,
                                  CT = CT)
        branch_ids <- unique(mutSigRef$Branch_ID)
        branch_sample_num <- lapply(branch_ids,function(x){
            s <- strsplit(x,"âˆ©")[[1]]
            num <- length(s)
        })
        trunkName <- branch_ids[which.max(branch_sample_num)]
        mtb_input <- list(sigsInput = sigsInput, 
                           trunkName = trunkName,
                          patientID = patient)
        mtb_output <- doMutTrunkBranch(mtb_input = mtb_input, CT = CT)
        if(plot){
            p <- doPlotTrunkBranch(mtb_output = mtb_output,  pvalue = pvalue, CT = CT)
            result[[patient]] <- list(mutTrunkBranch.res = mtb_output,mutTrunkBranch.plot = p)
        }else{
            result[[patient]] <- list(mutTrunkBranch.res = mtb_output)
        }
    }
    
    if(length(result) == 1){
        return(result[[1]])
    }else if(length(result) == 0){
        return(NA)
    }else{
        return(result)
    }
    return(result)
}