#' Class NJtree
#' 
#' @param a Maf object generated by readMaf function
#' @param minVaf the minimum value of vaf
#' @param maxVaf the maximum value of vaf
#' @param ccf.mutation.id manually specify which columns could be joint by ccf.mutation.sep to get the same format of mutation id in ccfy
#' @param ccf.mutation.sep manually specify the separator character.Values on each line of the ccf.mutation.id are separated by this character. I
#' 
#' 
#' @examples
#' maf.File <- system.file("extdata/maf", "311252.maf", package = "Meskit")
#' sampleInfo.File <- system.file("extdata", "sample_info.txt", package = "Meskit")
#' pyCloneCluster <- system.file("extdata/ccf", "311252.cluster.tsv", package = "Meskit")
#' pyCloneLoci <- system.file("extdata/ccf", "311252.loci.tsv", package = "Meskit")
#' maf <- readMaf(mafFile=maf.File, sampleInfoFile=sampleInfoFile, refBuild="hg19")
#' njtree <- getNJtree(maf)
#' @return  an object of class NJtree
#' @exportClass NJtree
#' @export getNJtree


# set NJtree object
getNJtree <- function(maf, 
                      method = "NJ",
                      minVaf=0.02, 
                      maxVaf=1, 
                      ccf.mutation.id = c("Hugo_Symbol","Chromosome","Start_Position"),
                      ccf.mutation.sep = ":"){
  
  maf.dat <- maf@data

  if (max(maf.dat$VAF, na.rm=TRUE) > 1){
    maf.dat$VAF <- maf.dat$VAF/100
  }
  maf.dat <- maf.dat[which(maf.dat$VAF > minVaf & maf.dat$VAF < maxVaf), ]
  ## information input
  patientID <- maf@patientID
  refBuild <- paste("BSgenome.Hsapiens.UCSC.", maf@ref.build, sep = "")
  mut_sort.id <- mut_binary_sort(maf.dat = maf.dat)
  mut_sort <- as.matrix(mut_sort.id[, -1])
  mut_dat <- t(mut_sort)
  mat.nj <- nj(dist.gene(mut_dat))
  if(method == "MP"){
      tree_dat <- as.phyDat(mut_dat, type="USER", levels = c(0, 1))
      tree_pars <- optim.parsimony(mat.nj, tree_dat)
      mat.nj <- acctran(tree_pars, tree_dat)
  }
  if(method == "ML"){
      tree_dat <- as.phyDat(mut_dat, type="USER", levels = c(0, 1))
      fitJC <- pml(mat.nj, tree_dat)
      fitJC <- optim.pml(fitJC)
      mat.nj <- fitJC$tree
  }
  branchAlias <- read.njtree(mat.nj)
  mut_branches <- .treeMutationalBranches(maf.dat, branchAlias, mut_sort.id)
  ccf_sort <- matrix()
  if(length(maf@ccf.loci)!= 0){
    ccf_sort <- mut_ccf_sort(maf.dat = maf.dat, ccf = maf@ccf.loci, ccf.mutation.id, ccf.mutation.sep)
  }
  
  njtree <- new('NJtree', patientID = patientID, nj = mat.nj, mut_sort = mut_sort, mut_branches = mut_branches, ccf_sort = ccf_sort, refBuild = maf@ref.build)
  return(njtree)
}
#Prevent class 'phylo' from not existing
setClass('phylo')
#Class NJtree
setClass('NJtree', slots = c(nj = 'phylo', patientID = 'character', mut_sort = 'matrix', mut_branches = 'list', ccf_sort = 'matrix', refBuild = 'character'))
#extract mut_sort from NJtree object
setGeneric("getMutSort", function(x, signature){standardGeneric("getMutSort")})
setMethod("getMutSort", 'NJtree', function(x){x@mut_sort})
#extract nj from NJtree object
setGeneric("getPhyloTree", function(x, signature){standardGeneric("getPhyloTree")})
setMethod("getPhyloTree",'NJtree', function(x){x@nj})
