#' @title plotCNA
#'  
#' @param seg Object generated by readSegment function.
#' @param patient.id Select the specific patients. Default: NULL, all patients are included.
#' @param chrSilent Chromosomes excluded in the analysis. e.g, 1, 2, 3 . Default NULL.
#' @param refBuild Human reference genome versions of hg18, hg19 or hg38 by UCSC. Default: "hg19".
#' @param show.GISTIC.gene Whether GISTIC gene in seg.Default FALSE.
#' @param sample.text.size Size of sample name.Default 11.
#' @param legend.text.size Size of legend text. Default 9.
#' @param legend.title.size Size of legend title.Default 11.
#' @param sample.bar.height Bar height of each sample .Default 0.5.
#' @param chrom.bar.height Bar height of each chromosome .Default 0.5.
#' 
#' 
#' @examples
#' segCN.file <- system.file("extdata", "HCC6046.seg.txt", package = "MesKit")
#' seg <- readSegment(segCN.file = segCN.file)
#' plotCNA(seg)
#' @import cowplot
#' @export plotCNA
#'


plotCNA <- function(seg,
                    patient.id = NULL,
                    chrSilent = NULL,
                    refBuild = "hg19",
                    show.GISTIC.gene = FALSE,
                    sample.text.size = 11,
                    chrom.text.size = 3,
                    legend.text.size = 9,
                    legend.title.size = 11,
                    sample.bar.height = 0.5,
                    chrom.bar.height = 0.5
){
    
    if(!is.null(chrSilent)){
        chr.setdiff <- setdiff(chrSilent, unique(seg$Chromosome))
        if(length(chr.setdiff) > 0){
            stop(paste0("Chromosome ", chr.setdiff, " can not be found in your data"))
        }
        seg <- seg[Chromosome %in% chrSilent]
    }
    
    ## select patients in patient.id 
    if(is.null(patient.id)|length(patient.id) > 1){
        if(length(patient.id) > 1){
            patient.setdiff <- setdiff(patient.id, unique(seg$Patient_ID))
            if(length(patient.setdiff) > 0){
                stop(paste0("Error: ", patient.setdiff, " can not be found in your data"))
            }
            seg <- seg[Patient_ID %in% patient.id]
        }
        seg$patient <- seg$Patient_ID
        seg <- dplyr::select(seg, -Patient_ID)
        seg$Patient_ID <- "ALL"
    }else if(length(patient.id) == 1){
        patient.setdiff <- setdiff(patient.id, unique(seg$Patient_ID))
        if(length(patient.setdiff) > 0){
            stop(paste0("Error: ", patient.setdiff, " can not be found in your data"))
        }
        seg <- seg[Patient_ID %in% patient.id]
    }
    
    if(show.GISTIC.gene){
      if(!"Gistic.type" %in% colnames(seg)){
        stop("Error: GISTIC genes information were not found. Please check readSegment")
      }
    }
    # if("Start" %in% colnames(seg)){
    #   seg <- dplyr::rename(seg,Start_Position = Start, End_Position = End)
    # }
    ref.options <- c("hg19","hg18","hg38")
    if(!refBuild %in% ref.options){
        stop("Error: refBuild should only be either 'hg18','hg19' or 'hg38'")
    }
    if(refBuild == 'hg19'){
        chrLens = c(249250621, 243199373, 198022430, 191154276, 180915260, 171115067, 159138663,
                     146364022, 141213431, 135534747, 135006516, 133851895, 115169878, 107349540,
                     102531392, 90354753, 81195210, 78077248, 59128983, 63025520, 48129895, 51304566,
                     155270560, 59373566)
    } else if(refBuild == 'hg18'){
        chrLens = c(247249719, 242951149, 199501827, 191273063, 180857866, 170899992,
                     158821424, 146274826, 140273252, 135374737, 134452384, 132349534,
                     114142980, 106368585, 100338915, 88827254, 78774742, 76117153,
                     63811651, 62435964, 46944323, 49691432, 154913754, 57772954)
    } else if(refBuild == 'hg38'){ #hg38
        chrLens = c(248956422, 242193529, 198295559, 190214555, 181538259, 170805979,
                     159345973, 145138636, 138394717, 133797422, 135086622, 133275309,
                     114364328, 107043718, 101991189, 90338345, 83257441, 80373285,
                     58617616, 64444167, 46709983, 50818468, 156040895, 57227415)
    } else{
        stop('"Error: available reference builds: hg18, hg19, hg38')
    }
    updatePosition <- function(x,pos,chrname,chrLens){
        return(as.numeric(x[pos]) + as.numeric(chrLens[as.numeric(x[chrname]) ]) )
    }
    chrLens =  append(cumsum(chrLens),1,after = 0)
    chrLabels = c(1:22)
    chrTable = data.table::data.table(chr = chrLabels, start = chrLens[1:(length(chrLens)-3)], end = chrLens[2:(length(chrLens)-2)])
    chrTable$color = rep(c('black','white'), length = nrow(chrTable))
    patients <- as.character(unique(seg$Patient_ID))
    dat.list <- split(seg, seg$Patient_ID)
    CNAplot.list <- list()
    
    for(patient in patients){
        patient.seg <- dat.list[[patient]]
        updateStart <- apply(patient.seg,1,updatePosition,"Start_Position","Chromosome",chrLens)
        updateEnd <- apply(patient.seg,1,updatePosition,"End_Position","Chromosome",chrLens)
        suppressWarnings(patient.seg[,Update_Start:= updateStart]) 
        suppressWarnings(patient.seg[,Update_End:= updateEnd])
        patient.seg[,hmin := 0]
        patient.seg[,hmax := 0]
        h <- chrom.bar.height
        
        if("patient" %in% colnames(patient.seg)& length(unique(patient.seg$patient)) > 1){
            
            ## sort sampleid 
            patient.seg <- patient.seg %>% 
                dplyr::arrange(patient,
                               plyr::desc(Tumor_Sample_Barcode),
                               Chromosome,
                               Start_Position) %>%
                as.data.table()
            
            patient.seg[,Sample_ID := paste(patient,Tumor_Sample_Barcode,sep = "&")]
            sampleids <- unique(patient.seg$Sample_ID)
            patient.rect.hmin <- c(h)
            patient.rect.hmax <- c()
            patient1 <- unique(patient.seg[Sample_ID == sampleids[1],]$patient)
            for(sampleid in sampleids){
                patient2 <- unique(patient.seg[Sample_ID == sampleid,]$patient)
                if(patient1 != patient2){
                    patient.rect.hmin <- append(patient.rect.hmin,h+ sample.bar.height*2/5)
                    patient.rect.hmax <- append(patient.rect.hmax,h)
                    h <- h + sample.bar.height*2/5
                }
                patient1 <- patient2
                patient.seg[Sample_ID == sampleid,]$hmin <- h
                patient.seg[Sample_ID == sampleid,]$hmax <- h + sample.bar.height
                h <- h + sample.bar.height + sample.bar.height/10
            }
            patient.rect.hmax <- append(patient.rect.hmax,max(patient.seg$hmax))
            patient.rect.table <- data.frame(hmin = patient.rect.hmin,
                                             hmax = patient.rect.hmax,
                                             patient = as.character(unique(patient.seg$patient)))
            patient.seg$patient <- factor(patient.seg$patient,levels = as.character(unique(patient.seg$patient)))
            
        }else{
            ## sort sampleid 
            patient.seg <- patient.seg %>% 
                dplyr::arrange(Patient_ID,
                               Tumor_Sample_Barcode,
                               Chromosome,
                               Start_Position) %>%
                as.data.table()
            patient.seg[,Sample_ID := paste(Patient_ID,Tumor_Sample_Barcode,sep = "&")]
            sampleids <- sort(unique(patient.seg$Sample_ID))
            for(sampleid in sampleids){
                patient.seg[Sample_ID == sampleid,]$hmin <- h
                patient.seg[Sample_ID == sampleid,]$hmax <- h + sample.bar.height
                h <- h + sample.bar.height + sample.bar.height/10
            }
        }
        
        min <- sort(unique(patient.seg$hmin))
        max <- sort(unique(patient.seg$hmax))
        CNADat <- patient.seg[Type != "Neutral",]
        patient.type <- unique(CNADat$Type)
        
        ## set color for CNA type
        type.all.levels <- c("Loss", "Deletion",  "Gain",  "Amplification")
        type.all.colors <- c("#6baed6", "#084594", "#f4a582", "#d73027")
        type.level <- type.all.levels[type.all.levels %in% patient.type]
        type.color <- type.all.colors[type.all.levels %in% patient.type]
        
        CNADat$Type <- factor(CNADat$Type, levels = type.level)
        segmentTable <- data.table::data.table(y = c(min(min),max(max)), yend = c(min(min),max(max)))
        seg.add <- 20000000
        text.add <- -200000000
        segmentTable$x <- -seg.add
        segmentTable$xend <- max(chrTable$end)+seg.add
        segmentTable <- rbind(segmentTable, list(min(min), max(max), -seg.add,-seg.add))
        segmentTable <- rbind(segmentTable, list(min(min), max(max), max(chrTable$end)+seg.add,max(chrTable$end)+seg.add))
        
       ## get tumor sample barcode of patient
        patient.tsbs <- unlist(lapply(unique(patient.seg$Sample_ID),
                                      function(x){return(strsplit(x,"&")[[1]][2])}))
        
        ## label on axis Y
        Y.text.table <- data.table::data.table(Pos = min + (max-min)/2, Tumor_Sample_Barcode = patient.tsbs)
        
        ## set background for the graph
        backgroundTable <- data.table::data.table(
            xmin = min(chrTable$start), 
            xmax = max(chrTable$end), 
            ymin = min, 
            ymax = max)
        
        ## set color for patient bar
        allScales <- type.color
        names(allScales) <- type.level
        if("patient" %in% colnames(patient.seg)&
           length(unique(patient.seg$patient)) > 1){
            set.seed(1234)
            patient_colors <- sample(colors(),length(patient.rect.table$patient),replace = FALSE)
            names(patient_colors) <- patient.rect.table$patient
        }
        p <- ggplot()+
            ## background color
            geom_rect(data = backgroundTable, 
                      mapping = aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), 
                      fill = "#f0f0f0")+
            
            ## CNA regions
            geom_rect(data = CNADat,
                      mapping = aes(xmin = Update_Start, xmax = Update_End, ymin = hmin, ymax = hmax, fill = Type))+
            scale_fill_manual(values = type.color, name = "Type")+
            
            ## chr bar
            geom_rect(data = chrTable,
                mapping = aes(xmin = start, xmax = end, ymin = 0, ymax = chrom.bar.height),
                fill = rep(c("black","white"), 11), color = "black")+
            geom_text(data = chrTable,
                mapping = aes(x = start + (end - start)/2, y = chrom.bar.height/2, label = chr), 
                size = chrom.text.size,
                color = rep(c("white","black"), 11))+
            
            ## vertical line 
            geom_segment(aes(x = chrTable$end[1:(nrow(chrTable)-1)],
                             xend = chrTable$end[1:(nrow(chrTable)-1)],
                             y = chrom.bar.height,
                             yend = max(backgroundTable$ymax)),
                         linetype = "dotted",
                         color = "#252525",
                         size = 0.7) +
            
            theme(
                axis.text.x = element_blank(),
                axis.text.y.left =  element_text(size = sample.text.size,
                                           margin = margin(l = 0),
                                           colour = "black"),
                axis.ticks = element_blank(),
                # axis.line.y = element_line(colour = "darkblue", size = 1, linetype = "solid"),
                panel.grid =element_blank(),
                panel.border = element_blank(),
                panel.background = element_blank(),
                axis.title.x = element_blank(),
                axis.title.y = element_blank(),
                # legend.key.width  = unit(1.1,'lines'),
                # legend.box.margin = margin(l = 0),
                # legend.title = element_text(size = legend.title.size),
                # legend.text = element_text(size = legend.text.size,margin = margin(b = 3))
                )+
            # geom_text(
            #     data = Y.text.table,
            #     mapping = aes(x = text.add, y = Pos, label = Tumor_Sample_Barcode))+
            scale_x_continuous(expand = c(0,0))+
            scale_y_continuous(breaks = Y.text.table$Pos,
                               labels = Y.text.table$Tumor_Sample_Barcode)+
            ggtitle("Copy number variant profile")+
            theme(plot.title = element_text(size = 13.5, face = "bold", hjust = 0.5, vjust = -2))
        ## patient bar
        if("patient" %in% colnames(patient.seg)&
           length(unique(patient.seg$patient)) > 1){
            p <- p + 
                geom_rect(data = patient.rect.table,
                          mapping = aes(xmax = -20000000, xmin = -40000000,ymin = hmin, ymax = hmax),
                          fill =  patient_colors)
            # print(CNADat)
            ## multi legend
            patient_legend <- (
                    ggplot()+
                    geom_rect(data = patient.rect.table,
                              mapping = aes(xmax = -20000000, xmin = -40000000,
                                            ymin = hmin, ymax = hmax,fill = patient)) +  
                    scale_fill_manual(values = patient_colors,name = "Patient") + 
                    theme(legend.background = element_blank(),
                          legend.title = element_text(size = legend.title.size),
                          legend.text = element_text(size = legend.text.size,margin = margin(b = 3)))
            )%>%
                ggplotGrob %>%
                {.$grobs[[which(sapply(.$grobs, function(x) {x$name}) == "guide-box")]]}
            
            type_legend <- (
                    ggplot()+
                   geom_rect(data = CNADat,mapping = aes(xmin = Update_Start, xmax = Update_End, ymin = hmin, ymax = hmax, fill = Type))+
                   scale_fill_manual(values = type.color, name = "Type")+
                   theme(legend.background = element_blank(),
                         legend.title = element_text(size = legend.title.size),
                         legend.text = element_text(size = legend.text.size,margin = margin(b = 3)))
            ) %>% 
                ggplotGrob %>%
                {.$grobs[[which(sapply(.$grobs, function(x) {x$name}) == "guide-box")]]}
            
            legends_column <-
                plot_grid(
                    patient_legend + theme(legend.position = "none"),
                    type_legend,
                    patient_legend + theme(legend.position = "none"),
                    patient_legend,
                    patient_legend + theme(legend.position = "none"),
                    ncol = 1,
                    rel_heights = c(1,1,.1,1,1),
                    align = "v")
            
            p <- plot_grid(p + theme(legend.position = "none"),
                           legends_column + theme(plot.margin = margin(r = 15)),
                           nrow = 1,
                           rel_widths = c(1,.2))
            
        }
        
        CNAplot.list[[patient]] <- p
    }
    
    if(length(CNAplot.list) == 1){
        return(CNAplot.list[[1]])
    }
    
    return(CNAplot.list)
}





