#' @title plotCNA
#'  
#' @param seg object generated by readSegment function.
#' @param refBuild human reference genome versions of hg18, hg19 or hg38 by UCSC. Default "hg19".
#' @param show.GISTIC.gene Whether GISTIC gene in seg.Default FALSE.
#' @param patient.id select the specific patients. Default: NULL, all patients are included.
#' 
#' @examples
#' segCN.file <- system.file("extdata", "HCC6046.seg.txt", package = "MesKit")
#' seg <- readSegment(segCN.file = segCN.file)
#' plotCNA(seg)
#'
#' @export plotCNA
#'


plotCNA <- function(seg, refBuild = "hg19", show.GISTIC.gene = FALSE, patient.id = NULL){
    if(is.null(patient.id)){
        seg$patient <- seg$Patient_ID
        seg <- dplyr::select(seg, -Patient_ID)
        seg$Patient_ID <- "ALL"
    }else{
        patient.setdiff <- setdiff(patient.id, unique(seg$Patient_ID))
        if(length(patient.setdiff) > 0){
            stop(paste0(patient.setdiff, " can not be found in your data"))
        }
        seg <- seg[Patient_ID %in% patient.id]
    }
    if(show.GISTIC.gene){
      if(!"Gistic.type" %in% colnames(seg)){
        stop("GISTIC genes information were not found. Please check readSegment")
      }
    }
    # if("Start" %in% colnames(seg)){
    #   seg <- dplyr::rename(seg,Start_Position = Start, End_Position = End)
    # }
    ref.options <- c("hg19","hg18","hg38")
    if(!refBuild %in% ref.options){
        stop("refBuild can only be either 'hg18','hg19' or 'hg38'")
    }
    if(refBuild == 'hg19'){
        chrLens = c(249250621, 243199373, 198022430, 191154276, 180915260, 171115067, 159138663,
                     146364022, 141213431, 135534747, 135006516, 133851895, 115169878, 107349540,
                     102531392, 90354753, 81195210, 78077248, 59128983, 63025520, 48129895, 51304566,
                     155270560, 59373566)
    } else if(refBuild == 'hg18'){
        chrLens = c(247249719, 242951149, 199501827, 191273063, 180857866, 170899992,
                     158821424, 146274826, 140273252, 135374737, 134452384, 132349534,
                     114142980, 106368585, 100338915, 88827254, 78774742, 76117153,
                     63811651, 62435964, 46944323, 49691432, 154913754, 57772954)
    } else if(refBuild == 'hg38'){ #hg38
        chrLens = c(248956422, 242193529, 198295559, 190214555, 181538259, 170805979,
                     159345973, 145138636, 138394717, 133797422, 135086622, 133275309,
                     114364328, 107043718, 101991189, 90338345, 83257441, 80373285,
                     58617616, 64444167, 46709983, 50818468, 156040895, 57227415)
    } else{
        stop('Available reference builds: hg18, hg19, hg38')
    }
    updatePosition <- function(x,pos,chrname,chrLens){
        return(as.numeric(x[pos]) + as.numeric(chrLens[as.numeric(x[chrname]) ]) )
    }
    chrLens =  append(cumsum(chrLens),1,after = 0)
    chrLabels = c(1:22)
    chrTable = data.table::data.table(chr = chrLabels, start = chrLens[1:(length(chrLens)-3)], end = chrLens[2:(length(chrLens)-2)])
    chrTable$color = rep(c('black','white'), length = nrow(chrTable))
    patients <- as.character(unique(seg$Patient_ID))
    dat.list <- split(seg, seg$Patient_ID)
    CNAplot.list <- list()
    for(patient in patients){
        patient.seg <- dat.list[[patient]]
        # patient.seg <- dat.list[[1]]
        updateStart <- apply(patient.seg,1,updatePosition,"Start_Position","Chromosome",chrLens)
        updateEnd <- apply(patient.seg,1,updatePosition,"End_Position","Chromosome",chrLens)
        suppressWarnings(patient.seg[,Update_Start:= updateStart]) 
        suppressWarnings(patient.seg[,Update_End:= updateEnd])
        patient.seg[,hmin := 0]
        patient.seg[,hmax := 0]
        samples <- sort(unique(patient.seg$Tumor_Sample_Barcode))  
        h <- 0.5
        if("patient" %in% colnames(patient.seg)&
           length(unique(patient.seg$patient)) > 1){
            patient.rect.hmin <- c(h)
            patient.rect.hmax <- c()
            patient1 <- unique(patient.seg[Tumor_Sample_Barcode == samples[1],]$patient)
            for(sample in samples){
                patient2 <- unique(patient.seg[Tumor_Sample_Barcode == sample,]$patient)
                if(patient1 != patient2){
                    patient.rect.hmin <- append(patient.rect.hmin,h+0.2)
                    patient.rect.hmax <- append(patient.rect.hmax,h)
                    h <- h + 0.2
                }
                patient1 <- patient2
                patient.seg[Tumor_Sample_Barcode == sample,]$hmin <- h
                patient.seg[Tumor_Sample_Barcode == sample,]$hmax <- h + 0.5
                h <- h + 0.55
            }
            patient.rect.hmax <- append(patient.rect.hmax,max(patient.seg$hmax))
            patient.rect.table <- data.frame(hmin = patient.rect.hmin,
                                             hmax = patient.rect.hmax,
                                             patient = as.character(unique(patient.seg$patient)))
            patient.seg$patient <- factor(patient.seg$patient,levels = as.character(unique(patient.seg$patient)))
            
        }
        else{
            for(sample in samples){
                patient.seg[Tumor_Sample_Barcode == sample,]$hmin <- h
                patient.seg[Tumor_Sample_Barcode == sample,]$hmax <- h + 0.5
                h <- h + 0.55
            }
        }
        min <- sort(unique(patient.seg$hmin))
        max <- sort(unique(patient.seg$hmax))
        CNADat <- patient.seg[Type != "Neutral",]
        patient.type <- unique(CNADat$Type)
        type.all.levels <- c("Loss", "Deletion",  "Gain",  "Amplification")
        type.all.colors <- c("#6baed6", "#084594", "#f4a582", "#d73027")
        type.level <- type.all.levels[type.all.levels %in% patient.type]
        type.color <- type.all.colors[type.all.levels %in% patient.type]
        CNADat$Type <- factor(CNADat$Type, levels = type.level)
        segmentTable <- data.table::data.table(y = c(min(min),max(max)), yend = c(min(min),max(max)))
        seg.add <- 20000000
        text.add <- -200000000
        segmentTable$x <- -seg.add
        segmentTable$xend <- max(chrTable$end)+seg.add
        segmentTable <- rbind(segmentTable, list(min(min), max(max), -seg.add,-seg.add))
        segmentTable <- rbind(segmentTable, list(min(min), max(max), max(chrTable$end)+seg.add,max(chrTable$end)+seg.add))
        textTable <- data.table::data.table(Pos = min + (max-min)/2, Tumor_Sample_Barcode = samples)
        backgroundTable <- data.table::data.table(
            xmin = min(chrTable$start), 
            xmax = max(chrTable$end), 
            ymin = min, 
            ymax = max)
        allScales <- type.color
        names(allScales) <- type.level
        if("patient" %in% colnames(patient.seg)&
           length(unique(patient.seg$patient)) > 1){
            patient.all.color <- c( "#E64B35B2","#4DBBD5B2","#00A087B2","#3C5488B2","#F39B7FB2",
                                    "#8491B4B2","#91D1C2B2","#DC0000B2","#7E6148B2","#91D1C2B2",
                                    "#1B9E77","#D95F02","#7570B3","#E7298A","#66A61E","#E6AB02","#A6761D","#666666")
            patient.color <- patient.all.color[1:length(patient.rect.table$patient)]
            names(patient.color) <- patient.rect.table$patient
            allScales <- append(allScales,patient.color)
        }
        p <- ggplot()+
            geom_rect(data = chrTable,
                mapping = aes(xmin = start, xmax = end, ymin = 0, ymax = 0.5),
                fill = rep(c("black","white"), 11), color = "black")+
            geom_text(data = chrTable,
                mapping = aes(x = start + (end - start)/2, y = 0.25, label = chr), 
                color = rep(c("white","black"), 11))+
            geom_rect(data = backgroundTable, 
                mapping = aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), 
                fill = "#f0f0f0")+
            theme(
                panel.grid =element_blank(), 
                # axis.text = element_blank(), 
                axis.text.x = element_blank(),
                axis.text.y = element_text(size = 10,margin = margin(r = -20),colour = "black"),
                axis.ticks = element_blank(),
                panel.border = element_blank(),
                panel.background = element_blank(),
                axis.title.x = element_blank(),
                axis.title.y = element_blank(),
                legend.key.width  = unit(1.1,'lines'),
                # legend.title = element_blank(),
                legend.box.margin = margin(l = -20),
                # legend.key.height = unit(.8,'lines'),
                # legend.key = element_rect(size = 1, color = NA, fill = NA),
                # legend.spacing = unit(1.0, 'lines'),
                # legend.spacing.y = unit(1, 'char'),
                legend.text = element_text(size = 10,margin = margin(b = 3)))+
            # guides(fill=guide_legend(
            #     keywidth=0.1,
            #     keyheight=0.1)
            # )+
            geom_rect(data = CNADat,
                mapping = aes(xmin = Update_Start, xmax = Update_End, ymin = hmin, ymax = hmax, fill = Type))+
            # scale_fill_manual(name = "Type",breaks = type.level, values = type.color)+
            scale_fill_manual(values = allScales)+
            # geom_text(
            #     data = textTable,
            #     mapping = aes(x = text.add, y = Pos, label = Tumor_Sample_Barcode))+
            scale_y_continuous(breaks = textTable$Pos,
                               labels = textTable$Tumor_Sample_Barcode)+
            # geom_segment(aes(x = x, y = y, xend = xend, yend = yend), 
            #     data = segmentTable, linetype="dashed")+
            ggtitle("Copy number variant profile")+
            theme(plot.title = element_text(size = 18, face = "bold", hjust = 0.5, vjust = -1))
        ## patient bar
        if("patient" %in% colnames(patient.seg)&
           length(unique(patient.seg$patient)) > 1){
            mainplot <- p + geom_rect(data = patient.rect.table,
                               mapping = aes(xmax = -20000000, xmin = -40000000,
                                             ymin = hmin, ymax = hmax,fill = patient))
            patient.rect.legend <- (ggplot()+
                                    geom_rect(data = patient.rect.table,
                                            mapping = aes(xmax = -20000000, 
                                                        xmin = -40000000,
                                                        ymin = hmin, 
                                                        ymax = hmax,
                                                        fill = patient)) +
                                    scale_fill_manual(values = allScales,name = "Patient")) %>% 
                                     ggplotGrob %>%
                 {.$grobs[[which(sapply(.$grobs, function(x) {x$name}) == "guide-box")]]}
            type.legend <- (ggplot() + 
                            geom_rect(data = CNADat,
                                      mapping = aes(xmin = Update_Start,
                                                    xmax = Update_End,
                                                    ymin = hmin, ymax = hmax, fill = Type))+
                            scale_fill_manual(values = allScales,name = "Type"))%>% 
                           ggplotGrob %>% {.$grobs[[which(sapply(.$grobs, function(x) {x$name}) == "guide-box")]]}
            legendColumn <-
                plot_grid(
                    type.legend + theme(legend.position = "none"),
                    type.legend,
                    patient.rect.legend,
                    type.legend + theme(legend.position = "none"),
                    ncol = 1,
                    align = "v") + theme(plot.margin = margin(l = -30)) 
            p <- plot_grid(mainplot + theme(legend.position = "none"),
                      legendColumn,
                      rel_widths = c(1,.15))
        }
        
        CNAplot.list[[patient]] <- p
    }
    return(CNAplot.list)
}

# plotCNA <- function(seg, sample.as.col = FALSE, show.GISTIC.gene = FALSE){
#     if(show.GISTIC.gene){
#         if(!"Gistic.type" %in% colnames(seg)){
#             stop("seg does not contain GISTIC gene information")
#         }
#     }
#     s <- seg[,round(mean(CopyNumber)),by = .(Tumor_Sample_Barcode,Gene)]
#     matDat <- dplyr::mutate(s, Type = unlist(lapply(s$V1, function(x){
#         if(x == 0){
#             return("Deletion")
#         }
#         else if(x == 1){
#             return("Loss")
#         }
#         else if(x == 2){
#             return(NA)
#         }
#         else if(x == 3){
#             return("Gain")
#         }
#         else if(x >=4){
#             return("Amplification")
#         }
#     })))%>% as.data.table() %>% dcast(Gene~ Tumor_Sample_Barcode, value.var = "Type")
#     geneOrder <- factor(matDat$Gene, levels = unique(s[,Gene]))
#     matDat <- matDat[order(geneOrder), ]
#     mat <- as.matrix(matDat[,-1])
#     # rownames(mat) <- matDat[,Gene]
#     colType <- c("Loss" = "#6baed6", "Deletion" = "#084594",
#                  "Gain" = "#f4a582","Amplification" = "#d73027")
#     if(sample.as.col){
#         ComplexHeatmap::Heatmap(mat, col = colType, name = "Type", na_col = "#f0f0f0", row_labels = rownames(mat))
#     }
#     else{
#         ComplexHeatmap::Heatmap(t(mat), col = colType, name = "Type", na_col = "#f0f0f0", row_names_side = "left", column_labels = colnames(t(mat)))
#     }
# }
# updateStart <- apply(seg,1,updatePosition,"Start_Position","Chromosome",chrLens)
# updateEnd <- apply(seg,1,updatePosition,"End_Position","Chromosome",chrLens)
# suppressWarnings(seg[,Update_Start:= updateStart]) 
# suppressWarnings(seg[,Update_End:= updateEnd])
# chrLabels = c(1:22)
# chrTable = data.table::data.table(chr = chrLabels, start = chrLens[1:(length(chrLens)-3)], end = chrLens[2:(length(chrLens)-2)])
# chrTable$color = rep(c('black','white'), length = nrow(chrTable))
# seg[,hmin := 0]
# seg[,hmax := 0]
# samples <- sort(unique(seg$Tumor_Sample_Barcode))  
# h = 0.2
# for(sample in samples){
#     h <- h + 0.1
#     seg[Tumor_Sample_Barcode == sample,]$hmin <- h
#     seg[Tumor_Sample_Barcode == sample,]$hmax <- h + 0.5
#     h <- h + 0.5
# }
# 
# min <- sort(unique(seg$hmin))
# max <- sort(unique(seg$hmax))
# CNADat <- seg[Type != "Neutral",]
# CNADat$Type <- factor(CNADat$Type, levels = c("Loss","Deletion","Gain","Amplification"))
# textTable <- data.table::data.table(Pos = min + (max-min)/2, Tumor_Sample_Barcode = samples)
# backgroundTable <- data.table::data.table(xmin = min(chrTable$start), xmax = max(chrTable$end), ymin = min, ymax = max)
# p <- ggplot()+geom_rect(data = chrTable,mapping = aes(xmin = start, xmax = end, ymin = 0, ymax = 0.3),fill = rep(c("black","white"),11),color = "black")+
#     geom_text(data = chrTable,mapping = aes(x = start + (end - start)/2, y = 0.15, label = chr), color = rep(c("white","black"),11))+
#     geom_rect(data = backgroundTable, mapping = aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), fill = "#f0f0f0")+
#     theme(panel.grid =element_blank(),axis.text = element_blank(),axis.ticks = element_blank(),
#           panel.border = element_blank(),panel.background = element_blank(),
#           axis.title.x = element_blank(),axis.title.y = element_blank())+
#     geom_rect(data = CNADat,mapping = aes(xmin = Update_Start, xmax = Update_End, ymin = hmin, ymax = hmax, fill = Type))+
#     scale_fill_manual(breaks = c("Loss","Deletion","Gain","Amplification"), values = c("#6baed6","#084594","#f4a582","#d73027"))+
#     geom_text(data = textTable,mapping = aes(x = -300000000, y = Pos, label = Tumor_Sample_Barcode))+
#     geom_segment(aes(x = -100000000, y = c(min,max(max)), xend = max(chrTable$end)+100000000, yend = c(min,max(max))), linetype="dashed")+
#     geom_segment(aes(x = c(-100000000,max(chrTable$end)+100000000), y = min(min), xend = c(-100000000,max(chrTable$end)+100000000), yend = max(max)), linetype="dashed")
