#' @title plotCNA
#'  
#' @param seg object generated by readSegment function.
#' @param refBuild human reference genome versions of hg18, hg19 or hg38 by UCSC. Default "hg19".
#' @param show.GISTIC.gene Whether GISTIC gene in seg.Default FALSE.
#' 
#' @examples
#' seg <- readSegment(segCN.file = segCN.file)
#' plotCNA(seg)
#'
#' @export plotCNA
#'

plotCNA <- function(seg, refBuild = "hg19", show.GISTIC.gene = FALSE){
    if(show.GISTIC.gene){
      if(!"Gistic.type" %in% colnames(seg)){
        stop("seg does not contain GISTIC gene information")
      }
    }
    if("Start" %in% colnames(seg)){
      seg <- dplyr::rename(seg,Start_Position = Start, End_Position = End)
    }
    if(refBuild == 'hg19'){
        chrLens = c(249250621, 243199373, 198022430, 191154276, 180915260, 171115067, 159138663,
                     146364022, 141213431, 135534747, 135006516, 133851895, 115169878, 107349540,
                     102531392, 90354753, 81195210, 78077248, 59128983, 63025520, 48129895, 51304566,
                     155270560, 59373566)
    } else if(refBuild == 'hg18'){
        chrLens = c(247249719, 242951149, 199501827, 191273063, 180857866, 170899992,
                     158821424, 146274826, 140273252, 135374737, 134452384, 132349534,
                     114142980, 106368585, 100338915, 88827254, 78774742, 76117153,
                     63811651, 62435964, 46944323, 49691432, 154913754, 57772954)
    } else if(refBuild == 'hg38'){ #hg38
        chrLens = c(248956422, 242193529, 198295559, 190214555, 181538259, 170805979,
                     159345973, 145138636, 138394717, 133797422, 135086622, 133275309,
                     114364328, 107043718, 101991189, 90338345, 83257441, 80373285,
                     58617616, 64444167, 46709983, 50818468, 156040895, 57227415)
    } else{
        stop('Available reference builds: hg18, hg19, hg38')
    }
    chrLens =  append(cumsum(chrLens),1,after = 0)
    updatePosition <- function(x,pos,chrname,chrLens){
        return(as.numeric(x[pos]) + as.numeric(chrLens[as.numeric(x[chrname]) ]) )
    }
    updateStart <- apply(seg,1,updatePosition,"Start_Position","Chromosome",chrLens)
    updateEnd <- apply(seg,1,updatePosition,"End_Position","Chromosome",chrLens)
    suppressWarnings(seg[,Update_Start:= updateStart]) 
    suppressWarnings(seg[,Update_End:= updateEnd])
    chrLabels = c(1:22)
    chrTable = data.table::data.table(chr = chrLabels, start = chrLens[1:(length(chrLens)-3)], end = chrLens[2:(length(chrLens)-2)])
    chrTable$color = rep(c('black','white'), length = nrow(chrTable))
    seg[,hmin := 0]
    seg[,hmax := 0]
    samples <- sort(unique(seg$Sample))  
    h = 0.2
    for(sample in samples){
        h <- h + 0.1
        seg[Sample == sample,]$hmin <- h
        seg[Sample == sample,]$hmax <- h + 0.5
        h <- h + 0.5
    }
    min <- sort(unique(seg$hmin))
    max <- sort(unique(seg$hmax))
    CNADat <- seg[Type != "Neutral",]
    CNADat$Type <- factor(CNADat$Type, levels = c("Loss","Deletion","Gain","Amplification"))
    textTable <- data.table::data.table(Pos = min + (max-min)/2, Sample = samples)
    backgroundTable <- data.table::data.table(xmin = min(chrTable$start), xmax = max(chrTable$end), ymin = min, ymax = max)
    p <- ggplot()+geom_rect(data = chrTable,mapping = aes(xmin = start, xmax = end, ymin = 0, ymax = 0.3),fill = rep(c("black","white"),11),color = "black")+
        geom_text(data = chrTable,mapping = aes(x = start + (end - start)/2, y = 0.15, label = chr), color = rep(c("white","black"),11))+
        geom_rect(data = backgroundTable, mapping = aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), fill = "#f0f0f0")+
        theme(panel.grid =element_blank(),axis.text = element_blank(),axis.ticks = element_blank(),
              panel.border = element_blank(),panel.background = element_blank(),
              axis.title.x = element_blank(),axis.title.y = element_blank())+
        geom_rect(data = CNADat,mapping = aes(xmin = Update_Start, xmax = Update_End, ymin = hmin, ymax = hmax, fill = Type))+
        scale_fill_manual(breaks = c("Loss","Deletion","Gain","Amplification"), values = c("#6baed6","#084594","#f4a582","#d73027"))+
        geom_text(data = textTable,mapping = aes(x = -300000000, y = Pos, label = Sample))+
        geom_segment(aes(x = -100000000, y = c(min,max(max)), xend = max(chrTable$end)+100000000, yend = c(min,max(max))), linetype="dashed")+
        geom_segment(aes(x = c(-100000000,max(chrTable$end)+100000000), y = min(min), xend = c(-100000000,max(chrTable$end)+100000000), yend = max(max)), linetype="dashed")
    return(p)
}

# plotCNA <- function(seg, sample.as.col = FALSE, show.GISTIC.gene = FALSE){
#     if(show.GISTIC.gene){
#         if(!"Gistic.type" %in% colnames(seg)){
#             stop("seg does not contain GISTIC gene information")
#         }
#     }
#     s <- seg[,round(mean(CopyNumber)),by = .(Sample,Gene)]
#     matDat <- dplyr::mutate(s, Type = unlist(lapply(s$V1, function(x){
#         if(x == 0){
#             return("Deletion")
#         }
#         else if(x == 1){
#             return("Loss")
#         }
#         else if(x == 2){
#             return(NA)
#         }
#         else if(x == 3){
#             return("Gain")
#         }
#         else if(x >=4){
#             return("Amplification")
#         }
#     })))%>% as.data.table() %>% dcast(Gene~ Sample, value.var = "Type")
#     geneOrder <- factor(matDat$Gene, levels = unique(s[,Gene]))
#     matDat <- matDat[order(geneOrder), ]
#     mat <- as.matrix(matDat[,-1])
#     # rownames(mat) <- matDat[,Gene]
#     colType <- c("Loss" = "#6baed6", "Deletion" = "#084594",
#                  "Gain" = "#f4a582","Amplification" = "#d73027")
#     if(sample.as.col){
#         ComplexHeatmap::Heatmap(mat, col = colType, name = "Type", na_col = "#f0f0f0", row_labels = rownames(mat))
#     }
#     else{
#         ComplexHeatmap::Heatmap(t(mat), col = colType, name = "Type", na_col = "#f0f0f0", row_names_side = "left", column_labels = colnames(t(mat)))
#     }
# }