#' @title plotPhyloTree
#' 
#' @param phyloTree a phyloTree object generated by getphyloTree function.
#' @param branchCol character."mutSig" for coloring branches by mutation signature; "mutType" for coloring branches by branch tumor type; NULL for coloring automatically.Default is 'mutSig'.
#' @param show.bootstrap logical.Whether to add bootstrap value on internal nodes.Default is TRUE.
#' @param use.box logical.Whether to add box around bootstrap value on tree. Default is TRUE.
#' @param min.ratio double (Default:1/30). If min.ratio is not NULL,
#'  all edge length of a phylogenetic tree should be greater than
#'  min.ratio*the longest edge length.
#'  If not, the edge length will be reset as min.ratio*longest edge length.
#' @param min.mut.count the threshold for the variants in a branch. Default 15.
#' @param signaturesRef The parameter used for deconstructSig. Default "cosmic_v2". Option: "genome_cosmic_v3","exome_cosmic_v3","nature2013".
#' @param patient.id select the specific patients. Default: NULL, all patients are included
#' @param geneList list of genes to restrict the analysis. Default NULL.
#' 
#' are mapped along the trees as indicated
#' @examples
#' plotPhyloTree(phyloTree)
#
#' @return return a list of phylotree graph .
#' @import reshape2 ape ggplot2 ggrepel
#' @export plotPhyloTree
#' 


plotPhyloTree <- function(phyloTree = NULL,
                          branchCol = "mutSig",
                          show.bootstrap = TRUE,
                          use.box = TRUE,
                          min.ratio = 1/30,
                          min.mut.count = 15,
                          signaturesRef="cosmic_v2",
                          patient.id = NULL){
   # heatmap.options <- c('binary','CCF')
   # if(!heatmap.type %in% heatmap.options){
   #    stop("Type of heatmap can only be 'binary' or 'CCF'")
   # }
   # if(heatmap.type == 'binary'){
   #   use.ccf = FALSE
   # }else if(heatmap.type == 'CCF'){
   #     use.ccf = TRUE 
   # }
   
   branchCol.options <- c("mutSig","mutType")
   if(!is.null(branchCol)){
       if(!branchCol %in% branchCol.options){
           stop("Type of branchCol can only be 'mutSig','mutType' or NULL")
       }
   }
   
   if(!is.null(patient.id)){
       patient.setdiff <- setdiff(patient.id, names(phyloTree))
       if(length(patient.setdiff) > 0){
           stop(paste0(patient.setdiff, " can not be found in your data"))
       }
       phyloTree <- phyloTree[names(phyloTree)  %in% patient.id] 
   }
   tree.list <- lapply(phyloTree, drawPhyloTree,
                       branchCol = branchCol,
                       show.bootstrap = show.bootstrap,
                       use.box = use.box,
                       min.ratio = min.ratio,
                       min.mut.count = min.mut.count,
                       signaturesRef = signaturesRef)
   return(tree.list)
}