#' @title readSegment
#'  
#' @param segCN.file character.Path to  copy number file. Column names should be Sample, Chromosome, Start_Position, End_Position, CopyNumber
#' @param gtf.dat character.Path to GTF data.Default NULL.
#' @param gisticAmpGenesFile Amplification Genes file generated by GISTIC. Default NULL.
#' @param gisticDelGenesFile Deletion Genes file generated by GISTIC. Default NULL.
#' @param verbose Whether to display details in the console.Default True.
#' @param cluster Default FALSE.
#' @param min.seg.size Mininum length of segs.Default is 500.
#' 
#' @examples
# segCN.file <- system.file("extdata", "HCC6046.seg.txt", package = "MesKit")
# # Load gtf.dat
# load(system.file("extdata", "Homo_sapiens.GRCh37.75.RData", package = "MesKit"))
# gisticAmpGenesFile <- system.file("extdata", "LIHC_amp_genes.conf_99.txt", package = "MesKit")
# gisticDelGenesFile <- system.file("extdata", "LIHC_del_genes.conf_99.txt", package = "MesKit")
# seg <- readSegment(segCN.file = segCN.file, gtf.dat = gtf.dat, gisticAmpGenesFile = gisticAmpGenesFile, gisticDelGenesFile = gisticDelGenesFile)
#'
#' @importFrom data.table setkey
#' @importFrom data.table as.data.table
#' @importFrom data.table data.table
#' @importFrom data.table foverlaps
#' @export readSegment
#'

readSegment <- function(segCN.file = NULL, gtf.dat = NULL,
                        gisticAmpGenesFile = NULL, 
                        gisticDelGenesFile = NULL,
                        verbose = TRUE, cluster = FALSE, min.seg.size = 500){
  seg <- suppressWarnings(data.table::fread(segCN.file, header=TRUE, sep="\t", stringsAsFactors = FALSE))
  
  # Replace chr with numeric value for better ordering
  seg$Chromosome = gsub(pattern = 'chr', replacement = '', x = seg$Chromosome, fixed = TRUE)
  seg$Chromosome = gsub(pattern = 'X', replacement = '23', x = seg$Chromosome, fixed = TRUE)
  seg$Chromosome = gsub(pattern = 'Y', replacement = '24', x = seg$Chromosome, fixed = TRUE) 
  
  if(!"CopyNumber" %in% colnames(seg)){
    if("SegmentMean" %in% colnames(seg)){
        suppressWarnings(seg[,CopyNumber := round(2^(SegmentMean)*2) ])
        seg <- dplyr::select(seg, Sample, Chromosome, Start, End, CopyNumber)
    }
    else{
        stop("segCN.file does not contain Copynumber or SegmentMean information")  
    }
  }
  seg[ ,Width := End - Start]
  seg <- dplyr::mutate(seg, Type = unlist(lapply(seg$CopyNumber, function(x){
      if(x == 0){
          return("Deletion")
      }
      else if(x == 1){
          return("Loss")
      }
      else if(x == 2){
          return("Neutral")
      }
      else if(x == 3){
          return("Gain")
      }
      else if(x >=4){
          return("Amplification")
      }
  }))) %>%
      dplyr::filter(Chromosome %in% seq(1:22)& Width > min.seg.size) %>%
      dplyr::select(Chromosome, Start , End, Sample, CopyNumber, Type) %>% 
       as.data.table()
  
  if(!is.null(gtf.dat)){
      segDat <- dplyr::rename(seg, seqnames = Chromosome, start = Start, end = End)
      genesDat <- gtf.dat
      data.table::setkey(x = genesDat, seqnames, start, end)
      data.table::setkey(x = segDat, seqnames, start, end)
      overlapsDat <- foverlaps(x = genesDat, y = segDat, by.x = c("seqnames", "start", "end"))
      overlapsDat <- na.omit(overlapsDat)
      overlapsDat <- overlapsDat[,.(seqnames, i.start, i.end, 
                                    gene_name,Sample,CopyNumber,  Type)]
      colnames(overlapsDat) <- c( "Chromosome", "Start_Position", "End_Position", 
                                 "Gene","Sample","CopyNumber", "Type")
      seg <- overlapsDat
      # read gistic results
      gisticCNVgenes <- data.table::data.table()
      if(!is.null(gisticAmpGenesFile)){
          if(verbose){
              cat(paste0('--Processing ', basename(gisticAmpGenesFile), '\n'))
          }
          gisticAmpGenes <- readGistic(gisticAmpGenesFile, "AMP")
          gisticCNVgenes <- rbind(gisticCNVgenes, gisticAmpGenes)
      }   
      if(!is.null(gisticDelGenesFile)){
          if(verbose){
              cat(paste0('--Processing ', basename(gisticDelGenesFile), '\n'))
          }
          gisticDelGenes <- readGistic(gisticDelGenesFile, "Del")
          gisticCNVgenes <- rbind(gisticCNVgenes, gisticDelGenes)
      } 
      if(nrow(gisticCNVgenes) != 0){
          ## combine segment file and gistic results
          gisticCNVgenes[,Chromosome := sapply(strsplit(x = gisticCNVgenes[,Wide_Peak_Boundaries], split = ':'), '[', 1)]
          gisticCNVgenes[,loc := sapply(strsplit(x = gisticCNVgenes[,Wide_Peak_Boundaries], split = ':'), '[', 2)]
          gisticCNVgenes[,Start_Position := as.integer(sapply(strsplit(x = gisticCNVgenes[,loc], split = '-'), '[', 1))]
          gisticCNVgenes[,End_Position := as.integer(sapply(strsplit(x = gisticCNVgenes[,loc], split = '-'), '[', 2))]
          gisticCNVgenes[,Chromosome := gsub(pattern = 'chr', replacement = '', x = gisticCNVgenes$Chromosome, fixed = TRUE)]
          gisticCNVgenes[,Chromosome := gsub(pattern = 'X', replacement = '23', x = gisticCNVgenes$Chromosome, fixed = TRUE)]
          gisticCNVgenes[,Chromosome := gsub(pattern = 'Y', replacement = '24', x = gisticCNVgenes$Chromosome, fixed = TRUE)]
          g <- dplyr::select(gisticCNVgenes, Gene, Chromosome, Start_Position, End_Position, Gistic.type, Cytoband)
          # seg <- as.data.table(seg) %>% plyr::rename(c("Start" = "Start_Position", "End" = "End_Position"))
          data.table::setkey(x = g, Chromosome, Start_Position, End_Position)
          data.table::setkey(x = seg, Chromosome, Start_Position, End_Position)
          mapDat <- data.table::foverlaps(seg, g, by.x = c("Chromosome", "Start_Position", "End_Position"))
          # mapDat$Gistic.type <- ifelse(test = is.na(mapDat$Gistic.type), yes = "NA", no = mapDat$Gistic.type)
          result <- dplyr::select(mapDat, Type, Gistic.type, Sample, i.Gene, CopyNumber, Chromosome, i.Start_Position, i.End_Position) %>%
              dplyr::distinct(.)
          result <- dplyr::rename(result, Gene = i.Gene, Start_Position = i.Start_Position, End_Position = i.End_Position) %>% 
              dplyr::select(Chromosome, Start_Position, End_Position, Gene, Sample, CopyNumber, Type, Gistic.type)
          seg <- result[(Type %in% c("Loss", "Deletion") & Gistic.type  == "Del")|(Type %in% c("Gain", "Amplification") & Gistic.type  == "AMP"),]
      }
   }  
  return(seg)
}


# extract recurrent CNA genes from gistic file
readGistic <- function(gisticGenesFile = NULL, Gistic.type = NULL){
    
    if(!is.null(gisticGenesFile)){
        gisticGenes <- data.table::fread(input = gisticGenesFile, stringsAsFactors = FALSE, header = TRUE)
        if(is.na(unique(gisticGenes[, ncol(gisticGenes), with = FALSE])[1])){     
            gisticGenes <- gisticGenes[,-ncol(gisticGenes), with = FALSE]
        }
        
        #wide peak boundaries
        wpb <- as.character(gisticGenes[3]) 
        wpb <- wpb[2:length(wpb)]
        
        # keep the information of peak boundaries and genes 
        gisticGenes <- gisticGenes[c(4:nrow(gisticGenes)),]
        gisticGenes$cytoband <- gsub(pattern = ' ', replacement = '_', x = gisticGenes$cytoband)
        colnames(gisticGenes) <- c( 'cytoband', paste0( colnames(gisticGenes)[2:length(colnames(gisticGenes))], '_',wpb))
        
        gisticGenes <- suppressWarnings(data.table::melt(gisticGenes, id.vars = 'cytoband'))
        gisticGenes <- gisticGenes[!value %in% '']
        
        #gisticGenes <- gisticGenes[!grep(pattern = '|', x = gisticGenes$value, fixed = TRUE)] #remove genes with ambiguous annotation
        gisticGenes <- gisticGenes[,.(variable, value)] %>%
            tidyr::separate(col = variable, into = c("Cytoband", "Wide_Peak_Boundaries"), 
                            sep = "_", remove = TRUE) %>%
            dplyr::mutate(Gene=sub("\\|.+", "", value), Gistic.type = Gistic.type) %>%
            dplyr::select(-value)
        
        return(gisticGenes)
    }
}