#' Subset Maf object
#'
#' @param maf Maf object generated by readMaf.
#' @param patient.id Select the specific patients. Default: NULL, all patients are included.
#' @param chrSilent Chromosomes excluded in the analysis. e.g, chrX, chrY. Default NULL.
#' @param mutType Select Proper variant classification you need. Default "All". Option: "nonSyn".
#' @param use.indel Logical value. Whether to use INDELs besides somatic SNVs. Default: TRUE.
#' @param min.vaf The minimum VAF for filtering variants. Default: 0.
#' @param max.vaf The maximum VAF for filtering variants. Default: 1.
#' @param min.average.vaf The minimum tumor average VAF for filtering variants. Default: 0.
#' @param min.average.adj.vaf The minimum tumor average ajust VAF for filtering variants. Default: 0.
#' @param min.ccf The minimum CCF for filtering variants. Default: 0.
#' @param min.ref.depth The minimum reference allele depth for filtering variants. Default: 0.
#' @param min.alt.depth The minimum alteratation allele depth for filtering variants. Default: 0.
#' @param min.total.depth The minimum total allele depth for filtering variants. Default: 0.
#' @param clonalStatus Subset by clonal status.Default: NULL.Option: "Clonal","Subclonal".
#' @param use.adjVAF Let VAF = VAF_adj, Tumor_Average_VAF = Tumor_Average_VAF_adj.Default: FALSE. 
#'
#'
#' @examples
#' maf <- readMaf(mafFile=maf.File, refBuild="hg19")
#' maf_data <- subsetMaf(maf)
#' @return Maf data.
#' @export

subsetMaf <- function(maf,
                      patient.id = NULL,
                      geneList = NULL,
                      chrSilent = NULL,
                      mutType = "All",
                      use.indel = TRUE,
                      min.vaf = 0,
                      max.vaf = 1,
                      min.average.vaf = NULL,
                      min.average.adj.vaf = NULL,
                      min.ccf = NULL,
                      min.ref.depth = 0,
                      min.alt.depth = 0,
                      min.total.depth = 0,
                      clonalStatus = NULL,
                      use.adjVAF = FALSE){
   
   maf_data <- maf@data
   nonSyn.vc <- maf@nonSyn.vc
   
   if(use.adjVAF){
       if(!"VAF_adj" %in% colnames(maf_data)){
           stop("Error:Adjusted VAF was not found in maf data.")
       }
       else{
           maf_data$VAF <- maf_data$VAF_adj
           maf_data$Tumor_Average_VAF <- maf_data$Tumor_Average_VAF_adj
       }
   }
   
   
   # ## patient filter
   # if(!is.null(patient.id)){
   #    patient.setdiff <- setdiff(patient.id, unique(maf_data$Patient_ID))
   #    if(length(patient.setdiff) > 0){
   #       stop(paste0("Patient ", patient.setdiff, " can not be found in your data"))
   #    }
   #    maf_data <- maf_data[Patient_ID %in% patient.id]
   # }
   
   ## chromosome filter
   if (!is.null(chrSilent)) {
      maf_data <- maf_data[Chromosome %in% chrSilent]
   }
   
   ## filter variant classification
   mutType <- match.arg(mutType, choices = c("All","nonSyn"), several.ok = FALSE)
   if (mutType == "nonSyn") {
      maf_data <-  maf_data[Variant_Classification %in% nonSyn.vc]
   }
   
   ## use.indel filter
   if (!use.indel) {
      maf_data <- maf_data[Variant_Type == "SNP"]
   }
   
   ## Select mutations in selected genes
   if(!is.null(geneList)){
      maf_data <- maf_data[Hugo_Symbol %in% geneList]
   }
   
   ## allele depth filter
   maf_data$Ref_allele_depth[is.na(maf_data$Ref_allele_depth)] <- 0
   maf_data$Alt_allele_depth[is.na(maf_data$Alt_allele_depth)] <- 0
   maf_data <- maf_data[Ref_allele_depth>=min.ref.depth&
                      Alt_allele_depth>=min.alt.depth &
                      Ref_allele_depth + Alt_allele_depth>=min.total.depth]
   
   ## vaf filter
   maf_data <- maf_data[VAF >= min.vaf & VAF <= max.vaf]
   if(!is.null(min.average.vaf)){
       if(min.average.vaf > 0){
           maf_data <- maf_data[Tumor_Average_VAF >=min.average.vaf]
       }
   }
   
   if("VAF_adj" %in% colnames(maf_data)){
       if(!is.null(min.average.adj.vaf)){
           if(min.average.adj.vaf > 0){
               maf_data <- maf_data[Tumor_Average_VAF_adj>=min.average.adj.vaf]
           }
       }
   }
   
   ## ccf filter
   if("CCF" %in% colnames(maf_data)){
      if(!is.null(min.ccf)){
          if(min.ccf > 0){
              maf_data <- maf_data[CCF >= min.ccf]
          }
      }
   }
   
   if(!is.null(clonalStatus)){
      if(!"Clonal_Status" %in% colnames(maf_data)){
         stop("Error:Missing information about clonal status in maf data")
      }
      clonalStatus <- match.arg(clonalStatus, choices = c("Clonal", "Subclonal"))
      maf_data <- maf_data[Clonal_Status == clonalStatus]
   }
   
   return(maf_data) 
}