#'  treeMutSig
#' @description Figure out the intersection of the variants on different branches 
#' in phyloTree object and output the treeMSOutput for further summary and plot functions.
#' 
#' 
#' @param phyloTree a phyloTree object generated by phyloTree function.
#' @param geneList list of genes to restrict the analysis. Default NULL.
#' @param min.mut.count the threshold for the variants in a branch. Default 15.
#' @param signaturesRef The parameter used for deconstructSig. Default "cosmic_v2". Option: "genome_cosmic_v3","exome_cosmic_v3","nature2013".
#' @param tri.counts.method The parameter used for deconstructSig.
#' @param plot output a list of diagrams that visualize mutational signature of trunk/branches in a phylogenetic tree.Default FALSE. 
#' @param patient.id select the specific patients. Default: NULL, all patients are included
#' @return a list of data frames, each one contains treeMSOutput, containing information about each set/branch's mutational signature.
#' 
#' @examples
#' treeMutSig(phyloTree, geneList=NULL, min.mut.count=15, signaturesRef="cosmic")
#' treeMutSig(phyloTree, plot = T)
#' @export  treeMutSig


## Mutational Signature function
treeMutSig <- function(phyloTree,
                       geneList=NULL,
                       min.mut.count=15,
                       signaturesRef="cosmic_v2",
                       tri.counts.method = "default",
                       withinType = FALSE,
                       plot = TRUE,
                       patient.id = NULL){
    
    signaturesRef.options <- c("cosmic_v2","nature2013","genome_cosmic_v3","exome_cosmic_v3")
    if(!signaturesRef %in% signaturesRef.options){
        stop("signaturesRef can only be either 'cosmic_v2','nature2013','genome_cosmic_v3' or 'exome_cosmic_v3'")
    } 
    
    if(!is.null(patient.id)){
        patient.setdiff <- setdiff(patient.id, names(phyloTree))
        if(length(patient.setdiff) > 0){
            stop(paste0(patient.setdiff, " can not be found in your data"))
        }
        phyloTree <- phyloTree[names(phyloTree)  %in% patient.id] 
    }
    
    treeMSOutput <- lapply(phyloTree, doTreeMutSig,
                           geneList = geneList,
                           min.mut.count = min.mut.count,
                           signaturesRef = signaturesRef,
                           tri.counts.method = tri.counts.method,
                           withinType = withinType)
    mutSigSummary <- lapply(treeMSOutput, doMutSigSummary, withinType)
    
    mutSig.plot <- NA
    if(plot){
        mutSig.plot <- lapply(treeMSOutput, doPlotMutSig, withinType)
        return(list(mutSig.summary = mutSigSummary, mutSig.plot = mutSig.plot))
    }
    
    return(list(mutSig.summary = mutSigSummary))
}