#'  treeMutSig
#' @description Figure out the intersection of the variants on different branches 
#' in phyloTree_list object and output the treeMSOutput for further summary and plot functions.
#' 
#' 
#' @param phyloTree_list a list of phyloTree object generated by getPhyloTree function.
#' @param geneList list of genes to restrict the analysis. Default NULL.
#' @param min.mut.count the threshold for the variants in a branch. Default 15.
#' @param signaturesRef Signature reference,Users can upload their own reference. Default "cosmic_v2". Option: "genome_cosmic_v3","exome_cosmic_v3","nature2013".
#' @param associated associated Vector of associated signatures. If given, will narrow the signatures reference to only the ones listed.Default NULL.
#' @param signature.cutoff Discard any signature contributions with a weight less than this amount.Default: 0.1.
#' @param patient.id select the specific patients. Default: NULL, all patients are included
#' @param withinTumor Exploring signatures within tumor.Default is FALSE.
#' @return a list of data frames, each one contains treeMSOutput, containing information about each set/branch's mutational signature.
#' 
#' @importFrom pracma lsqnonneg
#' @importFrom Biostrings getSeq
#' @examples
#' treeMutSig(phyloTree_list, geneList=NULL, min.mut.count=15, signaturesRef="cosmic")
#' treeMutSig(phyloTree_list, plot = T)
#' @export  treeMutSig


## Mutational Signature function
treeMutSig <- function(phyloTree_list,
                       geneList=NULL,
                       min.mut.count=15,
                       signaturesRef="cosmic_v2",
                       associated = c(), 
                       signature.cutoff = 0.1,
                       withinTumor = FALSE,
                       patient.id = NULL){
    
    

    if(!is.null(patient.id)){
        patient.setdiff <- setdiff(patient.id, names(phyloTree_list))
        if(length(patient.setdiff) > 0){
            stop(paste0(patient.setdiff, " can not be found in your data"))
        }
        phyloTree_list <- phyloTree_list[names(phyloTree_list)  %in% patient.id] 
    }
    treeMSOutput <- lapply(phyloTree_list, doTreeMutSig,
                           geneList = geneList,
                           min.mut.count = min.mut.count,
                           signaturesRef = signaturesRef,
                           withinTumor = withinTumor,
                           associated = associated, 
                           signature.cutoff = signature.cutoff)
    mutSig.summary <- lapply(treeMSOutput, doMutSigSummary, withinTumor)
    cos_sim.mat <- lapply(treeMSOutput,function(x)x$cos_sim.mat)
    mutSig.spectrum <- lapply(treeMSOutput,function(x)x$spectrum_df)
    
    return(list(mutSig.summary = mutSig.summary, cos_sim.mat = cos_sim.mat, mutSig.spectrum =  mutSig.spectrum))
}