#'  treeMutSig
#' @description Figure out the intersection of the variants on different branches 
#' in phyloTree object and output the treeMSOutput for further summary and plot functions.
#' 
#' 
#' @param phyloTree a phyloTree object generated by phyloTree function.
#' @param geneList list of genes to restrict the analysis. Default NULL.
#' @param min.mut.count the threshold for the variants in a branch. Default 15.
#' @param signaturesRef The parameter used for deconstructSig. Default "cosmic". Option: "nature2013". 
#' @param plot output a list of diagrams that visualize mutational signature of trunk/branches in a phylogenetic tree.Default FALSE. 
#' @param patient.id select the specific patients. Default: NULL, all patients are included
#' @return a list of data frames, each one contains treeMSOutput, containing information about each set/branch's mutational signature.
#' 
#' @examples
#' treeMutSig(phyloTree, geneList=NULL, min.mut.count=15, signaturesRef="cosmic")
#' treeMutSig(phyloTree, plot = T)
#' @export  treeMutSig


## Mutational Signature function
 treeMutSig <- function(phyloTree,
                        geneList=NULL,
                        min.mut.count=15,
                        signaturesRef="cosmic",
                        plot = FALSE,
                        patient.id = NULL){

     if(!is.null(patient.id)){
         patient.setdiff <- setdiff(patient.id, names(phyloTree))
         if(length(patient.setdiff) > 0){
             stop(paste0(patient.setdiff, " can not be found in your data"))
         }
         phyloTree <- phyloTree[names(phyloTree)  %in% patient.id] 
     }

     treeMSOutput <- lapply(phyloTree, doTreeMutSig,
                            geneList = geneList,
                           min.mut.count = min.mut.count,
                           signaturesRef = signaturesRef)
     mutSigSummary <- lapply(treeMSOutput, doMutSigSummary)

     mutSig.plot <- NA
     if(plot){
         mutSig.plot <- lapply(treeMSOutput, doPlotMutSig)
     }
         
     return(list(mutSig.summary = mutSigSummary, mutSig.plot = mutSig.plot))
}

doTreeMutSig <- function(phyloTree,
                         geneList=NULL,
                         min.mut.count=15,
                         signaturesRef="cosmic",
                         MTB = FALSE){

    refBuild <- phyloTree@refBuild
    ref.options = c('hg18', 'hg19', 'hg38')
    if(!refBuild %in% ref.options){
        stop("refBuild can only be either 'hg18', 'hg19' or 'hg38'")
    }else {
        refBuild <- paste("BSgenome.Hsapiens.UCSC.", refBuild, sep = "")
    }
    
    ## get branches information from phyloTree object
    mutBranches <- phyloTree@mut.branches

    patientID <- phyloTree@patientID
    branchesName <- names(mutBranches)
    branchesNameList <- strsplit(branchesName, split='∩')
    
    ## regain the data frame of all branches with Sample
    mutSigRef <- data.frame()
    for (branchName in branchesName){
        mutBranch <- mutBranches[[branchName]]
        mutSigRef <- rbind(mutBranch, mutSigRef)
    }
    colnames(mutSigRef) <- c("Sample",  "chr", "pos", "pos_end", 
                             "ref", "alt", "Hugo_Symbol", "mut_id", 
                             "alias")
    
    ## Select mutations in selected genes
    if(!is.null(geneList)){
        selectedGenes <- geneList
        mutSigRef <- mutSigRef  %>%
            dplyr::rowwise() %>%
            dplyr::filter(any(strsplit(Hugo_Symbol, ",|;")[[1]] %in% selectedGenes)) %>%
            as.data.frame()
        branchesNameList <- strsplit(unique(mutSigRef$Sample), split='∩')
        branchesName <- unique(mutSigRef$Sample)
        
    }
    
    ## get the mutational signature of the branch 
    mutSigsOutput <- data.frame()
    lsPicName <- c()
    ## deconstructSigs
    sigsInput <- suppressWarnings(
        deconstructSigs::mut.to.sigs.input(mut.ref=mutSigRef, 
                                           sample.id="Sample", 
                                           chr="chr", 
                                           pos="pos", 
                                           ref="ref", 
                                           alt="alt",
                                           bsg=get(refBuild)))
    if(MTB){
        trunkName <- unique(mutSigRef[which(mutSigRef$alias == "T"), ]$Sample) 
        treeMSOutput <- list(
            sigsInput = sigsInput, 
            trunkName = trunkName,
            patientID = patientID)
        return(treeMSOutput)
    }
    for (branchCounter in length(branchesNameList):1){
        ## generate a single branch
        branch <- Filter(Negate(is.na), 
                         branchesNameList[[branchCounter]])
        branchName <- branchesName[branchCounter]
        ## get the mutational signature of the branch
        if (length(mutSigRef[which(
            mutSigRef$Sample == branchName), 1]) < min.mut.count){
            sigsMaxName <- "No Signature"
            sigsMaxProb <- 0
            message(paste0("Warnings: ",
                          "mutation number of Branch " , branchName,
                          " is less than the min.mut.count argument! ",
                          "This branch will be skipped")
            )
            if (any(mutSigRef[which(mutSigRef$Sample == branchName), ]$mut_id == "NoSigTag")) {
                mut.count <- 0
            }
            else{
                mut.count <- length(mutSigRef[which(mutSigRef$Sample == branchName), 1])
            }
        }else{
            if (signaturesRef == "cosmic") {
                sigsWhich <- deconstructSigs::whichSignatures(tumor.ref=sigsInput, 
                                                              signatures.ref=deconstructSigs::signatures.cosmic, 
                                                              sample.id=branchName,
                                                              contexts.needed=TRUE)
            } else if (signaturesRef == "nature2013") {
                sigsWhich <- deconstructSigs::whichSignatures(tumor.ref=sigsInput, 
                                                              signatures.ref=deconstructSigs::signatures.nature2013, 
                                                              sample.id=branchName,
                                                              contexts.needed=TRUE)
            }
            
            
            ## get mutational signature with max weight
            sigsMax <- sigsWhich[["weights"]][which.max(sigsWhich[["weights"]])]
            sigsMaxName <- colnames(sigsMax)
            sigsMaxName <- gsub('[.]', ' ', sigsMaxName)
            sigsMaxProb <- sigsMax[,1]
            
            mut.count <- length(mutSigRef[which(mutSigRef$Sample == branchName), 1])
        }
        
        ## vectorize branch name
        # branch <- gsub(paste(patientID,"-",sep=""), "", branch)
        
        mutSigsBranch <- data.frame(
            branch=c(branchName), 
            alias=as.character(unique(mutSigRef[which(mutSigRef$Sample == branchName), ]$alias)), 
            sig=sigsMaxName, 
            mut.count=mut.count, 
            sig.prob=format(round(sigsMaxProb, digits = 3), nsmall = 3))
        
        ## restrict to selected genes
        if (!is.null(geneList)){
            selectedGenes <- geneList
            pdgMut <- mutSigRef[which(
                mutSigRef$Sample == branchName &
                    as.character(mutSigRef$Hugo_Symbol) %in% selectedGenes),]
            pdgBranch <- as.character(pdgMut$Hugo_Symbol)
            ## collect branches' mutataional signature and selected genes
            mutSigsBranch <- cbind(mutSigsBranch, selected_genes=c(paste(pdgBranch, collapse = ",")))
        }
        
        
        ## collect branches' mutataional signature information
        mutSigsOutput <- rbind(mutSigsOutput, mutSigsBranch)
    }
    
    
    ## calculation process(maybe could be replaced by lapply)
    if (signaturesRef =="cosmic") {
        ## Aetiology from https://cancer.sanger.ac.uk/cosmic/signatures_v2 emm actually the additional feature may matter
        df.aetiology <- data.frame(
            aeti=c(
                "An endogenous mutational process initiated by spontaneous deamination of 5-methylcytosine",
                "Activity of the AID/APOBEC family of cytidine deaminases", 
                "Failure of DNA double-strand break-repair by homologous recombination",
                "Tobacco mutagens", 
                "Unknown", 
                "Defective DNA mismatch repair and is found in microsatellite unstable tumours",
                "Ultraviolet light exposure", 
                "Unknown", 
                "Polymerase η, which is implicated with the activity of AID during somatic hypermutation", 
                "Recurrent POLE somatic mutations, viz., Pro286Arg and Val411Leu.",
                "Treatments with the alkylating agent temozolomide", 
                "Unknown", 
                "Activity of the AID/APOBEC family of cytidine deaminases(C > U)",
                "Unknown", 
                "Defective DNA mismatch repair", 
                "Unknown", 
                "Unknown",
                "Unknown", 
                "Unknown", 
                "Defective DNA mismatch repair", 
                "Unknown",
                "Exposures to aristolochic acid", 
                "Unknown", 
                "Exposures to aflatoxin",
                "Unknown", 
                "Defective DNA mismatch repair", 
                "Unknown", 
                "Unknown",
                "Tobacco chewing habit", 
                "Unknown", 
                "Unknown"), 
            sig=c("Signature 1", "Signature 2", "Signature 3", "Signature 4", "Signature 5", "Signature 6",
                  "Signature 7", "Signature 8", "Signature 9", "Signature 10", "Signature 11", "Signature 12", 
                  "Signature 13", "Signature 14", "Signature 15", "Signature 16", "Signature 17", "Signature 18",
                  "Signature 19", "Signature 20", "Signature 21", "Signature 22", "Signature 23", "Signature 24", 
                  "Signature 25", "Signature 26", "Signature 27", "Signature 28", "Signature 29", "Signature 30", 
                  "No Signature")
        )
    } else if (signaturesRef =="nature2013") {
        ## Aetiology from https://www.nature.com/articles/nature12477#s1
        df.aetiology <- data.frame(
            aeti=c(
                "Deamination of 5-methyl-cytosine",
                "Deamination of 5-methyl-cytosine", 
                "AID/APOBEC family of cytidine deaminases",
                "Defective homologous-recombination-based DNA double-strand break repair", 
                "Tobacco carcinogens", 
                "Smoking history, C>T and T>C mutations",
                "Defective DNA mismatch repair", 
                "Ultraviolet-light-induced mutations",
                "Exogenous carcinogens, transcription-coupled nucleotide excision repair acting on bulky DNA adducts", 
                "Polymerase η, which is implicated with the activity of AID during somatic hypermutation",
                "Polymerase ε, altered activity of the error-prone polymerase Pol ε", 
                "Alkylating agent temozolomide", 
                "Exogenous carcinogens, transcription-coupled nucleotide excision repair acting on bulky DNA adducts",
                "AID/APOBEC family of cytidine deaminases", 
                "Uncharacterized defects in DNA maintenance", 
                "Uncharacterized defects in DNA maintenance", 
                "Exogenous carcinogens, transcription-coupled nucleotide excision repair acting on bulky DNA adducts",
                "Unknown", 
                "Unknown", 
                "Unknown",
                "Unknown",
                "Uncharacterized defects in DNA maintenance", 
                "Unknown",
                "Unknown",
                "Unknown", 
                "Unknown", 
                "Unknown", 
                "Unknown"), 
            sig=c("Signature 1A", "Signature 1B", "Signature 2", "Signature 3", "Signature 4", "Signature 5", "Signature 6",
                  "Signature 7", "Signature 8", "Signature 9", "Signature 10", "Signature 11", "Signature 12", "Signature 13", 
                  "Signature 14", "Signature 15", "Signature 16", "Signature 17", "Signature 18", "Signature 19", "Signature 20", 
                  "Signature 21", "Signature R1", "Signature R2", "Signature R3", "Signature U1", "Signature U2", "No Signature"
            )
        )
    }
    
    message(paste0("Sample ", phyloTree@patientID, ": mutational signatures identified successfully!"))
    treeMSOutput <- list(
        sigsInput=sigsInput, 
        mutSigsOutput=mutSigsOutput, 
        df.aetiology=df.aetiology, 
        patientID = patientID)
    return(treeMSOutput)
} 

## Provide a summary data frame for signatures in different branches.
doMutSigSummary <- function(treeMSOutput){
    mutSigsOutput <- treeMSOutput$mutSigsOutput
    df.aetiology <- treeMSOutput$df.aetiology
    
    ## list of branch names
    ls.branchesName <- mutSigsOutput$branch
    ls.aeti <- c()
    
    for (branch in ls.branchesName) {
        signature <- as.character(mutSigsOutput[which(mutSigsOutput$branch == branch), ]$sig)
        aetiology <- as.character(df.aetiology[which(df.aetiology$sig == signature), ]$aeti)
        ls.aeti <- c(ls.aeti, aetiology)
    }
    
    ## rearrange the order of columns
    if (is.null(mutSigsOutput$selected_genes)) {
        mutSigsOutput <- data.frame(branch=mutSigsOutput$branch,
                                    alias=mutSigsOutput$alias, 
                                    mut.count=mutSigsOutput$mut.count, 
                                    sig=mutSigsOutput$sig, 
                                    sig.prob=mutSigsOutput$sig.prob, 
                                    aeti=ls.aeti)
        colnames(mutSigsOutput) <- c("Branch", "Alias", "Mutation number", "Signature", "Signature weight", "Aetiology")
    } else {
        mutSigsOutput <- data.frame(branch=mutSigsOutput$branch,
                                    alias=mutSigsOutput$alias, 
                                    mut.count=mutSigsOutput$mut.count, 
                                    sig=mutSigsOutput$sig, 
                                    sig.prob=mutSigsOutput$sig.prob,
                                    aeti=ls.aeti,
                                    selected_genes=mutSigsOutput$selected_genes)
        colnames(mutSigsOutput) <- c("Branch", "Alias", "Mutation number", "Signature", "Signature weight", "Aetiology", "Selected gene")
    }
    
    return(mutSigsOutput)
}


## Visualize the mutational signatures of trunk/branches in a phylogenetic tree.
doPlotMutSig <- function(tree.mutSig) {
    sigsInput <- tree.mutSig$sigsInput
    mutSigsOutput <- tree.mutSig$mutSigsOutput
    df.aetiology <- tree.mutSig$df.aetiology
    
    ## calculate the Mutation Probability
    sigsInputSum <- as.data.frame(apply(sigsInput, 1, function(x) sum(x)))
    sigsInputTrans <- as.data.frame(t(sigsInput))
    ls.branchesName <- as.character(rownames(sigsInput))
    ls.mutationType <-as.character(rownames(sigsInputTrans))
    ls.mutationGroup <- c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G")
    sigsInputBranches <- data.frame()
    
    ## calculation process(maybe could be replaced by lapply)
    for (branch in ls.branchesName) {
        sigsInputTrans[branch] <- sigsInputTrans[branch]/sigsInputSum[branch, ]
        signature <- as.character(mutSigsOutput[which(mutSigsOutput$branch == branch), ]$sig)
        alias <- as.character(mutSigsOutput[which(mutSigsOutput$branch == branch), ]$alias)
        sigsWeight <- mutSigsOutput[which(mutSigsOutput$branch == branch), ]$sig.prob
        aetiology <- as.character(df.aetiology[which(df.aetiology$sig == signature), ]$aeti)
        sigsInputBranch <- data.frame(sigsInputTrans[branch], 
                                      rep(branch, length(sigsInputTrans[branch])), 
                                      rep(alias, length(sigsInputTrans[branch])), 
                                      rep(signature, length(sigsInputTrans[branch])), 
                                      rep(sigsWeight, length(sigsInputTrans[branch])), 
                                      rep(aetiology , length(sigsInputTrans[branch])), 
                                      stringsAsFactors = FALSE)
        colnames(sigsInputBranch) <- c("Mutation_Probability", "Branch", "Alias", "Signature", "SigsWeight", "Aetiology")
        sigsInputBranches <- rbind(sigsInputBranches, sigsInputBranch)
    }
    
    df.sigsInputTrans <- data.frame(Mutational_Type=ls.mutationType, 
                                    Group=ls.mutationType, 
                                    sigsInputBranches, 
                                    stringsAsFactors = FALSE)
    
    ## generate Mutation Type for every column
    for (mutationGroup in ls.mutationGroup) {
        df.sigsInputTrans$Group[which(grepl(mutationGroup, df.sigsInputTrans$Group))] <- mutationGroup
    }
    
    ## specific the label order of x axis
    orderlist <- c(ls.mutationType)
    df.sigsInputTrans <- transform(df.sigsInputTrans, Mutational_Type = factor(Mutational_Type, levels = orderlist))
    df.sigsInputTrans <- df.sigsInputTrans[which(df.sigsInputTrans$Signature != "No Signature"), ]
    if(nrow(df.sigsInputTrans) == 0){
        warning(paste0("Patient ", tree.mutSig$patientID, ": There is no enough eligible mutations can be used."))
        return(NA)
    }
    df.sigsInputText <- dplyr::distinct(df.sigsInputTrans, Branch, .keep_all = TRUE)
    
    CA <- grid::textGrob(expression(bold("C > A")),
                         gp=grid::gpar(fontsize=7, fontface="bold"), vjust=0,hjust=1)
    CG <- grid::textGrob(expression(bold("C > G")),
                         gp=grid::gpar(fontsize=7, fontface="bold"), vjust=0,hjust=1)
    CT <- grid::textGrob(expression(bold("C > T")),
                         gp=grid::gpar(fontsize=7, fontface="bold"), vjust=0,hjust=1)
    TA <- grid::textGrob(expression(bold("T > A")),
                         gp=grid::gpar(fontsize=7, fontface="bold"), vjust=0,hjust=1)
    TC <- grid::textGrob(expression(bold("T > C")),
                         gp=grid::gpar(fontsize=7, fontface="bold"), vjust=0,hjust=1)
    TG <- grid::textGrob(expression(bold("T > G")),
                         gp=grid::gpar(fontsize=7, fontface="bold"), vjust=0,hjust=1)
    
    group.colors <- c("#E64B35FF", "#4DBBD5FF", "#00A087FF",
                      "#3C5488FF", "#F39B7FFF", "#8491B4FF")
    pic <- ggplot(df.sigsInputTrans, 
                  aes(x=Mutational_Type, y=Mutation_Probability, group=Group, fill=Group)
    ) + 
        geom_bar(stat="identity")+ 
        theme(panel.grid=element_blank(), 
              panel.border=element_blank(), 
              panel.background = element_blank(), 
              legend.position='none', 
              # axis.text.x=element_text(size=3, angle = 45, hjust = 1, vjust = 1), 
              plot.title = element_text(size = 13, face = "bold", hjust = 0.5,vjust = 0),
              axis.text.x=element_blank(), 
              axis.ticks.x=element_blank(),
              axis.line.y = element_blank(),
              axis.ticks.length.y = unit(0.2, "cm"),
              axis.text.y=element_text(size=6, color = "black")
              ) +
        annotate(geom = "segment", x=-1, xend = -1, y = 0, yend = 0.2, size = 0.6)+
        ## background colors
        geom_rect(aes(xmin=0, xmax=16.5, ymin=0, ymax=Inf),
                  fill="#fce7e4", alpha=0.15) + 
        geom_rect(aes(xmin=16.5, xmax=32.5, ymin=0, ymax=Inf),
                  fill="#ecf8fa", alpha=0.25) + 
        geom_rect(aes(xmin=32.5, xmax=48.5, ymin=0, ymax=Inf),
                  fill="#dbfff9", alpha=0.05) + 
        geom_rect(aes(xmin=48.5, xmax=64.5, ymin=0, ymax=Inf),
                  fill="#e4e8f3", alpha=0.08) + 
        geom_rect(aes(xmin=64.5, xmax=80.5, ymin=0, ymax=Inf),
                  fill="#fdefeb", alpha=0.15) + 
        geom_rect(aes(xmin=80.5, xmax=96.5, ymin=0, ymax=Inf),
                  fill="#e5e8ef", alpha=0.1) + 
        ## barplot
        geom_bar(stat="identity") + 
        ## combine different results
        facet_grid(Alias ~ .) + 
        ## color setting
        scale_fill_manual(values=group.colors) + 
        ## axis setting
        xlab("Mutational type") + 
        ylab("Mutation probability") + 
        ggtitle(paste0("Mutational signatures of ",tree.mutSig$patientID,"'s phylogenetic tree ") )+
        scale_y_continuous(limits=c(-0.03, 0.2), breaks=seq(0, 0.2, 0.1)) + 
        ## signature notes and text parts
        geom_text(data = df.sigsInputText, 
                  aes(x=-Inf, y=Inf, label=paste(Signature, ": ", 
                                                 round(as.numeric(levels(SigsWeight)[SigsWeight]), 3), 
                                                 "    ", 
                                                 "Aetiology: ", Aetiology, sep="")), 
                  hjust = -0.02, vjust = 1.5, colour="#2B2B2B", fontface = "bold", size=3) + 
        ## Mutational Type Labels
        annotation_custom(grob = CA,  xmin = 10, xmax = 10, ymin = -0.065, ymax = -0) + 
        annotation_custom(grob = CG,  xmin = 27, xmax = 27, ymin = -0.065, ymax = -0) + 
        annotation_custom(grob = CT,  xmin = 43, xmax = 43, ymin = -0.065, ymax = -0) + 
        annotation_custom(grob = TA,  xmin = 59, xmax = 59, ymin = -0.065, ymax = -0) + 
        annotation_custom(grob = TC,  xmin = 75, xmax = 75, ymin = -0.065, ymax = -0) + 
        annotation_custom(grob = TG,  xmin = 91, xmax = 91, ymin = -0.065, ymax = -0) + 
        ## x axis bar
        geom_rect(aes(xmin=0, xmax=16.405, ymin=-0.01, ymax=-0.005),
                  fill=group.colors[1], alpha=1) + 
        geom_rect(aes(xmin=16.595, xmax=32.405, ymin=-0.01, ymax=-0.005),
                  fill=group.colors[2], alpha=0.25) + 
        geom_rect(aes(xmin=32.595, xmax=48.405, ymin=-0.01, ymax=-0.005),
                  fill=group.colors[3], alpha=0.05) + 
        geom_rect(aes(xmin=48.595, xmax=64.405, ymin=-0.01, ymax=-0.005),
                  fill=group.colors[4], alpha=0.08) + 
        geom_rect(aes(xmin=64.595, xmax=80.405, ymin=-0.01, ymax=-0.005),
                  fill=group.colors[5], alpha=0.15) + 
        geom_rect(aes(xmin=80.595, xmax=96.5, ymin=-0.01, ymax=-0.005),
                  fill=group.colors[6], alpha=0.1)
    message("Mutational signature plot generation done!")
    return(pic)
}

