#' vafCluster
#' @description Generate variant allele frequency(VAF) frequency distribution
#'  curve in different mode.
#'  
#' @param maf  Maf or MafList object generated by \code{\link{readMaf}} function.
#' @param patient.id Select the specific patients. Default: NULL, all patients are included.
#' @param segFile The segment file.
#' @param min.vaf The minimum value of VAF. Default: 0. Option: on the scale of 0 to 1
#' @param max.vaf The maximum value of VAF. Default: 0. Option: on the scale of 0 to 1
#' @param withinTumor Cluster VAF within tumors in each patients,default is FALSE.
#' @param ... Other options passed to \code{\link{subMaf}}
#' 
#' @examples
#' vafCluster(maf, plotOption="compare")
#' vafCluster(maf, plotOption="combine")
#' 
#'
#' @return clustering plots of vaf
#' @export vafCluster

## Main function for VAF plot
vafCluster <-function(maf,
                      patient.id = NULL,
                      segFile = NULL,
                      min.vaf = 0.02,
                      max.vaf = 1,
                      withinTumor = FALSE,
                      ...){
  # plotOption <- match.arg(plotOption, choices = c('combine', 'compare'), several.ok = FALSE)
  ## check input data
  maf_list <- checkMafInput(maf, patient.id = patient.id)
  
  result <- list()
  for(m in maf_list){
      
      ## remove mutation in CNA regions
      if(!is.null(segFile)){
          seg <- readSegment(segFile = segFile)
          seg <- seg[!Chromosome %in% c("X","Y")]
          m <- copyNumberFilter(m,seg)
      }
      
      maf_data <- subMaf(m,
                            min.vaf = min.vaf,
                            max.vaf = max.vaf,
                            ...)
      patient <- getMafPatient(m)
      if(nrow(maf_data) == 0){
            message("Warning :there was no mutation in ", patient, " after filtering.")
            next
      }
      
      ## extract vaf info
      if(withinTumor){
          maf_data$VAF <- maf_data$Tumor_Average_VAF
          maf_data$ID <- maf_data$Tumor_ID
          id_list <- unique(maf_data$ID)
      }else{
          maf_data$ID <- maf_data$Tumor_Sample_Barcode
          id_list <- unique(maf_data$ID)
      }
      
      ## general data process for all samples 
      plot_list <- list()
      
      rebuild_data <- data.frame()
      for (id in id_list){
          subdata <- maf_data[ID == id]  
          ## data cleaning
          if (nrow(subdata) < 3) {
              message(paste0("Error: mutations of Sample ", id, " are not enough for clustering"))
              next()
          }
          
          ## infer possible cluster from maf_data
          message(paste("Processing ", id," of ", patient, sep = ""))
          cluster_result <- mclust::densityMclust(subdata$VAF, G=seq_len(7), verbose=FALSE)
          subdata$cluster <- as.character(cluster_result$classification)
          
          ## define outfilter
          out_vaf <- boxplot.stats(subdata$VAF)$out
          subdata[subdata$VAF %in% out_vaf]$cluster <- "outlier"
          
          rebuild_data <- rbind(rebuild_data, subdata)
          
          # prepare separated pictures for later combination 
          p  <- drawVAFCombine(subdata)
          # if(length(id_list) == 1){
          #         p <- p + ggtitle(label = paste("VAF clustering of ", patient, sep=""),
          #                          subtitle = id)+
          #         theme(plot.title = element_text(face = "bold",size = 18,hjust = 0.5),
          #               plot.subtitle = element_text(size = 16,hjust = 1,vjust = -5))
          # }
          plot_list[[id]] <- p
      }
      
      if(nrow(rebuild_data) == 0){
        next
      }
      
      rebuild_data <-  rebuild_data %>% 
          dplyr::select(Patient_ID, Tumor_Sample_Barcode, Hugo_Symbol,
                        Chromosome, Start_Position, End_Position, cluster) %>% 
          dplyr::arrange(cluster) %>% 
          dplyr::rename(Cluster = cluster)
      
      ## combine: print VAF pictures for all samples in one document
      # pic <- cowplot::plot_grid(plotlist = plot_list,ncol = 2)
      
      # combineTitle <-  ggdraw() + draw_label(
      #     paste("VAF clustering of ", patient, sep=""),
      #     fontface = 'bold',
      #     x = 0.5,
      #     hjust = 0.5,
      #     size = 13.5
      # )
      # pic <- cowplot::plot_grid(combineTitle,
      #                           pic,
      #                           ncol = 1,
      #                           # rel_heights values control vertical title margins
      #                            rel_heights = c(0.1, 1))
      result[[patient]] <- list(cluster.data = rebuild_data, cluster.plot = plot_list) 
      
  }
  if(length(result) > 1){
      return(result)
  }else if(length(result) == 0){
      return(NA)
  }else{
      return(result[[1]])
  }
  
}




# ## plot all samples' vaf distribution with ggridges
# else if (plotOption == "compare"){
#     all_cluster_result <- c()
#     rebuild_data <- data.frame()
#     for (id in id_list){
#         subdata <- maf_data[ID == id]  
#         if (nrow(subdata) < 3) {
#             message(paste0("Error: mutations of Sample ", id, " are not enough for clustering"))
#             maf_data <- maf_data[ID != id]
#             next
#         }
#         ## generate data from different Tumor_Sample_Barcode
#         message(paste("Processing ", id," of ", patient, sep = ""))
#         cluster_result <- mclust::densityMclust(subdata$VAF, G=1:7, verbose=FALSE)
#         subdata$cluster <- as.character(cluster_result$classification)
#         
#         ## define outlier
#         out_vaf <- boxplot.stats(subdata$VAF)$out
#         subdata[subdata$VAF %in% out_vaf]$cluster <- "outlier"
#         
#         rebuild_data <- rbind(subdata, rebuild_data)
#     }
#     
#     # print(all_cluster_result)
#     maf_data <- rebuild_data
#     pic <- suppressMessages(drawVAFCompare(maf_data, withinTumor = withinTumor))
#     result[[patient]] <- pic
#     next
# }