#' vafCluster
#' @description Generate variant allele frequency(VAF) frequency distribution
#'  curve in different mode.
#'  
#' @param maf a Maf object generated by readMaf function.
#' @param patient.id Select the specific patients. Default: NULL, all patients are included.
#' @param chrSilent Select chromosomes needed to be dismissed. Default NULL.
#' @param mutType select Proper variant classification you need. Default "All". Option: "nonSyn".
#' @param use.indel Logical value. Whether to use INDELs besides somatic SNVs. Default: TRUE.
#' @param segCN.file the segment file.
#' @param min.vaf the minimum value of VAF. Default: 0. Option: on the scale of 0 to 1
#' @param max.vaf the maximum value of VAF. Default: 0. Option: on the scale of 0 to 1
#' @param showMATH logical. Whether to show MATH Scores in the plot. Default TRUE 
#' @param plotOption character. Three options for displaying clustering of VAFs in samples. Default "compare"
#' "compare": prints comparison of VAFs distribution among samples;
#' "combine" prints VAFs distribution of samples all together;
#' @param use.adjVAF cluster by adjusted VAF.Default: FALSE. 
#' @examples
#' vafCluster(maf, plotOption="compare")
#' vafCluster(maf, plotOption="combine", showMATH=FALSE)
#' 
#'
#' @export vafCluster

## Main function for VAF plot
vafCluster <-function(maf,
                      segCN.file = NULL,
                      patient.id = NULL,
                      chrSilent = NULL,
                      mutType = "All",
                      use.indel = TRUE,
                      min.vaf=0.02,
                      max.vaf=1,
                      showMATH=TRUE, 
                      plotOption="combine",
                      use.adjVAF = FALSE){
  plotOption <- match.arg(plotOption, choices = c('combine', 'compare'), several.ok = FALSE)
   
  maf <- subsetMaf(maf,
                   patient.id = patient.id,
                   chrSilent = chrSilent,
                   mutType = mutType,
                   use.indel = use.indel,
                   min.vaf = min.vaf,
                   max.vaf = max.vaf,
                   use.adjVAF = use.adjVAF)
  
  ## remove mutation in CNV region
  if(!is.null(segCN.file)){
    seg <- readSegment(segCN.file = segCN.file)
    maf <- copyNumberFilter(maf, seg)
  }
  
  
  dat.list <- split(maf@data, maf@data$Patient_ID)
  result.list <- lapply(dat.list, doVafCluster,
                        maf = maf,
                        min.vaf = min.vaf,
                        max.vaf = max.vaf,
                        showMATH = showMATH, 
                        plotOption = plotOption,
                        use.adjVAF = use.adjVAF)
  return(result.list)
}